{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For Gene_Data\n",
    "[This](https://dataguide.nlm.nih.gov/eutilities/utilities.html) has been extremely useful, especially the part about efetch and the tables it links (which are very hard to find).\n",
    "\n",
    "We will need two separate Entrez query methods:\n",
    "    1. One for normally annotated species\n",
    "    2. Another for metagenomes of newly added species\n",
    "Both should return an xml which will be parsed differently according the !DOCTYPE tag (**Entrezgene-Set** or **GBSet**)\n",
    "\n",
    "The default setting for *retmax* is 20 which was limiting our geneID results. The maximum value that can be used is 100000. For higher numbers we need to use different approach, found [here](https://www.ncbi.nlm.nih.gov/books/NBK25499/). But I don't think it will be necessary since none of our species will have that many genes.\n",
    "\n",
    "Tags leading to the Gene ID:\n",
    "\n",
    "    <Entrezgene-Set>\n",
    "    <Entrezgene>\n",
    "      <Entrezgene_track-info>\n",
    "        <Gene-track>\n",
    "          <Gene-track_geneid>38094981</Gene-track_geneid>\n",
    "\n",
    "Tags leading to the gene name:\n",
    "\n",
    "    <Entrezgene_gene>\n",
    "        <Gene-ref>\n",
    "            <Gene-ref_locus>ytiA</Gene-ref_locus>\n",
    "\n",
    "Gene description and protein accession:\n",
    "\n",
    "    <Gene-commentary_products>\n",
    "        <Gene-commentary>\n",
    "          <Gene-commentary_type value=\"peptide\">8</Gene-commentary_type>\n",
    "          <Gene-commentary_label>uncharacterized protein YtiA</Gene-commentary_label>\n",
    "          <Gene-commentary_accession>YP_009518833</Gene-commentary_accession>\n",
    "          <Gene-commentary_version>1</Gene-commentary_version>\n",
    "\n",
    "There are several other tags that would be of interest, like chromosome, sequence interval from-to, strand direction.\n",
    "\n",
    "<span style=\"color:red\">There are entries from multiple databases within &lt;Entrezgene&gt;, defining the same gene.</span> Ideally we are to use the reference ones &lt;Gene-ref&gt;, &lt;Prot-ref&gt;, &lt;Entrezgene_locus&gt;."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "Entrez.email = \"A.N.Other@example.com\"  # Always tell NCBI who you are\n",
    "handle = Entrez.esearch(db=\"gene\", term=\"txid511145[Organism]\",retmax=1)\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(record)\n",
    "handle = Entrez.efetch(db=\"gene\", id=record['IdList'],rettype=\"xml\")\n",
    "genes=handle.read()\n",
    "\n",
    "print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For new species/metagenomes\n",
    "Proteins for species, determined from environmental samples are not present in the *gene* NCBI database. However, their protein sequences have been determined and exist in the *protein* database. We can use that to extract gi numbers, symbols, and descriptions.\n",
    "\n",
    "    <GBSeq_definition>Chain A, Uncharacterized protein</GBSeq_definition>\n",
    "    <GBSeq_primary-accession>5YEE_A</GBSeq_primary-accession>\n",
    "    <GBSeq_accession-version>5YEE_A</GBSeq_accession-version>\n",
    "    <GBSeq_other-seqids>\n",
    "      <GBSeqid>pdb|5YEE|A</GBSeqid>\n",
    "      <GBSeqid>gi|1482416786</GBSeqid>\n",
    "    </GBSeq_other-seqids>           <----- gi is more important than the tag in htis case\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Entrez' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-3a37db93731b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mesearch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdb\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"protein\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mterm\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"txid1538547[Organism]\"\u001b[0m\u001b[0;34m)\u001b[0m   \u001b[0;31m#Lokiarchaeum\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mrecord\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mhandle\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mhandle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mEntrez\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mefetch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdb\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"protein\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mid\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrecord\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'IdList'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mrettype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"gp\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mretmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"xml\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Entrez' is not defined"
     ]
    }
   ],
   "source": [
    "handle = Entrez.esearch(db=\"protein\", term=\"txid1538547[Organism]\")   #Lokiarchaeum\n",
    "record = Entrez.read(handle)\n",
    "handle.close()\n",
    "print(record)\n",
    "handle = Entrez.efetch(db=\"protein\", id=record['IdList'],rettype=\"gp\", retmode=\"xml\")\n",
    "genes=handle.read()\n",
    "\n",
    "print(genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene_Data table \n",
    "This table contains information regarding the GeneID, StrainID, the nomgd_ID(which is the nomenclature ID generated in the nomenclature table), gene symbol and gene description. It gets the strain ID from the species table and the gene information such as the gene ID, gene symbol and gene description from NCBI database.\n",
    "\n",
    "We need to build functions for cases where there are no hits for the query i.e., if we don't find any gene information for a particular strain ID.\n",
    "\n",
    "There are three strainIDs that do not have any entries in the gene database for bacteria. So, these three strain IDs showed up in the output as no records exist :\n",
    "\n",
    "Record doesn't exist 165597 - Crocosphaera watsonii WH 8501;\n",
    "Record doesn't exist 272627 - Magnetospirillum magnetotacticum MS-1;\n",
    "Record doesn't exist 575564 - Acinetobacter sp. RUH2624."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "User name: vchivukula\n",
      "Password: ········\n",
      "Record doesn't exist 165597\n",
      "Record doesn't exist 272627\n",
      "Record doesn't exist 575564\n"
     ]
    }
   ],
   "source": [
    "#! /usr/bin/env python3\n",
    "\n",
    "from Bio import Entrez\n",
    "from xml.dom.minidom import parse, parseString\n",
    "import xml.dom.minidom as minidom\n",
    "import subprocess\n",
    "import re\n",
    "import mysql.connector\n",
    "import getpass\n",
    "\n",
    "uname = input(\"User name: \")\n",
    "pw = getpass.getpass(\"Password: \")\n",
    "cnx = mysql.connector.connect(user=uname, password=pw, host='130.207.36.75', database='DESIRE')\n",
    "cursor = cnx.cursor()\n",
    "\n",
    "def get_strainid():\n",
    "\t'''\n",
    "\tfunction to get a list of strain IDs from the TaxGroup table\n",
    "\t'''\n",
    "\tstrainID_list = []\n",
    "\tcursor.execute(\"SELECT DESIRE.Species.strain_id FROM DESIRE.Species_TaxGroup\\\n",
    "\t\t\t\tINNER JOIN DESIRE.TaxGroups ON DESIRE.Species_TaxGroup.taxgroup_id=DESIRE.TaxGroups.taxgroup_id\\\n",
    "\t\t\t\tINNER JOIN DESIRE.Species ON DESIRE.Species_TaxGroup.strain_id=DESIRE.Species.strain_id\\\n",
    "\t\t\t\tWHERE DESIRE.TaxGroups.groupName = 'Bacteria'\")\n",
    "\tresults = cursor.fetchall()\n",
    "\tfor i in results:\n",
    "\t\tstrainID_list.append(int(i[0]))\n",
    "\t#print(len(strainID_list))\n",
    "\treturn strainID_list\n",
    "\n",
    "def superkingdom_info(ID):\n",
    "\t'''\n",
    "\tfunction to get the superkingdoms for the strain IDs\n",
    "\t'''\n",
    "\t#print(ID)\n",
    "\tcursor.execute(\"SELECT DESIRE.TaxGroups.groupName FROM DESIRE.Species_TaxGroup\\\n",
    "\t\tINNER JOIN DESIRE.TaxGroups ON DESIRE.Species_TaxGroup.taxgroup_id=DESIRE.TaxGroups.taxgroup_id\\\n",
    "\t\tINNER JOIN DESIRE.Species ON DESIRE.Species_TaxGroup.strain_id=DESIRE.Species.strain_id\\\n",
    "\t\tWHERE DESIRE.TaxGroups.groupLevel = 'superkingdom' AND DESIRE.Species.strain_id = '\"+ID+\"'\")\n",
    "\tresults = cursor.fetchall()\n",
    "\tsuperkingdom=(results[0][0])\n",
    "\t#print(superkingdom)\n",
    "\treturn superkingdom\n",
    "\n",
    "def nomid_oldname(domain_letter):\n",
    "\t'''\n",
    "\tfunction to get the nomo_id and old_name for the superkingdom obtained from superkingdom_info function\n",
    "\t'''\n",
    "\tcursor.execute(\"SELECT DESIRE.Old_name.nomo_id, DESIRE.Old_name.old_name FROM DESIRE.Old_name\\\n",
    "\t\tINNER JOIN DESIRE.Nomenclature ON DESIRE.Old_name.nomo_id=DESIRE.Nomenclature.nom_id\\\n",
    "\t\tWHERE DESIRE.Nomenclature.occurrence = '\"+domain_letter+\"' AND DESIRE.Old_name.N_B_Y_H_A = '\"+domain_letter+\"'\")\n",
    "\toldname = cursor.fetchall()\n",
    "\t# print(oldname)\n",
    "\treturn oldname\n",
    "\n",
    "def gene_info(taxid):\n",
    "\t'''\n",
    "\tfunction to get gene_id, gene_symbol, and gene_desc for the Gene_Data table using the strain ID\n",
    "\t'''\n",
    "\t#print(taxid)\n",
    "\tgeneinfo_output = subprocess.check_output(\"esearch -db gene -query 'txid\"+taxid+\"[Organism] ribosomal' | efetch -format tabular\", shell=True)\n",
    "\t#print(geneinfo_output)\n",
    "\tgene_out = geneinfo_output.decode('utf-8')\n",
    "\t# print(gene_out)\n",
    "\tif re.match(\".*https\\:\\/\\/eutils\\.st-va\\.ncbi\\.nlm\\.nih\\.gov\\.*\", gene_out):\n",
    "\t\tprint(\"There is an error\")\n",
    "\t\tpass\n",
    "\telif re.match(\".*https\\:\\/\\/eutils\\.be-md\\.ncbi\\.nlm\\.nih\\.gov\\.*\", gene_out):\n",
    "\t\tprint(\"There is an error\")\n",
    "\t\tpass\n",
    "\telse:\n",
    "\t\t# print(gene_out)\n",
    "\t\tgene_out = gene_out.split('\\n')\n",
    "\t\tregex = re.compile('^tax_id\\\\tOrg_name\\\\tGeneID\\\\t.*')\n",
    "\t\tgene_out = [i for i in gene_out if not regex.match(i)] \n",
    "\t\tregex = re.compile('.*NEWENTRY.*')\n",
    "\t\tgene_out = [i for i in gene_out if not regex.match(i)]\n",
    "\t\t# print(gene_out)\n",
    "\t\tgenesym_desc=[]\n",
    "\t\tfor item in gene_out[:-2]:\n",
    "\t\t\tgene_list=item.split('\\t')\n",
    "\t\t\tif gene_list[7] == 'pseudo' or gene_list[7] == 'hypothetical protein':\n",
    "\t\t\t\tpass\n",
    "\t\t\telse:\n",
    "\t\t\t\tgeneinfo = (gene_list[2], gene_list[5], gene_list[7]) # gene_id, gene_symbol, gene_desc\n",
    "\t\t\t\t# print(geneinfo)\n",
    "\t\t\t\tgenesym_desc.append(geneinfo)\n",
    "\t\t#print(genesym_desc)\n",
    "\t\treturn genesym_desc\n",
    "\n",
    "def upload_data(name, desc, strainID):\n",
    "\tfor i in name:\n",
    "\t\told_name = i[1]\n",
    "\t\tnom_id=i[0]\n",
    "\t\tfor j in desc:\n",
    "\t\t\tword = j[2]\n",
    "\t\t\tregexp = re.compile(r''+old_name+'$')\n",
    "\t\t\tregexp_RNA = re.compile(r'ribosomal RNA$')\n",
    "\t\t\tif regexp.search(word): \n",
    "\t\t\t\tquery = (\"INSERT INTO `DESIRE`.`Gene_Data`(`GI`,`strain_ID`,`nomgd_id`, `GeneSymbol`, `GeneDescription`) VALUES('\"+str(j[0])+\"','\"+str(strainID)+\"','\"+str(i[0])+\"','\"+str(j[1])+\"','\"+str(j[2])+\"')\")\n",
    "\t\t\t\tcursor.execute(query)\n",
    "\t\t\t\t#print(query)\n",
    "\t\t\tif regexp_RNA.search(word):\n",
    "\t\t\t\tquery = (\"INSERT INTO `DESIRE`.`Gene_Data`(`GI`,`strain_ID`, `GeneSymbol`, `GeneDescription`) VALUES('\"+str(j[0])+\"','\"+str(strainID)+\"','\"+str(j[1])+\"','\"+str(j[2])+\"')\")\n",
    "\t\t\t\tcursor.execute(query)\n",
    "\t\t\t\t#print(query)\n",
    "\n",
    "\n",
    "strainID_list=get_strainid()\n",
    "\t\n",
    "for strainID in strainID_list:\n",
    "\tgenesym_desc=gene_info(str(strainID))\n",
    "\t\n",
    "\tif len(genesym_desc) == 0:\n",
    "\t\tpass #change this - use another function to access protein db rather than gene db\n",
    "\t\tprint(\"Record doesn't exist\", strainID)\n",
    "\n",
    "\tsuperkingdom=superkingdom_info(str(strainID))\n",
    "\tnom_oldList = nomid_oldname(str(superkingdom[0]))\n",
    "\tupload_data(nom_oldList, genesym_desc, strainID)\n",
    "cnx.commit()\n",
    "cursor.close()\n",
    "cnx.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
