<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: Free Energy Evaluation</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__eval.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#groups">Modules</a> &#124;
<a href="#files">Files</a> &#124;
<a href="#define-members">Macros</a>  </div>
  <div class="headertitle">
<div class="title">Free Energy Evaluation</div>  </div>
</div><!--header-->
<div class="contents">

<p>Functions and variables related to free energy evaluation of sequence/structure pairs.  
<a href="#details">More...</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Functions and variables related to free energy evaluation of sequence/structure pairs. </p>
<p>Several different functions to evaluate the free energy of a particular secondary structure under a particular set of parameters and the Nearest Neighbor Energy model are available. For most of them, two different forms of representations for the secondary structure may be used:</p><ul>
<li>The Dot-Bracket string</li>
<li>A pair table representation</li>
</ul>
<p>Furthermore, the evaluation functions are divided into <code>basic</code> and <code>simplified</code> variants, where <code>basic</code> functions require the use of a <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure holding the sequence string, and model configuration (settings and parameters). The <code>simplified</code> functions, on the other hand, provide often used default model settings that may be called directly with only sequence and structure data.</p>
<p>Finally, <code>verbose</code> options exist for some functions that allow one to print the (individual) free energy contributions to some <code>FILE</code> stream. </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Free Energy Evaluation:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__eval.svg" width="478" height="410"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="groups"></a>
Modules</h2></td></tr>
<tr class="memitem:group__eval__loops"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval__loops.html">Energy Evaluation for Individual Loops</a></td></tr>
<tr class="memdesc:group__eval__loops"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions to evaluate the free energy of particular types of loops. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:group__eval__move"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval__move.html">Energy Evaluation for Atomic Moves</a></td></tr>
<tr class="memdesc:group__eval__move"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions to evaluate the free energy change of a structure after application of (a set of) atomic moves. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:group__eval__deprecated"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval__deprecated.html">Deprecated Interface for Free Energy Evaluation</a></td></tr>
<tr class="memdesc:group__eval__deprecated"><td class="mdescLeft">&#160;</td><td class="mdescRight">Deprecated Energy Evaluation functions. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:eval_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="eval_8h.html">eval.h</a></td></tr>
<tr class="memdesc:eval_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions and variables related to energy evaluation of sequence/structure pairs. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:all_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="all_8h.html">all.h</a></td></tr>
<tr class="memdesc:all_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Energy evaluation for MFE and partition function calculations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:external_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="external_8h.html">external.h</a></td></tr>
<tr class="memdesc:external_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Energy evaluation of exterior loops for MFE and partition function calculations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:hairpin_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="hairpin_8h.html">hairpin.h</a></td></tr>
<tr class="memdesc:hairpin_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Energy evaluation of hairpin loops for MFE and partition function calculations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:internal_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="internal_8h.html">internal.h</a></td></tr>
<tr class="memdesc:internal_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Energy evaluation of interior loops for MFE and partition function calculations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:multibranch_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="multibranch_8h.html">multibranch.h</a></td></tr>
<tr class="memdesc:multibranch_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Energy evaluation of multibranch loops for MFE and partition function calculations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:gaf4afe19780b61b4962c613bde324128b"><td class="memItemLeft" align="right" valign="top"><a id="gaf4afe19780b61b4962c613bde324128b"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b">VRNA_VERBOSITY_QUIET</a>&#160;&#160;&#160;-1</td></tr>
<tr class="memdesc:gaf4afe19780b61b4962c613bde324128b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Quiet level verbosity setting. <br /></td></tr>
<tr class="separator:gaf4afe19780b61b4962c613bde324128b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga47430d9e875084cfb983b22612e3abdf"><td class="memItemLeft" align="right" valign="top"><a id="ga47430d9e875084cfb983b22612e3abdf"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga47430d9e875084cfb983b22612e3abdf">VRNA_VERBOSITY_DEFAULT</a>&#160;&#160;&#160;1</td></tr>
<tr class="memdesc:ga47430d9e875084cfb983b22612e3abdf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Default level verbosity setting. <br /></td></tr>
<tr class="separator:ga47430d9e875084cfb983b22612e3abdf"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Basic Energy Evaluation Interface with Dot-Bracket Structure String</h2></td></tr>
<tr class="memitem:ga58f199f1438d794a265f3b27fc8ea631"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631">vrna_eval_structure</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *structure)</td></tr>
<tr class="memdesc:ga58f199f1438d794a265f3b27fc8ea631"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631">More...</a><br /></td></tr>
<tr class="separator:ga58f199f1438d794a265f3b27fc8ea631"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga6cea75c0eb9857fb59172be54cab09e0"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0">vrna_eval_covar_structure</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *structure)</td></tr>
<tr class="memdesc:ga6cea75c0eb9857fb59172be54cab09e0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.  <a href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0">More...</a><br /></td></tr>
<tr class="separator:ga6cea75c0eb9857fb59172be54cab09e0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga0928d699d310178f84ee2351034e5cb5"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5">vrna_eval_structure_verbose</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *structure, FILE *file)</td></tr>
<tr class="memdesc:ga0928d699d310178f84ee2351034e5cb5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA and print contributions on a per-loop base.  <a href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5">More...</a><br /></td></tr>
<tr class="separator:ga0928d699d310178f84ee2351034e5cb5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab12e6b1226227670322150df018734f8"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gab12e6b1226227670322150df018734f8">vrna_eval_structure_v</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gab12e6b1226227670322150df018734f8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA and print contributions on a per-loop base.  <a href="group__eval.html#gab12e6b1226227670322150df018734f8">More...</a><br /></td></tr>
<tr class="separator:gab12e6b1226227670322150df018734f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1d49318a535422b723e2c47520ac090a"><td class="memItemLeft" align="right" valign="top"><a id="ga1d49318a535422b723e2c47520ac090a"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><b>vrna_eval_structure_cstr</b> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *structure, int verbosity_level, vrna_cstr_t output_stream)</td></tr>
<tr class="separator:ga1d49318a535422b723e2c47520ac090a"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Basic Energy Evaluation Interface with Structure Pair Table</h2></td></tr>
<tr class="memitem:gadbd09372ddfd7a450bbd590c96a6bfe4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4">vrna_eval_structure_pt</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const short *pt)</td></tr>
<tr class="memdesc:gadbd09372ddfd7a450bbd590c96a6bfe4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4">More...</a><br /></td></tr>
<tr class="separator:gadbd09372ddfd7a450bbd590c96a6bfe4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8a517cfeeae8c376ae7b1e0c401d38b4"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4">vrna_eval_structure_pt_verbose</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const short *pt, FILE *file)</td></tr>
<tr class="memdesc:ga8a517cfeeae8c376ae7b1e0c401d38b4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4">More...</a><br /></td></tr>
<tr class="separator:ga8a517cfeeae8c376ae7b1e0c401d38b4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2c6533ba0afe4c88d335d8f1e0e2a48e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga2c6533ba0afe4c88d335d8f1e0e2a48e">vrna_eval_structure_pt_v</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const short *pt, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:ga2c6533ba0afe4c88d335d8f1e0e2a48e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga2c6533ba0afe4c88d335d8f1e0e2a48e">More...</a><br /></td></tr>
<tr class="separator:ga2c6533ba0afe4c88d335d8f1e0e2a48e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Simplified Energy Evaluation with Sequence and Dot-Bracket Strings</h2></td></tr>
<tr class="memitem:ga7e5273464b775d4130245681312c1369"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369">vrna_eval_structure_simple</a> (const char *string, const char *structure)</td></tr>
<tr class="memdesc:ga7e5273464b775d4130245681312c1369"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga7e5273464b775d4130245681312c1369">More...</a><br /></td></tr>
<tr class="separator:ga7e5273464b775d4130245681312c1369"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3e05a23ddf9b083f4e69881e440d4866"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866">vrna_eval_circ_structure</a> (const char *string, const char *structure)</td></tr>
<tr class="memdesc:ga3e05a23ddf9b083f4e69881e440d4866"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a sequence/structure pair where the sequence is circular.  <a href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866">More...</a><br /></td></tr>
<tr class="separator:ga3e05a23ddf9b083f4e69881e440d4866"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3263504825ef4b523eba797c99921df4"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4">vrna_eval_gquad_structure</a> (const char *string, const char *structure)</td></tr>
<tr class="memdesc:ga3263504825ef4b523eba797c99921df4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.  <a href="group__eval.html#ga3263504825ef4b523eba797c99921df4">More...</a><br /></td></tr>
<tr class="separator:ga3263504825ef4b523eba797c99921df4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga9dba2fc5d7e6ad1359a7c2f350589c0e"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga9dba2fc5d7e6ad1359a7c2f350589c0e">vrna_eval_circ_gquad_structure</a> (const char *string, const char *structure)</td></tr>
<tr class="memdesc:ga9dba2fc5d7e6ad1359a7c2f350589c0e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a sequence/structure pair where the sequence is circular and the structure may contain G-Quadruplexes.  <a href="group__eval.html#ga9dba2fc5d7e6ad1359a7c2f350589c0e">More...</a><br /></td></tr>
<tr class="separator:ga9dba2fc5d7e6ad1359a7c2f350589c0e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf928bfd96767e1b8033a95a4cc432e39"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gaf928bfd96767e1b8033a95a4cc432e39">vrna_eval_structure_simple_verbose</a> (const char *string, const char *structure, FILE *file)</td></tr>
<tr class="memdesc:gaf928bfd96767e1b8033a95a4cc432e39"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA and print contributions per loop.  <a href="group__eval.html#gaf928bfd96767e1b8033a95a4cc432e39">More...</a><br /></td></tr>
<tr class="separator:gaf928bfd96767e1b8033a95a4cc432e39"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacd6278343e77d13f1d53588e50d303bc"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc">vrna_eval_structure_simple_v</a> (const char *string, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gacd6278343e77d13f1d53588e50d303bc"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA and print contributions per loop.  <a href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc">More...</a><br /></td></tr>
<tr class="separator:gacd6278343e77d13f1d53588e50d303bc"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac3fb44e0773a51be8efc5f4f595a94a7"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gac3fb44e0773a51be8efc5f4f595a94a7">vrna_eval_circ_structure_v</a> (const char *string, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gac3fb44e0773a51be8efc5f4f595a94a7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate free energy of a sequence/structure pair, assume sequence to be circular and print contributions per loop.  <a href="group__eval.html#gac3fb44e0773a51be8efc5f4f595a94a7">More...</a><br /></td></tr>
<tr class="separator:gac3fb44e0773a51be8efc5f4f595a94a7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaeaa2bdbc1b5d78c667e735fbdff87fff"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gaeaa2bdbc1b5d78c667e735fbdff87fff">vrna_eval_gquad_structure_v</a> (const char *string, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gaeaa2bdbc1b5d78c667e735fbdff87fff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate free energy of a sequence/structure pair, allow for G-Quadruplexes in the structure and print contributions per loop.  <a href="group__eval.html#gaeaa2bdbc1b5d78c667e735fbdff87fff">More...</a><br /></td></tr>
<tr class="separator:gaeaa2bdbc1b5d78c667e735fbdff87fff"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab96a6c59923ff06c35f8c2fd2c239727"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gab96a6c59923ff06c35f8c2fd2c239727">vrna_eval_circ_gquad_structure_v</a> (const char *string, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gab96a6c59923ff06c35f8c2fd2c239727"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate free energy of a sequence/structure pair, assume sequence to be circular, allow for G-Quadruplexes in the structure, and print contributions per loop.  <a href="group__eval.html#gab96a6c59923ff06c35f8c2fd2c239727">More...</a><br /></td></tr>
<tr class="separator:gab96a6c59923ff06c35f8c2fd2c239727"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Simplified Energy Evaluation with Sequence Alignments and Consensus Structure Dot-Bracket String</h2></td></tr>
<tr class="memitem:ga7762c3a7bdcbc3a14ef93259d322c7d6"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6">vrna_eval_consensus_structure_simple</a> (const char **alignment, const char *structure)</td></tr>
<tr class="memdesc:ga7762c3a7bdcbc3a14ef93259d322c7d6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA sequence alignment.  <a href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6">More...</a><br /></td></tr>
<tr class="separator:ga7762c3a7bdcbc3a14ef93259d322c7d6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac96577cf232c71160f762737a994b7c6"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gac96577cf232c71160f762737a994b7c6">vrna_eval_circ_consensus_structure</a> (const char **alignment, const char *structure)</td></tr>
<tr class="memdesc:gac96577cf232c71160f762737a994b7c6"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequences are circular.  <a href="group__eval.html#gac96577cf232c71160f762737a994b7c6">More...</a><br /></td></tr>
<tr class="separator:gac96577cf232c71160f762737a994b7c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf09a326b3d57a4b30c27bd0e216198ac"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gaf09a326b3d57a4b30c27bd0e216198ac">vrna_eval_gquad_consensus_structure</a> (const char **alignment, const char *structure)</td></tr>
<tr class="memdesc:gaf09a326b3d57a4b30c27bd0e216198ac"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the structure may contain G-Quadruplexes.  <a href="group__eval.html#gaf09a326b3d57a4b30c27bd0e216198ac">More...</a><br /></td></tr>
<tr class="separator:gaf09a326b3d57a4b30c27bd0e216198ac"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac673ebb9ae2a29f54d201e2ac5b85540"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gac673ebb9ae2a29f54d201e2ac5b85540">vrna_eval_circ_gquad_consensus_structure</a> (const char **alignment, const char *structure)</td></tr>
<tr class="memdesc:gac673ebb9ae2a29f54d201e2ac5b85540"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence is circular and the structure may contain G-Quadruplexes.  <a href="group__eval.html#gac673ebb9ae2a29f54d201e2ac5b85540">More...</a><br /></td></tr>
<tr class="separator:gac673ebb9ae2a29f54d201e2ac5b85540"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1c07851f6b665c3461a19e9e4eb33d26"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga1c07851f6b665c3461a19e9e4eb33d26">vrna_eval_consensus_structure_simple_verbose</a> (const char **alignment, const char *structure, FILE *file)</td></tr>
<tr class="memdesc:ga1c07851f6b665c3461a19e9e4eb33d26"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributions per loop.  <a href="group__eval.html#ga1c07851f6b665c3461a19e9e4eb33d26">More...</a><br /></td></tr>
<tr class="separator:ga1c07851f6b665c3461a19e9e4eb33d26"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad88927c62ab0a8b534e078e44be1b36e"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e">vrna_eval_consensus_structure_simple_v</a> (const char **alignment, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gad88927c62ab0a8b534e078e44be1b36e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributions per loop.  <a href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e">More...</a><br /></td></tr>
<tr class="separator:gad88927c62ab0a8b534e078e44be1b36e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gae89240c230e4740b22a703ee953396b9"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gae89240c230e4740b22a703ee953396b9">vrna_eval_circ_consensus_structure_v</a> (const char **alignment, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gae89240c230e4740b22a703ee953396b9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences and print contributions per loop.  <a href="group__eval.html#gae89240c230e4740b22a703ee953396b9">More...</a><br /></td></tr>
<tr class="separator:gae89240c230e4740b22a703ee953396b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga8abc794fc48d43268ced5e8cde017baa"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga8abc794fc48d43268ced5e8cde017baa">vrna_eval_gquad_consensus_structure_v</a> (const char **alignment, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:ga8abc794fc48d43268ced5e8cde017baa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a consensus structure for an RNA sequence alignment, allow for annotated G-Quadruplexes in the structure and print contributions per loop.  <a href="group__eval.html#ga8abc794fc48d43268ced5e8cde017baa">More...</a><br /></td></tr>
<tr class="separator:ga8abc794fc48d43268ced5e8cde017baa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaecd3e17292a0b3927277434019a5e187"><td class="memItemLeft" align="right" valign="top">float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gaecd3e17292a0b3927277434019a5e187">vrna_eval_circ_gquad_consensus_structure_v</a> (const char **alignment, const char *structure, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gaecd3e17292a0b3927277434019a5e187"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences, allow for annotated G-Quadruplexes in the structure and print contributions per loop.  <a href="group__eval.html#gaecd3e17292a0b3927277434019a5e187">More...</a><br /></td></tr>
<tr class="separator:gaecd3e17292a0b3927277434019a5e187"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Simplified Energy Evaluation with Sequence String and Structure Pair Table</h2></td></tr>
<tr class="memitem:ga0bba59b4d6e53461088666ff4aece7b0"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga0bba59b4d6e53461088666ff4aece7b0">vrna_eval_structure_pt_simple</a> (const char *string, const short *pt)</td></tr>
<tr class="memdesc:ga0bba59b4d6e53461088666ff4aece7b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga0bba59b4d6e53461088666ff4aece7b0">More...</a><br /></td></tr>
<tr class="separator:ga0bba59b4d6e53461088666ff4aece7b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga76e152ee9a02be23da14cdddf52b4e44"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#ga76e152ee9a02be23da14cdddf52b4e44">vrna_eval_structure_pt_simple_verbose</a> (const char *string, const short *pt, FILE *file)</td></tr>
<tr class="memdesc:ga76e152ee9a02be23da14cdddf52b4e44"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#ga76e152ee9a02be23da14cdddf52b4e44">More...</a><br /></td></tr>
<tr class="separator:ga76e152ee9a02be23da14cdddf52b4e44"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac40b813d35289da9816d0c1eec94faa5"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gac40b813d35289da9816d0c1eec94faa5">vrna_eval_structure_pt_simple_v</a> (const char *string, const short *pt, int verbosity_level, FILE *file)</td></tr>
<tr class="memdesc:gac40b813d35289da9816d0c1eec94faa5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Calculate the free energy of an already folded RNA.  <a href="group__eval.html#gac40b813d35289da9816d0c1eec94faa5">More...</a><br /></td></tr>
<tr class="separator:gac40b813d35289da9816d0c1eec94faa5"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Simplified Energy Evaluation with Sequence Alignment and Consensus Structure Pair Table</h2></td></tr>
<tr class="memitem:gabbb4d2a7aa324ec9cce8f47ce61ab8af"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__eval.html#gabbb4d2a7aa324ec9cce8f47ce61ab8af">vrna_eval_consensus_structure_pt_simple</a> (const char **alignment, const short *pt)</td></tr>
<tr class="memdesc:gabbb4d2a7aa324ec9cce8f47ce61ab8af"><td class="mdescLeft">&#160;</td><td class="mdescRight">Evaluate the Free Energy of a Consensus Secondary Structure given a Sequence Alignment.  <a href="group__eval.html#gabbb4d2a7aa324ec9cce8f47ce61ab8af">More...</a><br /></td></tr>
<tr class="separator:gabbb4d2a7aa324ec9cce8f47ce61ab8af"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2769e4369d023ad2d5c5f4d2ee825c23"><td class="memItemLeft" align="right" valign="top"><a id="ga2769e4369d023ad2d5c5f4d2ee825c23"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>vrna_eval_consensus_structure_pt_simple_verbose</b> (const char **alignment, const short *pt, FILE *file)</td></tr>
<tr class="separator:ga2769e4369d023ad2d5c5f4d2ee825c23"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf2d227b3d54bf9b693a3df52faf5e2e4"><td class="memItemLeft" align="right" valign="top"><a id="gaf2d227b3d54bf9b693a3df52faf5e2e4"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>vrna_eval_consensus_structure_pt_simple_v</b> (const char **alignment, const short *pt, int verbosity_level, FILE *file)</td></tr>
<tr class="separator:gaf2d227b3d54bf9b693a3df52faf5e2e4"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga58f199f1438d794a265f3b27fc8ea631"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga58f199f1438d794a265f3b27fc8ea631">&#9670;&nbsp;</a></span>vrna_eval_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float vrna_eval_structure </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function allows for energy evaluation of a given pair of structure and sequence (alignment). Model details, energy parameters, and possibly soft constraints are used as provided via the parameter 'vc'. The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> does not need to contain any DP matrices, but requires all most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a>(sequence, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
<div class="ttc" id="agroup__fold__compound_html_ga283cf113a37614a75c1388c4ec6d7b2f"><div class="ttname"><a href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a></div><div class="ttdeci">vrna_fold_compound_t * vrna_fold_compound(const char *sequence, const vrna_md_t *md_p, unsigned int options)</div><div class="ttdoc">Retrieve a vrna_fold_compound_t data structure for single sequences and hybridizing sequences.</div></div>
<div class="ttc" id="agroup__fold__compound_html_ga61469c423131552c8483229f8b6c7e0e"><div class="ttname"><a href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a></div><div class="ttdeci">#define VRNA_OPTION_EVAL_ONLY</div><div class="ttdoc">Option flag to specify that neither MFE, nor PF DP matrices are required.</div><div class="ttdef"><b>Definition:</b> fold_compound.h:387</div></div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>Accepts vrna_fold_compound_t of type <a class="el" href="group__fold__compound.html#gga01a4ff86fa71deaaa5d1abbd95a1447da7e264dd3cf2dc9b6448caabcb7763cd6">VRNA_FC_TYPE_SINGLE</a> and <a class="el" href="group__fold__compound.html#gga01a4ff86fa71deaaa5d1abbd95a1447dab821ce46ea3cf665be97df22a76f5023">VRNA_FC_TYPE_COMPARATIVE</a></dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>, <a class="el" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f" title="Retrieve a vrna_fold_compound_t data structure for single sequences and hybridizing sequences.">vrna_fold_compound()</a>, <a class="el" href="group__fold__compound.html#gad6bacc816af274922b13d947f708aa0c" title="Retrieve a vrna_fold_compound_t data structure for sequence alignments.">vrna_fold_compound_comparative()</a>, <a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000043">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>eval_structure()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga6cea75c0eb9857fb59172be54cab09e0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga6cea75c0eb9857fb59172be54cab09e0">&#9670;&nbsp;</a></span>vrna_eval_covar_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float vrna_eval_covar_structure </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences. </p>
<p>Consensus structure prediction is driven by covariance scores of base pairs in rows of the provided alignment. This function allows one to retrieve the total amount of this covariance pseudo energy scores. The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> does not need to contain any DP matrices, but requires all most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#gad6bacc816af274922b13d947f708aa0c">vrna_fold_compound_comparative</a>(alignment, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
<div class="ttc" id="agroup__fold__compound_html_gad6bacc816af274922b13d947f708aa0c"><div class="ttname"><a href="group__fold__compound.html#gad6bacc816af274922b13d947f708aa0c">vrna_fold_compound_comparative</a></div><div class="ttdeci">vrna_fold_compound_t * vrna_fold_compound_comparative(const char **sequences, vrna_md_t *md_p, unsigned int options)</div><div class="ttdoc">Retrieve a vrna_fold_compound_t data structure for sequence alignments.</div></div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>Accepts vrna_fold_compound_t of type <a class="el" href="group__fold__compound.html#gga01a4ff86fa71deaaa5d1abbd95a1447dab821ce46ea3cf665be97df22a76f5023">VRNA_FC_TYPE_COMPARATIVE</a> only!</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__fold__compound.html#gad6bacc816af274922b13d947f708aa0c" title="Retrieve a vrna_fold_compound_t data structure for sequence alignments.">vrna_fold_compound_comparative()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary (consensus) structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The covariance pseudo energy score of the input structure given the input sequence alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000047">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>eval_covar_structure()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga0928d699d310178f84ee2351034e5cb5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0928d699d310178f84ee2351034e5cb5">&#9670;&nbsp;</a></span>vrna_eval_structure_verbose()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float vrna_eval_structure_verbose </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA and print contributions on a per-loop base. </p>
<p>This function is a simplyfied version of <a class="el" href="group__eval.html#gab12e6b1226227670322150df018734f8" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_v()</a> that uses the <em>default</em> verbosity level.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>,</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000045">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>eval_structure_verbose()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="gab12e6b1226227670322150df018734f8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab12e6b1226227670322150df018734f8">&#9670;&nbsp;</a></span>vrna_eval_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">float vrna_eval_structure_v </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA and print contributions on a per-loop base. </p>
<p>This function allows for detailed energy evaluation of a given sequence/structure pair. In contrast to <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a> this function prints detailed energy contributions based on individual loops to a file handle. If NULL is passed as file handle, this function defaults to print to stdout. Any positive <code>verbosity_level</code> activates potential warning message of the energy evaluting functions, while values <img class="formulaInl" alt="$ \ge 1 $" src="form_73.png" width="21" height="12"/> allow for detailed control of what data is printed. A negative parameter <code>verbosity_level</code> turns off printing all together.</p>
<p>Model details, energy parameters, and possibly soft constraints are used as provided via the parameter 'vc'. The fold_compound does not need to contain any DP matrices, but all the most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a>(sequence, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
</div><!-- fragment --><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>,</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol </dd></dl>

</div>
</div>
<a id="gadbd09372ddfd7a450bbd590c96a6bfe4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gadbd09372ddfd7a450bbd590c96a6bfe4">&#9670;&nbsp;</a></span>vrna_eval_structure_pt()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function allows for energy evaluation of a given sequence/structure pair where the structure is provided in pair_table format as obtained from <a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>. Model details, energy parameters, and possibly soft constraints are used as provided via the parameter 'vc'. The fold_compound does not need to contain any DP matrices, but all the most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a>(sequence, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
</div><!-- fragment --><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000044">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>eval_structure_pt()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga8a517cfeeae8c376ae7b1e0c401d38b4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8a517cfeeae8c376ae7b1e0c401d38b4">&#9670;&nbsp;</a></span>vrna_eval_structure_pt_verbose()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt_verbose </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function is a simplyfied version of <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> that uses the <em>default</em> verbosity level.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga2c6533ba0afe4c88d335d8f1e0e2a48e" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_v()</a>, <a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000046">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>eval_structure_pt_verbose()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga2c6533ba0afe4c88d335d8f1e0e2a48e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2c6533ba0afe4c88d335d8f1e0e2a48e">&#9670;&nbsp;</a></span>vrna_eval_structure_pt_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt_v </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function allows for energy evaluation of a given sequence/structure pair where the structure is provided in pair_table format as obtained from <a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>. Model details, energy parameters, and possibly soft constraints are used as provided via the parameter 'vc'. The fold_compound does not need to contain any DP matrices, but all the most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a>(sequence, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
</div><!-- fragment --><p> In contrast to <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a> this function prints detailed energy contributions based on individual loops to a file handle. If NULL is passed as file handle, this function defaults to print to stdout. Any positive <code>verbosity_level</code> activates potential warning message of the energy evaluting functions, while values <img class="formulaInl" alt="$ \ge 1 $" src="form_73.png" width="21" height="12"/> allow for detailed control of what data is printed. A negative parameter <code>verbosity_level</code> turns off printing all together.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>A vrna_fold_compound_t containing the energy parameters and model details </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol </dd></dl>

</div>
</div>
<a id="ga7e5273464b775d4130245681312c1369"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7e5273464b775d4130245681312c1369">&#9670;&nbsp;</a></span>vrna_eval_structure_simple()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_simple </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function allows for energy evaluation of a given sequence/structure pair. In contrast to <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a> this function assumes default model details and default energy parameters in order to evaluate the free energy of the secondary structure. Therefore, it serves as a simple interface function for energy evaluation for situations where no changes on the energy model are required.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>,</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000053">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function serves as a wrapper for <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga3e05a23ddf9b083f4e69881e440d4866"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3e05a23ddf9b083f4e69881e440d4866">&#9670;&nbsp;</a></span>vrna_eval_circ_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_structure </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a sequence/structure pair where the sequence is circular. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>, <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>, <a class="el" href="group__eval.html#gac96577cf232c71160f762737a994b7c6" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_consensus_structure()</a>, <a class="el" href="group__eval.html#gac3fb44e0773a51be8efc5f4f595a94a7" title="Evaluate free energy of a sequence/structure pair, assume sequence to be circular and print contribut...">vrna_eval_circ_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the structure given the circular input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000054">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function serves as a wrapper for <a class="el" href="group__eval.html#gac3fb44e0773a51be8efc5f4f595a94a7" title="Evaluate free energy of a sequence/structure pair, assume sequence to be circular and print contribut...">vrna_eval_circ_structure_v()</a> and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga3263504825ef4b523eba797c99921df4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga3263504825ef4b523eba797c99921df4">&#9670;&nbsp;</a></span>vrna_eval_gquad_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_gquad_structure </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes. </p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>, <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>, <a class="el" href="group__eval.html#gaf09a326b3d57a4b30c27bd0e216198ac" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the structur...">vrna_eval_gquad_consensus_structure()</a>, <a class="el" href="group__eval.html#gaeaa2bdbc1b5d78c667e735fbdff87fff" title="Evaluate free energy of a sequence/structure pair, allow for G-Quadruplexes in the structure and prin...">vrna_eval_gquad_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the structure including contributions of G-quadruplexes in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000055">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function serves as a wrapper for <a class="el" href="group__eval.html#gaeaa2bdbc1b5d78c667e735fbdff87fff" title="Evaluate free energy of a sequence/structure pair, allow for G-Quadruplexes in the structure and prin...">vrna_eval_gquad_structure_v()</a> and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga9dba2fc5d7e6ad1359a7c2f350589c0e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga9dba2fc5d7e6ad1359a7c2f350589c0e">&#9670;&nbsp;</a></span>vrna_eval_circ_gquad_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_gquad_structure </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a sequence/structure pair where the sequence is circular and the structure may contain G-Quadruplexes. </p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>, <a class="el" href="group__eval.html#gac673ebb9ae2a29f54d201e2ac5b85540" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_gquad_consensus_structure()</a>, <a class="el" href="group__eval.html#gab96a6c59923ff06c35f8c2fd2c239727" title="Evaluate free energy of a sequence/structure pair, assume sequence to be circular,...">vrna_eval_circ_gquad_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the structure including contributions of G-quadruplexes in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000056">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function serves as a wrapper for <a class="el" href="group__eval.html#gab96a6c59923ff06c35f8c2fd2c239727" title="Evaluate free energy of a sequence/structure pair, assume sequence to be circular,...">vrna_eval_circ_gquad_structure_v()</a> and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gaf928bfd96767e1b8033a95a4cc432e39"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaf928bfd96767e1b8033a95a4cc432e39">&#9670;&nbsp;</a></span>vrna_eval_structure_simple_verbose()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_simple_verbose </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA and print contributions per loop. </p>
<p>This function is a simplyfied version of <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> that uses the <em>default</em> verbosity level.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000061">SWIG Wrapper Notes:</a></b></dt><dd>This function is not available. Use <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> instead! </dd></dl>

</div>
</div>
<a id="gacd6278343e77d13f1d53588e50d303bc"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gacd6278343e77d13f1d53588e50d303bc">&#9670;&nbsp;</a></span>vrna_eval_structure_simple_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_simple_v </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA and print contributions per loop. </p>
<p>This function allows for detailed energy evaluation of a given sequence/structure pair. In contrast to <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a> this function prints detailed energy contributions based on individual loops to a file handle. If NULL is passed as file handle, this function defaults to print to stdout. Any positive <code>verbosity_level</code> activates potential warning message of the energy evaluting functions, while values <img class="formulaInl" alt="$ \ge 1 $" src="form_73.png" width="21" height="12"/> allow for detailed control of what data is printed. A negative parameter <code>verbosity_level</code> turns off printing all together.</p>
<p>In contrast to <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a> this function assumes default model details and default energy parameters in order to evaluate the free energy of the secondary structure. Threefore, it serves as a simple interface function for energy evaluation for situations where no changes on the energy model are required.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>,</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000063">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>. The last two arguments for this function are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gac3fb44e0773a51be8efc5f4f595a94a7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac3fb44e0773a51be8efc5f4f595a94a7">&#9670;&nbsp;</a></span>vrna_eval_circ_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate free energy of a sequence/structure pair, assume sequence to be circular and print contributions per loop. </p>
<p>This function is the same as <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> but assumes the input sequence to be circularized.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a>, <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000064">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>. The last two arguments for this function are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gaeaa2bdbc1b5d78c667e735fbdff87fff"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaeaa2bdbc1b5d78c667e735fbdff87fff">&#9670;&nbsp;</a></span>vrna_eval_gquad_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_gquad_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate free energy of a sequence/structure pair, allow for G-Quadruplexes in the structure and print contributions per loop. </p>
<p>This function is the same as <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> but allows for annotated G-Quadruplexes in the dot-bracket structure input.</p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a>, <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000065">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>. The last two arguments for this function are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gab96a6c59923ff06c35f8c2fd2c239727"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab96a6c59923ff06c35f8c2fd2c239727">&#9670;&nbsp;</a></span>vrna_eval_circ_gquad_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_gquad_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate free energy of a sequence/structure pair, assume sequence to be circular, allow for G-Quadruplexes in the structure, and print contributions per loop. </p>
<p>This function is the same as <a class="el" href="group__eval.html#gacd6278343e77d13f1d53588e50d303bc" title="Calculate the free energy of an already folded RNA and print contributions per loop.">vrna_eval_structure_simple_v()</a> but assumes the input sequence to be circular and allows for annotated G-Quadruplexes in the dot-bracket structure input.</p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000066">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga9dba2fc5d7e6ad1359a7c2f350589c0e" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular and the structur...">vrna_eval_circ_gquad_structure()</a>. The last two arguments for this function are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga7762c3a7bdcbc3a14ef93259d322c7d6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga7762c3a7bdcbc3a14ef93259d322c7d6">&#9670;&nbsp;</a></span>vrna_eval_consensus_structure_simple()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_consensus_structure_simple </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA sequence alignment. </p>
<p>This function allows for energy evaluation for a given multiple sequence alignment and consensus structure pair. In contrast to <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a> this function assumes default model details and default energy parameters in order to evaluate the free energy of the secondary structure. Therefore, it serves as a simple interface function for energy evaluation for situations where no changes on the energy model are required.</p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters and hyphen ('-') to denote gaps </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus Secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the input alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000057">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloadeded version of <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument </dd></dl>

</div>
</div>
<a id="gac96577cf232c71160f762737a994b7c6"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac96577cf232c71160f762737a994b7c6">&#9670;&nbsp;</a></span>vrna_eval_circ_consensus_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_consensus_structure </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequences are circular. </p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a>, <a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6" title="Calculate the free energy of an already folded RNA sequence alignment.">vrna_eval_consensus_structure_simple()</a>, <a class="el" href="group__eval.html#gaf09a326b3d57a4b30c27bd0e216198ac" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the structur...">vrna_eval_gquad_consensus_structure()</a>, <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>, <a class="el" href="group__eval.html#gae89240c230e4740b22a703ee953396b9" title="Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences and prin...">vrna_eval_circ_consensus_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the circular input sequence in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000058">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloadeded version of <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument </dd></dl>

</div>
</div>
<a id="gaf09a326b3d57a4b30c27bd0e216198ac"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaf09a326b3d57a4b30c27bd0e216198ac">&#9670;&nbsp;</a></span>vrna_eval_gquad_consensus_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_gquad_consensus_structure </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the structure may contain G-Quadruplexes. </p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a>, <a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6" title="Calculate the free energy of an already folded RNA sequence alignment.">vrna_eval_consensus_structure_simple()</a>, <a class="el" href="group__eval.html#gac96577cf232c71160f762737a994b7c6" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_consensus_structure()</a>, <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>, <a class="el" href="group__eval.html#ga8abc794fc48d43268ced5e8cde017baa" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment, allow for annotated ...">vrna_eval_gquad_consensus_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure including contributions of G-quadruplexes in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000059">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloadeded version of <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument </dd></dl>

</div>
</div>
<a id="gac673ebb9ae2a29f54d201e2ac5b85540"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac673ebb9ae2a29f54d201e2ac5b85540">&#9670;&nbsp;</a></span>vrna_eval_circ_gquad_consensus_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_gquad_consensus_structure </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence is circular and the structure may contain G-Quadruplexes. </p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a>, <a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6" title="Calculate the free energy of an already folded RNA sequence alignment.">vrna_eval_consensus_structure_simple()</a>, <a class="el" href="group__eval.html#gac96577cf232c71160f762737a994b7c6" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_consensus_structure()</a>, <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>, <a class="el" href="group__eval.html#gaecd3e17292a0b3927277434019a5e187" title="Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences,...">vrna_eval_circ_gquad_consensus_structure_v()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure including contributions of G-quadruplexes in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000060">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloadeded version of <a class="el" href="group__eval.html#ga9dba2fc5d7e6ad1359a7c2f350589c0e" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular and the structur...">vrna_eval_circ_gquad_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument </dd></dl>

</div>
</div>
<a id="ga1c07851f6b665c3461a19e9e4eb33d26"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga1c07851f6b665c3461a19e9e4eb33d26">&#9670;&nbsp;</a></span>vrna_eval_consensus_structure_simple_verbose()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_consensus_structure_simple_verbose </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributions per loop. </p>
<p>This function is a simplyfied version of <a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a> that uses the <em>default</em> verbosity level.</p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a>, <a class="el" href="group__eval.html#ga0928d699d310178f84ee2351034e5cb5" title="Calculate the free energy of an already folded RNA and print contributions on a per-loop base.">vrna_eval_structure_verbose()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the conensus structure given the aligned input sequences in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000062">SWIG Wrapper Notes:</a></b></dt><dd>This function is not available. Use <a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a> instead! </dd></dl>

</div>
</div>
<a id="gad88927c62ab0a8b534e078e44be1b36e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad88927c62ab0a8b534e078e44be1b36e">&#9670;&nbsp;</a></span>vrna_eval_consensus_structure_simple_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_consensus_structure_simple_v </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributions per loop. </p>
<p>This function allows for detailed energy evaluation of a given sequence alignment/consensus structure pair. In contrast to <a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6" title="Calculate the free energy of an already folded RNA sequence alignment.">vrna_eval_consensus_structure_simple()</a> this function prints detailed energy contributions based on individual loops to a file handle. If NULL is passed as file handle, this function defaults to print to stdout. Any positive <code>verbosity_level</code> activates potential warning message of the energy evaluting functions, while values <img class="formulaInl" alt="$ \ge 1 $" src="form_73.png" width="21" height="12"/> allow for detailed control of what data is printed. A negative parameter <code>verbosity_level</code> turns off printing all together.</p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd>vrna_eval_consensus_structure(), <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the sequence alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000067">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gae89240c230e4740b22a703ee953396b9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gae89240c230e4740b22a703ee953396b9">&#9670;&nbsp;</a></span>vrna_eval_circ_consensus_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_consensus_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences and print contributions per loop. </p>
<p>This function is identical with <a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a> but assumed the aligned sequences to be circular.</p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a>, <a class="el" href="group__eval.html#gac96577cf232c71160f762737a994b7c6" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_consensus_structure()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the sequence alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000068">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga3e05a23ddf9b083f4e69881e440d4866" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular.">vrna_eval_circ_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga8abc794fc48d43268ced5e8cde017baa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga8abc794fc48d43268ced5e8cde017baa">&#9670;&nbsp;</a></span>vrna_eval_gquad_consensus_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_gquad_consensus_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a consensus structure for an RNA sequence alignment, allow for annotated G-Quadruplexes in the structure and print contributions per loop. </p>
<p>This function is identical with <a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a> but allows for annotated G-Quadruplexes in the consensus structure.</p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a>, <a class="el" href="group__eval.html#gaf09a326b3d57a4b30c27bd0e216198ac" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the structur...">vrna_eval_gquad_consensus_structure()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the sequence alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000069">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga3263504825ef4b523eba797c99921df4" title="Evaluate the free energy of a sequence/structure pair where the structure may contain G-Quadruplexes.">vrna_eval_gquad_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="gaecd3e17292a0b3927277434019a5e187"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaecd3e17292a0b3927277434019a5e187">&#9670;&nbsp;</a></span>vrna_eval_circ_gquad_consensus_structure_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_circ_gquad_consensus_structure_v </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the free energy of a consensus structure for an alignment of circular RNA sequences, allow for annotated G-Quadruplexes in the structure and print contributions per loop. </p>
<p>This function is identical with <a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a> but assumes the sequences in the alignment to be circular and allows for annotated G-Quadruplexes in the consensus structure.</p>
<p>G-Quadruplexes are annotated as plus signs ('+') for each G involved in the motif. Linker sequences must be denoted by dots ('.') as they are considered unpaired. Below is an example of a 2-layer G-quadruplex: </p><div class="fragment"><div class="line">GGAAGGAAAGGAGG</div>
<div class="line">++..++...++.++</div>
</div><!-- fragment --><dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gad88927c62ab0a8b534e078e44be1b36e" title="Evaluate the free energy of a consensus structure for an RNA sequence alignment and print contributio...">vrna_eval_consensus_structure_simple_v()</a>, <a class="el" href="group__eval.html#gac673ebb9ae2a29f54d201e2ac5b85540" title="Evaluate the free energy of a multiple sequence alignment/consensus structure pair where the sequence...">vrna_eval_circ_gquad_consensus_structure()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">structure</td><td>Consensus secondary structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the consensus structure given the sequence alignment in kcal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000070">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloaded version of <a class="el" href="group__eval.html#ga9dba2fc5d7e6ad1359a7c2f350589c0e" title="Evaluate the free energy of a sequence/structure pair where the sequence is circular and the structur...">vrna_eval_circ_gquad_structure()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga0bba59b4d6e53461088666ff4aece7b0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0bba59b4d6e53461088666ff4aece7b0">&#9670;&nbsp;</a></span>vrna_eval_structure_pt_simple()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt_simple </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>In contrast to <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a> this function assumes default model details and default energy parameters in order to evaluate the free energy of the secondary structure. Threefore, it serves as a simple interface function for energy evaluation for situations where no changes on the energy model are required.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000071">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function serves as a wrapper for <a class="el" href="group__eval.html#ga2c6533ba0afe4c88d335d8f1e0e2a48e" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_v()</a> and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to <a class="el" href="group__eval.html#gaf4afe19780b61b4962c613bde324128b" title="Quiet level verbosity setting.">VRNA_VERBOSITY_QUIET</a> and NULL, respectively. </dd></dl>

</div>
</div>
<a id="ga76e152ee9a02be23da14cdddf52b4e44"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga76e152ee9a02be23da14cdddf52b4e44">&#9670;&nbsp;</a></span>vrna_eval_structure_pt_simple_verbose()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt_simple_verbose </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function is a simplyfied version of <a class="el" href="group__eval.html#gac40b813d35289da9816d0c1eec94faa5" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_simple_v()</a> that uses the <em>default</em> verbosity level.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#gac40b813d35289da9816d0c1eec94faa5" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_simple_v()</a>, <a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a>, <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol </dd></dl>

</div>
</div>
<a id="gac40b813d35289da9816d0c1eec94faa5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac40b813d35289da9816d0c1eec94faa5">&#9670;&nbsp;</a></span>vrna_eval_structure_pt_simple_v()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_structure_pt_simple_v </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>string</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>verbosity_level</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>file</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Calculate the free energy of an already folded RNA. </p>
<p>This function allows for energy evaluation of a given sequence/structure pair where the structure is provided in pair_table format as obtained from <a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>. Model details, energy parameters, and possibly soft constraints are used as provided via the parameter 'vc'. The fold_compound does not need to contain any DP matrices, but all the most basic init values as one would get from a call like this: </p><div class="fragment"><div class="line">vc = <a class="code" href="group__fold__compound.html#ga283cf113a37614a75c1388c4ec6d7b2f">vrna_fold_compound</a>(sequence, NULL, <a class="code" href="group__fold__compound.html#ga61469c423131552c8483229f8b6c7e0e">VRNA_OPTION_EVAL_ONLY</a>);</div>
</div><!-- fragment --><p> In contrast to <a class="el" href="group__eval.html#ga8a517cfeeae8c376ae7b1e0c401d38b4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_verbose()</a> this function assumes default model details and default energy parameters in order to evaluate the free energy of the secondary structure. Threefore, it serves as a simple interface function for energy evaluation for situations where no changes on the energy model are required.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__struct__utils__pair__table.html#gab124ba58014a97d2fb8c21831e19f107" title="Create a pair table from a dot-bracket notation of a secondary structure.">vrna_ptable()</a>, <a class="el" href="group__eval.html#ga2c6533ba0afe4c88d335d8f1e0e2a48e" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_v()</a>, <a class="el" href="group__eval.html#ga7e5273464b775d4130245681312c1369" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_simple()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">string</td><td>RNA sequence in uppercase letters </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure as pair_table </td></tr>
    <tr><td class="paramname">verbosity_level</td><td>The level of verbosity of this function </td></tr>
    <tr><td class="paramname">file</td><td>A file handle where this function should print to (may be NULL). </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The free energy of the input structure given the input sequence in 10cal/mol </dd></dl>

</div>
</div>
<a id="gabbb4d2a7aa324ec9cce8f47ce61ab8af"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabbb4d2a7aa324ec9cce8f47ce61ab8af">&#9670;&nbsp;</a></span>vrna_eval_consensus_structure_pt_simple()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_eval_consensus_structure_pt_simple </td>
          <td>(</td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>alignment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="eval_8h.html">ViennaRNA/eval.h</a>&gt;</code></p>

<p>Evaluate the Free Energy of a Consensus Secondary Structure given a Sequence Alignment. </p>
<dl class="section note"><dt>Note</dt><dd>The free energy returned from this function already includes the covariation pseudo energies that is used fir comparative structure prediction within this library.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__eval.html#ga7762c3a7bdcbc3a14ef93259d322c7d6" title="Calculate the free energy of an already folded RNA sequence alignment.">vrna_eval_consensus_structure_simple()</a>, <a class="el" href="group__eval.html#gadbd09372ddfd7a450bbd590c96a6bfe4" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt()</a>, <a class="el" href="group__eval.html#ga58f199f1438d794a265f3b27fc8ea631" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure()</a>, <a class="el" href="group__eval.html#ga6cea75c0eb9857fb59172be54cab09e0" title="Calculate the pseudo energy derived by the covariance scores of a set of aligned sequences.">vrna_eval_covar_structure()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">alignment</td><td>RNA sequence alignment in uppercase letters. Gaps are denoted by hyphens ('-') </td></tr>
    <tr><td class="paramname">pt</td><td>Secondary structure in pair table format </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Free energy of the consensus structure in 10cal/mol</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000072">SWIG Wrapper Notes:</a></b></dt><dd>This function is available through an overloadeded version of <a class="el" href="group__eval.html#ga0bba59b4d6e53461088666ff4aece7b0" title="Calculate the free energy of an already folded RNA.">vrna_eval_structure_pt_simple()</a>. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument </dd></dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:30 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
