<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: SHAPE Reactivity Data</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__SHAPE__reactivities.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#files">Files</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">SHAPE Reactivity Data<div class="ingroups"><a class="el" href="group__probing__data.html">Experimental Structure Probing Data</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Incorporate SHAPE reactivity structure probing data into the folding recursions by means of soft constraints.  
<a href="#details">More...</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Incorporate SHAPE reactivity structure probing data into the folding recursions by means of soft constraints. </p>
<p>Details for our implementation to incorporate SHAPE reactivity data to guide secondary structure prediction can be found in <a class="el" href="citelist.html#CITEREF_lorenz:2016a">[17]</a> </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for SHAPE Reactivity Data:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__SHAPE__reactivities.svg" width="376" height="51"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:SHAPE_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="SHAPE_8h.html">SHAPE.h</a></td></tr>
<tr class="memdesc:SHAPE_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">This module provides function to incorporate SHAPE reactivity data into the folding recursions by means of soft constraints. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga57d612b58e1c61dd6cfcb5a843f8f1b3"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__SHAPE__reactivities.html#ga57d612b58e1c61dd6cfcb5a843f8f1b3">vrna_sc_add_SHAPE_deigan</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const double *reactivities, double m, double b, unsigned int options)</td></tr>
<tr class="memdesc:ga57d612b58e1c61dd6cfcb5a843f8f1b3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add SHAPE reactivity data as soft constraints (Deigan et al. method)  <a href="group__SHAPE__reactivities.html#ga57d612b58e1c61dd6cfcb5a843f8f1b3">More...</a><br /></td></tr>
<tr class="separator:ga57d612b58e1c61dd6cfcb5a843f8f1b3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga04ba85da63d8c793bb8001d1e6f800ba"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__SHAPE__reactivities.html#ga04ba85da63d8c793bb8001d1e6f800ba">vrna_sc_add_SHAPE_deigan_ali</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char **shape_files, const int *shape_file_association, double m, double b, unsigned int options)</td></tr>
<tr class="memdesc:ga04ba85da63d8c793bb8001d1e6f800ba"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add SHAPE reactivity data from files as soft constraints for consensus structure prediction (Deigan et al. method)  <a href="group__SHAPE__reactivities.html#ga04ba85da63d8c793bb8001d1e6f800ba">More...</a><br /></td></tr>
<tr class="separator:ga04ba85da63d8c793bb8001d1e6f800ba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaf3c65a045060aef5c4e41693d30af58c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__SHAPE__reactivities.html#gaf3c65a045060aef5c4e41693d30af58c">vrna_sc_add_SHAPE_zarringhalam</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const double *reactivities, double b, double default_value, const char *shape_conversion, unsigned int options)</td></tr>
<tr class="memdesc:gaf3c65a045060aef5c4e41693d30af58c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add SHAPE reactivity data as soft constraints (Zarringhalam et al. method)  <a href="group__SHAPE__reactivities.html#gaf3c65a045060aef5c4e41693d30af58c">More...</a><br /></td></tr>
<tr class="separator:gaf3c65a045060aef5c4e41693d30af58c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga67675b3ed48744489a3bcfa4174197cb"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__SHAPE__reactivities.html#ga67675b3ed48744489a3bcfa4174197cb">vrna_sc_SHAPE_to_pr</a> (const char *shape_conversion, double *values, int length, double default_value)</td></tr>
<tr class="memdesc:ga67675b3ed48744489a3bcfa4174197cb"><td class="mdescLeft">&#160;</td><td class="mdescRight">Convert SHAPE reactivity values to probabilities for being unpaired.  <a href="group__SHAPE__reactivities.html#ga67675b3ed48744489a3bcfa4174197cb">More...</a><br /></td></tr>
<tr class="separator:ga67675b3ed48744489a3bcfa4174197cb"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga57d612b58e1c61dd6cfcb5a843f8f1b3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga57d612b58e1c61dd6cfcb5a843f8f1b3">&#9670;&nbsp;</a></span>vrna_sc_add_SHAPE_deigan()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_sc_add_SHAPE_deigan </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double *&#160;</td>
          <td class="paramname"><em>reactivities</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="SHAPE_8h.html">ViennaRNA/constraints/SHAPE.h</a>&gt;</code></p>

<p>Add SHAPE reactivity data as soft constraints (Deigan et al. method) </p>
<p>This approach of SHAPE directed RNA folding uses the simple linear ansatz </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ \Delta G_{\text{SHAPE}}(i) = m \ln(\text{SHAPE reactivity}(i)+1)+ b \]" src="form_131.png" width="300" height="15"/>
</p>
<p> to convert SHAPE reactivity values to pseudo energies whenever a nucleotide <img class="formulaInl" alt="$ i $" src="form_3.png" width="6" height="10"/> contributes to a stacked pair. A positive slope <img class="formulaInl" alt="$ m $" src="form_21.png" width="14" height="7"/> penalizes high reactivities in paired regions, while a negative intercept <img class="formulaInl" alt="$ b $" src="form_132.png" width="6" height="11"/> results in a confirmatory `&lsquo;bonus&rsquo;' free energy for correctly predicted base pairs. Since the energy evaluation of a base pair stack involves two pairs, the pseudo energies are added for all four contributing nucleotides. Consequently, the energy term is applied twice for pairs inside a helix and only once for pairs adjacent to other structures. For all other loop types the energy model remains unchanged even when the experimental data highly disagrees with a certain motif.</p>
<dl class="section see"><dt>See also</dt><dd>For further details, we refer to <a class="el" href="citelist.html#CITEREF_deigan:2009">[6]</a>. </dd>
<dd>
<a class="el" href="group__soft__constraints.html#ga73cdc07b9a199c614367bebef0f2c41a" title="Remove soft constraints from vrna_fold_compound_t.">vrna_sc_remove()</a>, <a class="el" href="group__SHAPE__reactivities.html#gaf3c65a045060aef5c4e41693d30af58c" title="Add SHAPE reactivity data as soft constraints (Zarringhalam et al. method)">vrna_sc_add_SHAPE_zarringhalam()</a>, <a class="el" href="group__perturbation.html#gaa124bdc20d88001c38ade590c4bcc3c4" title="Find a vector of perturbation energies that minimizes the discripancies between predicted and observe...">vrna_sc_minimize_pertubation()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> the soft constraints are associated with </td></tr>
    <tr><td class="paramname">reactivities</td><td>A vector of normalized SHAPE reactivities </td></tr>
    <tr><td class="paramname">m</td><td>The slope of the conversion function </td></tr>
    <tr><td class="paramname">b</td><td>The intercept of the conversion function </td></tr>
    <tr><td class="paramname">options</td><td>The options flag indicating how/where to store the soft constraints </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>1 on successful extraction of the method, 0 on errors</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000030">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>sc_add_SHAPE_deigan()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga04ba85da63d8c793bb8001d1e6f800ba"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga04ba85da63d8c793bb8001d1e6f800ba">&#9670;&nbsp;</a></span>vrna_sc_add_SHAPE_deigan_ali()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_sc_add_SHAPE_deigan_ali </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>shape_files</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int *&#160;</td>
          <td class="paramname"><em>shape_file_association</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>m</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="SHAPE_8h.html">ViennaRNA/constraints/SHAPE.h</a>&gt;</code></p>

<p>Add SHAPE reactivity data from files as soft constraints for consensus structure prediction (Deigan et al. method) </p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> the soft constraints are associated with </td></tr>
    <tr><td class="paramname">shape_files</td><td>A set of filenames that contain normalized SHAPE reactivity data </td></tr>
    <tr><td class="paramname">shape_file_association</td><td>An array of integers that associate the files with sequences in the alignment </td></tr>
    <tr><td class="paramname">m</td><td>The slope of the conversion function </td></tr>
    <tr><td class="paramname">b</td><td>The intercept of the conversion function </td></tr>
    <tr><td class="paramname">options</td><td>The options flag indicating how/where to store the soft constraints </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>1 on successful extraction of the method, 0 on errors</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000031">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>sc_add_SHAPE_deigan_ali()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="gaf3c65a045060aef5c4e41693d30af58c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaf3c65a045060aef5c4e41693d30af58c">&#9670;&nbsp;</a></span>vrna_sc_add_SHAPE_zarringhalam()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_sc_add_SHAPE_zarringhalam </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const double *&#160;</td>
          <td class="paramname"><em>reactivities</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>b</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>default_value</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>shape_conversion</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="SHAPE_8h.html">ViennaRNA/constraints/SHAPE.h</a>&gt;</code></p>

<p>Add SHAPE reactivity data as soft constraints (Zarringhalam et al. method) </p>
<p>This method first converts the observed SHAPE reactivity of nucleotide <img class="formulaInl" alt="$ i $" src="form_3.png" width="6" height="10"/> into a probability <img class="formulaInl" alt="$ q_i $" src="form_133.png" width="10" height="10"/> that position <img class="formulaInl" alt="$ i $" src="form_3.png" width="6" height="10"/> is unpaired by means of a non-linear map. Then pseudo-energies of the form </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ \Delta G_{\text{SHAPE}}(x,i) = \beta\ |x_i - q_i| \]" src="form_134.png" width="168" height="15"/>
</p>
<p> are computed, where <img class="formulaInl" alt="$ x_i=0 $" src="form_135.png" width="39" height="12"/> if position <img class="formulaInl" alt="$ i $" src="form_3.png" width="6" height="10"/> is unpaired and <img class="formulaInl" alt="$ x_i=1 $" src="form_136.png" width="39" height="12"/> if <img class="formulaInl" alt="$ i $" src="form_3.png" width="6" height="10"/> is paired in a given secondary structure. The parameter <img class="formulaInl" alt="$ \beta $" src="form_137.png" width="10" height="14"/> serves as scaling factor. The magnitude of discrepancy between prediction and experimental observation is represented by <img class="formulaInl" alt="$ |x_i - q_i| $" src="form_138.png" width="47" height="15"/>.</p>
<dl class="section see"><dt>See also</dt><dd>For further details, we refer to <a class="el" href="citelist.html#CITEREF_zarringhalam:2012">[26]</a> </dd>
<dd>
<a class="el" href="group__soft__constraints.html#ga73cdc07b9a199c614367bebef0f2c41a" title="Remove soft constraints from vrna_fold_compound_t.">vrna_sc_remove()</a>, <a class="el" href="group__SHAPE__reactivities.html#ga57d612b58e1c61dd6cfcb5a843f8f1b3" title="Add SHAPE reactivity data as soft constraints (Deigan et al. method)">vrna_sc_add_SHAPE_deigan()</a>, <a class="el" href="group__perturbation.html#gaa124bdc20d88001c38ade590c4bcc3c4" title="Find a vector of perturbation energies that minimizes the discripancies between predicted and observe...">vrna_sc_minimize_pertubation()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> the soft constraints are associated with </td></tr>
    <tr><td class="paramname">reactivities</td><td>A vector of normalized SHAPE reactivities </td></tr>
    <tr><td class="paramname">b</td><td>The scaling factor <img class="formulaInl" alt="$ \beta $" src="form_137.png" width="10" height="14"/> of the conversion function </td></tr>
    <tr><td class="paramname">default_value</td><td>The default value for a nucleotide where reactivity data is missing for </td></tr>
    <tr><td class="paramname">shape_conversion</td><td>A flag that specifies how to convert reactivities to probabilities </td></tr>
    <tr><td class="paramname">options</td><td>The options flag indicating how/where to store the soft constraints </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>1 on successful extraction of the method, 0 on errors</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000032">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>sc_add_SHAPE_zarringhalam()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga67675b3ed48744489a3bcfa4174197cb"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga67675b3ed48744489a3bcfa4174197cb">&#9670;&nbsp;</a></span>vrna_sc_SHAPE_to_pr()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_sc_SHAPE_to_pr </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>shape_conversion</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double *&#160;</td>
          <td class="paramname"><em>values</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>default_value</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="SHAPE_8h.html">ViennaRNA/constraints/SHAPE.h</a>&gt;</code></p>

<p>Convert SHAPE reactivity values to probabilities for being unpaired. </p>
<p>This function parses the informations from a given file and stores the result in the preallocated string sequence and the <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126" title="Typename for floating point number in partition function computations.">FLT_OR_DBL</a> array values.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats.html#ga646ebf45450a69a7f2533f9ecd283a32" title="Read data from a given SHAPE reactivity input file.">vrna_file_SHAPE_read()</a> </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">shape_conversion</td><td>String definining the method used for the conversion process </td></tr>
    <tr><td class="paramname">values</td><td>Pointer to an array of SHAPE reactivities </td></tr>
    <tr><td class="paramname">length</td><td>Length of the array of SHAPE reactivities </td></tr>
    <tr><td class="paramname">default_value</td><td>Result used for position with invalid/missing reactivity values </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:34 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
