<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: Multiple Sequence Alignments</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__file__formats__msa.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#files">Files</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Multiple Sequence Alignments<div class="ingroups"><a class="el" href="group__utils.html">Utilities</a> &raquo; <a class="el" href="group__file__utils.html">Files and I/O</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Functions to read/write multiple sequence alignments (MSA) in various file formats.  
<a href="#details">More...</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Functions to read/write multiple sequence alignments (MSA) in various file formats. </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Multiple Sequence Alignments:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><div class="zoom"><iframe scrolling="no" frameborder="0" src="group__file__formats__msa.svg" width="100%" height="300"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe></div>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:io_2file__formats__msa_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="io_2file__formats__msa_8h.html">file_formats_msa.h</a></td></tr>
<tr class="memdesc:io_2file__formats__msa_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions dealing with file formats for Multiple Sequence Alignments (MSA) <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:ga79a23de2c7249f2cccd762e475c81859"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga79a23de2c7249f2cccd762e475c81859">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>&#160;&#160;&#160;1U</td></tr>
<tr class="memdesc:ga79a23de2c7249f2cccd762e475c81859"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating ClustalW formatted files.  <a href="group__file__formats__msa.html#ga79a23de2c7249f2cccd762e475c81859">More...</a><br /></td></tr>
<tr class="separator:ga79a23de2c7249f2cccd762e475c81859"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga62be992445cd8ab2ad7a8fded944338b"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>&#160;&#160;&#160;2U</td></tr>
<tr class="memdesc:ga62be992445cd8ab2ad7a8fded944338b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating Stockholm 1.0 formatted files.  <a href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b">More...</a><br /></td></tr>
<tr class="separator:ga62be992445cd8ab2ad7a8fded944338b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gacf6274a2c825f34e4131404665b00604"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a>&#160;&#160;&#160;4U</td></tr>
<tr class="memdesc:gacf6274a2c825f34e4131404665b00604"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating FASTA (Pearson) formatted files.  <a href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604">More...</a><br /></td></tr>
<tr class="separator:gacf6274a2c825f34e4131404665b00604"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab3186d774c01570d2a47d0308eec4927"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a>&#160;&#160;&#160;8U</td></tr>
<tr class="memdesc:gab3186d774c01570d2a47d0308eec4927"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating MAF formatted files.  <a href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927">More...</a><br /></td></tr>
<tr class="separator:gab3186d774c01570d2a47d0308eec4927"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga494488a771aa0c602fb4cf445be34d47"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga494488a771aa0c602fb4cf445be34d47">VRNA_FILE_FORMAT_MSA_MIS</a>&#160;&#160;&#160;16U</td></tr>
<tr class="memdesc:ga494488a771aa0c602fb4cf445be34d47"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating most informative sequence (MIS) output.  <a href="group__file__formats__msa.html#ga494488a771aa0c602fb4cf445be34d47">More...</a><br /></td></tr>
<tr class="separator:ga494488a771aa0c602fb4cf445be34d47"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4acc255373831856a8417b68de8a94c0"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0">VRNA_FILE_FORMAT_MSA_DEFAULT</a></td></tr>
<tr class="memdesc:ga4acc255373831856a8417b68de8a94c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating the set of default file formats.  <a href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0">More...</a><br /></td></tr>
<tr class="separator:ga4acc255373831856a8417b68de8a94c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga229fb3778ecabea4782902b69fa48fd1"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1">VRNA_FILE_FORMAT_MSA_NOCHECK</a>&#160;&#160;&#160;4096U</td></tr>
<tr class="memdesc:ga229fb3778ecabea4782902b69fa48fd1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag to disable validation of the alignment.  <a href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1">More...</a><br /></td></tr>
<tr class="separator:ga229fb3778ecabea4782902b69fa48fd1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gabdc948f547e550125de3e7c65878400c"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gabdc948f547e550125de3e7c65878400c">VRNA_FILE_FORMAT_MSA_UNKNOWN</a>&#160;&#160;&#160;8192U</td></tr>
<tr class="memdesc:gabdc948f547e550125de3e7c65878400c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Return flag of <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> to indicate unknown or malformatted alignment.  <a href="group__file__formats__msa.html#gabdc948f547e550125de3e7c65878400c">More...</a><br /></td></tr>
<tr class="separator:gabdc948f547e550125de3e7c65878400c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga1577ea0f497d9c501549c863a4f2c089"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga1577ea0f497d9c501549c863a4f2c089">VRNA_FILE_FORMAT_MSA_APPEND</a>&#160;&#160;&#160;16384U</td></tr>
<tr class="memdesc:ga1577ea0f497d9c501549c863a4f2c089"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag indicating to append data to a multiple sequence alignment file rather than overwriting it.  <a href="group__file__formats__msa.html#ga1577ea0f497d9c501549c863a4f2c089">More...</a><br /></td></tr>
<tr class="separator:ga1577ea0f497d9c501549c863a4f2c089"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga73aa3701a22c5acfcb2e8927287a0554"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#ga73aa3701a22c5acfcb2e8927287a0554">VRNA_FILE_FORMAT_MSA_QUIET</a>&#160;&#160;&#160;32768U</td></tr>
<tr class="memdesc:ga73aa3701a22c5acfcb2e8927287a0554"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag to suppress unnecessary spam messages on <code>stderr</code>  <a href="group__file__formats__msa.html#ga73aa3701a22c5acfcb2e8927287a0554">More...</a><br /></td></tr>
<tr class="separator:ga73aa3701a22c5acfcb2e8927287a0554"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad98ca08968d19f221f29d2a1faf26887"><td class="memItemLeft" align="right" valign="top">#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gad98ca08968d19f221f29d2a1faf26887">VRNA_FILE_FORMAT_MSA_SILENT</a>&#160;&#160;&#160;65536U</td></tr>
<tr class="memdesc:gad98ca08968d19f221f29d2a1faf26887"><td class="mdescLeft">&#160;</td><td class="mdescRight">Option flag to completely silence any warnings on <code>stderr</code>  <a href="group__file__formats__msa.html#gad98ca08968d19f221f29d2a1faf26887">More...</a><br /></td></tr>
<tr class="separator:gad98ca08968d19f221f29d2a1faf26887"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:gad02d5d12bda54611c915a1019323b7be"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be">vrna_file_msa_read</a> (const char *filename, char ***names, char ***aln, char **id, char **structure, unsigned int options)</td></tr>
<tr class="memdesc:gad02d5d12bda54611c915a1019323b7be"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a multiple sequence alignment from file.  <a href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be">More...</a><br /></td></tr>
<tr class="separator:gad02d5d12bda54611c915a1019323b7be"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad4203a438622b2df7bc2f16578d54799"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799">vrna_file_msa_read_record</a> (FILE *fp, char ***names, char ***aln, char **id, char **structure, unsigned int options)</td></tr>
<tr class="memdesc:gad4203a438622b2df7bc2f16578d54799"><td class="mdescLeft">&#160;</td><td class="mdescRight">Read a multiple sequence alignment from file handle.  <a href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799">More...</a><br /></td></tr>
<tr class="separator:gad4203a438622b2df7bc2f16578d54799"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="memItemLeft" align="right" valign="top">unsigned int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3">vrna_file_msa_detect_format</a> (const char *filename, unsigned int options)</td></tr>
<tr class="memdesc:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Detect the format of a multiple sequence alignment file.  <a href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3">More...</a><br /></td></tr>
<tr class="separator:gade4fa8136ebb2d0f7eb3f8b59a8658e3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaccca55b71d34def88e18ed9a51ccb2bf"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__file__formats__msa.html#gaccca55b71d34def88e18ed9a51ccb2bf">vrna_file_msa_write</a> (const char *filename, const char **names, const char **aln, const char *id, const char *structure, const char *source, unsigned int options)</td></tr>
<tr class="memdesc:gaccca55b71d34def88e18ed9a51ccb2bf"><td class="mdescLeft">&#160;</td><td class="mdescRight">Write multiple sequence alignment file.  <a href="group__file__formats__msa.html#gaccca55b71d34def88e18ed9a51ccb2bf">More...</a><br /></td></tr>
<tr class="separator:gaccca55b71d34def88e18ed9a51ccb2bf"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Macro Definition Documentation</h2>
<a id="ga79a23de2c7249f2cccd762e475c81859"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga79a23de2c7249f2cccd762e475c81859">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_CLUSTAL</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_CLUSTAL&#160;&#160;&#160;1U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating ClustalW formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="ga62be992445cd8ab2ad7a8fded944338b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga62be992445cd8ab2ad7a8fded944338b">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_STOCKHOLM</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_STOCKHOLM&#160;&#160;&#160;2U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating Stockholm 1.0 formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="gacf6274a2c825f34e4131404665b00604"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gacf6274a2c825f34e4131404665b00604">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_FASTA</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_FASTA&#160;&#160;&#160;4U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating FASTA (Pearson) formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="gab3186d774c01570d2a47d0308eec4927"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab3186d774c01570d2a47d0308eec4927">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_MAF</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_MAF&#160;&#160;&#160;8U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating MAF formatted files. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="ga494488a771aa0c602fb4cf445be34d47"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga494488a771aa0c602fb4cf445be34d47">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_MIS</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_MIS&#160;&#160;&#160;16U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating most informative sequence (MIS) output. </p>
<p>The default reference sequence output for an alignment is simply a consensus sequence. This flag allows to write the most informative equence (MIS) instead.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gaccca55b71d34def88e18ed9a51ccb2bf" title="Write multiple sequence alignment file.">vrna_file_msa_write()</a> </dd></dl>

</div>
</div>
<a id="ga4acc255373831856a8417b68de8a94c0"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga4acc255373831856a8417b68de8a94c0">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_DEFAULT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_DEFAULT</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>
<b>Value:</b><div class="fragment"><div class="line">    ( \</div>
<div class="line">    VRNA_FILE_FORMAT_MSA_CLUSTAL \</div>
<div class="line">    | <a class="code" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a> \</div>
<div class="line">    | <a class="code" href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a> \</div>
<div class="line">    | <a class="code" href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a> \</div>
<div class="line">    )</div>
<div class="ttc" id="agroup__file__formats__msa_html_ga62be992445cd8ab2ad7a8fded944338b"><div class="ttname"><a href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_STOCKHOLM</div><div class="ttdoc">Option flag indicating Stockholm 1.0 formatted files.</div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:28</div></div>
<div class="ttc" id="agroup__file__formats__msa_html_gab3186d774c01570d2a47d0308eec4927"><div class="ttname"><a href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927">VRNA_FILE_FORMAT_MSA_MAF</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_MAF</div><div class="ttdoc">Option flag indicating MAF formatted files.</div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:40</div></div>
<div class="ttc" id="agroup__file__formats__msa_html_gacf6274a2c825f34e4131404665b00604"><div class="ttname"><a href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604">VRNA_FILE_FORMAT_MSA_FASTA</a></div><div class="ttdeci">#define VRNA_FILE_FORMAT_MSA_FASTA</div><div class="ttdoc">Option flag indicating FASTA (Pearson) formatted files.</div><div class="ttdef"><b>Definition:</b> file_formats_msa.h:34</div></div>
</div><!-- fragment -->
<p>Option flag indicating the set of default file formats. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="ga229fb3778ecabea4782902b69fa48fd1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga229fb3778ecabea4782902b69fa48fd1">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_NOCHECK</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_NOCHECK&#160;&#160;&#160;4096U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag to disable validation of the alignment. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a> </dd></dl>

</div>
</div>
<a id="gabdc948f547e550125de3e7c65878400c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gabdc948f547e550125de3e7c65878400c">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_UNKNOWN</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_UNKNOWN&#160;&#160;&#160;8192U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Return flag of <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> to indicate unknown or malformatted alignment. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> </dd></dl>

</div>
</div>
<a id="ga1577ea0f497d9c501549c863a4f2c089"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga1577ea0f497d9c501549c863a4f2c089">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_APPEND</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_APPEND&#160;&#160;&#160;16384U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag indicating to append data to a multiple sequence alignment file rather than overwriting it. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gaccca55b71d34def88e18ed9a51ccb2bf" title="Write multiple sequence alignment file.">vrna_file_msa_write()</a> </dd></dl>

</div>
</div>
<a id="ga73aa3701a22c5acfcb2e8927287a0554"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga73aa3701a22c5acfcb2e8927287a0554">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_QUIET</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_QUIET&#160;&#160;&#160;32768U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag to suppress unnecessary spam messages on <code>stderr</code> </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a> </dd></dl>

</div>
</div>
<a id="gad98ca08968d19f221f29d2a1faf26887"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad98ca08968d19f221f29d2a1faf26887">&#9670;&nbsp;</a></span>VRNA_FILE_FORMAT_MSA_SILENT</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">#define VRNA_FILE_FORMAT_MSA_SILENT&#160;&#160;&#160;65536U</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Option flag to completely silence any warnings on <code>stderr</code> </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a> </dd></dl>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="gad02d5d12bda54611c915a1019323b7be"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad02d5d12bda54611c915a1019323b7be">&#9670;&nbsp;</a></span>vrna_file_msa_read()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_msa_read </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Read a multiple sequence alignment from file. </p>
<p>This function reads the (first) multiple sequence alignment from an input file. The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The <code>options</code> parameter allows to specify the set of alignment file formats that should be used to retrieve the data. If 0 is passed as option, the list of alignment file formats defaults to <a class="el" href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats.">VRNA_FILE_FORMAT_MSA_DEFAULT</a>.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p><ul>
<li><a class="el" href="file_formats.html#msa-formats-clustal">ClustalW format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-fasta">FASTA (Pearson) format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-maf">MAF format</a></li>
</ul>
<dl class="section note"><dt>Note</dt><dd>After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing <a class="el" href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment.">VRNA_FILE_FORMAT_MSA_NOCHECK</a> in the <code>options</code> parameter.</dd>
<dd>
It is the users responsibility to free any memory occupied by the output arguments <code>names</code>, <code>aln</code>, <code>id</code>, and <code>structure</code> after calling this function. The function automatically sets the latter two arguments to <code>NULL</code> in case no corresponding data could be retrieved from the input alignment.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad4203a438622b2df7bc2f16578d54799" title="Read a multiple sequence alignment from file handle.">vrna_file_msa_read_record()</a>, <a class="el" href="group__file__formats__msa.html#ga79a23de2c7249f2cccd762e475c81859" title="Option flag indicating ClustalW formatted files.">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>, <a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>, <a class="el" href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604" title="Option flag indicating FASTA (Pearson) formatted files.">VRNA_FILE_FORMAT_MSA_FASTA</a>, <a class="el" href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927" title="Option flag indicating MAF formatted files.">VRNA_FILE_FORMAT_MSA_MAF</a>, <a class="el" href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats.">VRNA_FILE_FORMAT_MSA_DEFAULT</a>, <a class="el" href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment.">VRNA_FILE_FORMAT_MSA_NOCHECK</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>The name of input file that contains the alignment </td></tr>
    <tr><td class="paramname">names</td><td>An address to the pointer where sequence identifiers should be written to </td></tr>
    <tr><td class="paramname">aln</td><td>An address to the pointer where aligned sequences should be written to </td></tr>
    <tr><td class="paramname">id</td><td>An address to the pointer where the alignment ID should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">structure</td><td>An address to the pointer where consensus structure information should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences in the alignment, or -1 if no alignment record could be found</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000078">SWIG Wrapper Notes:</a></b></dt><dd><p class="startdd">In the target scripting language, only the first and last argument, <code>filename</code> and <code>options</code>, are passed to the corresponding function. The other arguments, which serve as output in the C-library, are available as additional return values. Hence, a function call in python may look like this:</p>
<p class="enddd"></p>
</dd></dl>
<div class="fragment"><div class="line">num_seq, names, aln, id, structure = RNA.file_msa_read(<span class="stringliteral">&quot;msa.stk&quot;</span>, RNA.FILE_FORMAT_MSA_STOCKHOLM)</div>
</div><!-- fragment --><p>After successfully reading the first record, the variable <code>num_seq</code> contains the number of sequences in the alignment (the actual return value of the C-function), while the variables <code>names</code>, <code>aln</code>, <code>id</code>, and <code>structure</code> are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the <code>SS_cons</code> line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.</p>
<p>This function exists as an overloaded version where the <code>options</code> parameter may be omitted! In that case, the <code>options</code> parameter defaults to <a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>. </p>

</div>
</div>
<a id="gad4203a438622b2df7bc2f16578d54799"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad4203a438622b2df7bc2f16578d54799">&#9670;&nbsp;</a></span>vrna_file_msa_read_record()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_msa_read_record </td>
          <td>(</td>
          <td class="paramtype">FILE *&#160;</td>
          <td class="paramname"><em>fp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char ***&#160;</td>
          <td class="paramname"><em>aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">char **&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Read a multiple sequence alignment from file handle. </p>
<p>Similar to <a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, this function reads a multiple sequence alignment from an input file handle. Since using a file handle, this function is not limited to the first alignment record, but allows for looping over all alignments within the input.</p>
<p>The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The <code>options</code> parameter allows to specify the alignment file format used to retrieve the data. A single format must be specified here, see <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a> for helping to determine the correct MSA file format.</p>
<p>Currently, the list of parsable multiple sequence alignment file formats consists of:</p><ul>
<li><a class="el" href="file_formats.html#msa-formats-clustal">ClustalW format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-fasta">FASTA (Pearson) format</a></li>
<li><a class="el" href="file_formats.html#msa-formats-maf">MAF format</a></li>
</ul>
<dl class="section note"><dt>Note</dt><dd>After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing <a class="el" href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment.">VRNA_FILE_FORMAT_MSA_NOCHECK</a> in the <code>options</code> parameter.</dd>
<dd>
It is the users responsibility to free any memory occupied by the output arguments <code>names</code>, <code>aln</code>, <code>id</code>, and <code>structure</code> after calling this function. The function automatically sets the latter two arguments to <code>NULL</code> in case no corresponding data could be retrieved from the input alignment.</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, <a class="el" href="group__file__formats__msa.html#gade4fa8136ebb2d0f7eb3f8b59a8658e3" title="Detect the format of a multiple sequence alignment file.">vrna_file_msa_detect_format()</a>, <a class="el" href="group__file__formats__msa.html#ga79a23de2c7249f2cccd762e475c81859" title="Option flag indicating ClustalW formatted files.">VRNA_FILE_FORMAT_MSA_CLUSTAL</a>, <a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>, <a class="el" href="group__file__formats__msa.html#gacf6274a2c825f34e4131404665b00604" title="Option flag indicating FASTA (Pearson) formatted files.">VRNA_FILE_FORMAT_MSA_FASTA</a>, <a class="el" href="group__file__formats__msa.html#gab3186d774c01570d2a47d0308eec4927" title="Option flag indicating MAF formatted files.">VRNA_FILE_FORMAT_MSA_MAF</a>, <a class="el" href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats.">VRNA_FILE_FORMAT_MSA_DEFAULT</a>, <a class="el" href="group__file__formats__msa.html#ga229fb3778ecabea4782902b69fa48fd1" title="Option flag to disable validation of the alignment.">VRNA_FILE_FORMAT_MSA_NOCHECK</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fp</td><td>The file pointer the data will be retrieved from </td></tr>
    <tr><td class="paramname">names</td><td>An address to the pointer where sequence identifiers should be written to </td></tr>
    <tr><td class="paramname">aln</td><td>An address to the pointer where aligned sequences should be written to </td></tr>
    <tr><td class="paramname">id</td><td>An address to the pointer where the alignment ID should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">structure</td><td>An address to the pointer where consensus structure information should be written to (Maybe NULL) </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The number of sequences in the alignment, or -1 if no alignment record could be found</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000079">SWIG Wrapper Notes:</a></b></dt><dd><p class="startdd">In the target scripting language, only the first and last argument, <code>fp</code> and <code>options</code>, are passed to the corresponding function. The other arguments, which serve as output in the C-library, are available as additional return values. Hence, a function call in python may look like this:</p>
<p class="enddd"></p>
</dd></dl>
<div class="fragment"><div class="line">f = open(<span class="stringliteral">&#39;msa.stk&#39;</span>, <span class="charliteral">&#39;r&#39;</span>)</div>
<div class="line">num_seq, names, aln, <span class="keywordtype">id</span>, structure = RNA.file_msa_read_record(f, RNA.FILE_FORMAT_MSA_STOCKHOLM)</div>
<div class="line">f.close() </div>
</div><!-- fragment --><p>After successfully reading the first record, the variable <code>num_seq</code> contains the number of sequences in the alignment (the actual return value of the C-function), while the variables <code>names</code>, <code>aln</code>, <code>id</code>, and <code>structure</code> are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the <code>SS_cons</code> line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.</p>
<p>This function exists as an overloaded version where the <code>options</code> parameter may be omitted! In that case, the <code>options</code> parameter defaults to <a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>. </p>

</div>
</div>
<a id="gade4fa8136ebb2d0f7eb3f8b59a8658e3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gade4fa8136ebb2d0f7eb3f8b59a8658e3">&#9670;&nbsp;</a></span>vrna_file_msa_detect_format()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">unsigned int vrna_file_msa_detect_format </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Detect the format of a multiple sequence alignment file. </p>
<p>This function attempts to determine the format of a file that supposedly contains a multiple sequence alignment (MSA). This is useful in cases where a MSA file contains more than a single record and therefore <a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a> can not be applied, since it only retrieves the first. Here, one can try to guess the correct file format using this function and then loop over the file, record by record using one of the low-level record retrieval functions for the corresponding MSA file format.</p>
<dl class="section note"><dt>Note</dt><dd>This function parses the entire first record within the specified file. As a result, it returns <a class="el" href="group__file__formats__msa.html#gabdc948f547e550125de3e7c65878400c" title="Return flag of vrna_file_msa_detect_format() to indicate unknown or malformatted alignment.">VRNA_FILE_FORMAT_MSA_UNKNOWN</a> not only if it can't detect the file's format, but also in cases where the file doesn't contain sequences!</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#gad02d5d12bda54611c915a1019323b7be" title="Read a multiple sequence alignment from file.">vrna_file_msa_read()</a>, vrna_file_stockholm_read_record(), vrna_file_clustal_read_record(), <a class="el" href="group__file__formats.html#ga8cfb7e271efc9e1f34640acb85475639">vrna_file_fasta_read_record()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>The name of input file that contains the alignment </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The MSA file format, or <a class="el" href="group__file__formats__msa.html#gabdc948f547e550125de3e7c65878400c" title="Return flag of vrna_file_msa_detect_format() to indicate unknown or malformatted alignment.">VRNA_FILE_FORMAT_MSA_UNKNOWN</a></dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000077">SWIG Wrapper Notes:</a></b></dt><dd>This function exists as an overloaded version where the <code>options</code> parameter may be omitted! In that case, the <code>options</code> parameter defaults to <a class="el" href="group__file__formats__msa.html#ga4acc255373831856a8417b68de8a94c0" title="Option flag indicating the set of default file formats.">VRNA_FILE_FORMAT_MSA_DEFAULT</a>. </dd></dl>

</div>
</div>
<a id="gaccca55b71d34def88e18ed9a51ccb2bf"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaccca55b71d34def88e18ed9a51ccb2bf">&#9670;&nbsp;</a></span>vrna_file_msa_write()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_file_msa_write </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>names</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char **&#160;</td>
          <td class="paramname"><em>aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>id</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>source</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>options</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="io_2file__formats__msa_8h.html">ViennaRNA/io/file_formats_msa.h</a>&gt;</code></p>

<p>Write multiple sequence alignment file. </p>
<dl class="section note"><dt>Note</dt><dd>Currently, we only support <a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a> output</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a>, <a class="el" href="group__file__formats__msa.html#ga1577ea0f497d9c501549c863a4f2c089" title="Option flag indicating to append data to a multiple sequence alignment file rather than overwriting i...">VRNA_FILE_FORMAT_MSA_APPEND</a>, <a class="el" href="group__file__formats__msa.html#ga494488a771aa0c602fb4cf445be34d47" title="Option flag indicating most informative sequence (MIS) output.">VRNA_FILE_FORMAT_MSA_MIS</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">filename</td><td>The output filename </td></tr>
    <tr><td class="paramname">names</td><td>The array of sequence names / identifies </td></tr>
    <tr><td class="paramname">aln</td><td>The array of aligned sequences </td></tr>
    <tr><td class="paramname">id</td><td>An optional ID for the alignment </td></tr>
    <tr><td class="paramname">structure</td><td>An optional consensus structure </td></tr>
    <tr><td class="paramname">source</td><td>A string describing the source of the alignment </td></tr>
    <tr><td class="paramname">options</td><td>Options to manipulate the behavior of this function </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Non-null upon successfully writing the alignment to file</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000080">SWIG Wrapper Notes:</a></b></dt><dd>In the target scripting language, this function exists as a set of overloaded versions, where the last four parameters may be omitted. If the <code>options</code> parameter is missing the options default to (<a class="el" href="group__file__formats__msa.html#ga62be992445cd8ab2ad7a8fded944338b" title="Option flag indicating Stockholm 1.0 formatted files.">VRNA_FILE_FORMAT_MSA_STOCKHOLM</a> | <a class="el" href="group__file__formats__msa.html#ga1577ea0f497d9c501549c863a4f2c089" title="Option flag indicating to append data to a multiple sequence alignment file rather than overwriting i...">VRNA_FILE_FORMAT_MSA_APPEND</a>). </dd></dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:34 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
