<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: Predicting various thermodynamic properties</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__thermodynamics.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#files">Files</a> &#124;
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#typedef-members">Typedefs</a>  </div>
  <div class="headertitle">
<div class="title">Predicting various thermodynamic properties<div class="ingroups"><a class="el" href="group__pf__fold.html">Partition Function and Equilibrium Properties</a> &raquo; <a class="el" href="group__part__func__global.html">Global Partition Function and Equilibrium Probabilities</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Compute various thermodynamic properties using the partition function.  
<a href="#details">More...</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Compute various thermodynamic properties using the partition function. </p>
<p>Many thermodynamic properties can be derived from the partition function </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ Q = \sum_{s \in \omega} e^{\frac{-E(s)}{kT}}. \]" src="form_43.png" width="92" height="33"/>
</p>
<p> In particular, for nucleic acids in equilibrium the probabilty <img class="formulaInl" alt="$ p(F) $" src="form_44.png" width="29" height="15"/> of a particular structural feature <img class="formulaInl" alt="$ F $" src="form_45.png" width="11" height="10"/> follows Boltzmanns law, i.e. </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ p(F) \propto \sum_{s \mid F \in s} e^{\frac{-E(s)}{kT}}. \]" src="form_46.png" width="121" height="36"/>
</p>
<p> The actual probabilities can then be obtained from the ratio of those structures containing <img class="formulaInl" alt="$ F $" src="form_45.png" width="11" height="10"/> and <em>all</em> structures, i.e. </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ p(F) = \frac{1}{Q} \sum_{s \mid F \in s} e^{\frac{-E(s)}{kT}}. \]" src="form_47.png" width="136" height="40"/>
</p>
<p>Consequently, a particular secondary structure <img class="formulaInl" alt="$ s $" src="form_35.png" width="6" height="7"/> has equilibrium probability </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ p(s) = \frac{1}{Q} e^{\frac{-E(s)}{kT}} \]" src="form_48.png" width="92" height="32"/>
</p>
<p> which can be easily computed once <img class="formulaInl" alt="$ Q $" src="form_49.png" width="11" height="14"/> and <img class="formulaInl" alt="$ E(s) $" src="form_36.png" width="26" height="15"/> are known.</p>
<p>On the other hand, efficient dynamic programming algorithms exist to compute the equilibrium probabilities </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ p_{ij} = \frac{1}{Q} \sum_{s \mid (i,j) \in s} e^{\frac{-E(s)}{kT}} \]" src="form_50.png" width="132" height="40"/>
</p>
<p> of base pairs <img class="formulaInl" alt="$ (i,j) $" src="form_13.png" width="26" height="15"/> without the need for exhaustive enumeration of <img class="formulaInl" alt="$ s $" src="form_35.png" width="6" height="7"/>.</p>
<p>This interface provides the functions for all thermodynamic property computations implemented in <em>RNAlib</em>. </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Predicting various thermodynamic properties:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__thermodynamics.svg" width="426" height="51"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:equilibrium__probs_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="equilibrium__probs_8h.html">equilibrium_probs.h</a></td></tr>
<tr class="memdesc:equilibrium__probs_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Equilibrium Probability implementations. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:heat__capacity_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="heat__capacity_8h.html">heat_capacity.h</a></td></tr>
<tr class="memdesc:heat__capacity_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute heat capacity for an RNA. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:structvrna__heat__capacity__s"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s">vrna_heat_capacity_s</a></td></tr>
<tr class="memdesc:structvrna__heat__capacity__s"><td class="mdescLeft">&#160;</td><td class="mdescRight">A single result from heat capacity computations.  <a href="group__thermodynamics.html#structvrna__heat__capacity__s">More...</a><br /></td></tr>
<tr class="separator:structvrna__heat__capacity__s"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:ga18316e4657502e9203931f924d93732f"><td class="memItemLeft" align="right" valign="top">typedef void()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga18316e4657502e9203931f924d93732f">vrna_heat_capacity_callback</a>(float temp, float heat_capacity, void *data)</td></tr>
<tr class="memdesc:ga18316e4657502e9203931f924d93732f"><td class="mdescLeft">&#160;</td><td class="mdescRight">The callback for heat capacity predictions.  <a href="group__thermodynamics.html#ga18316e4657502e9203931f924d93732f">More...</a><br /></td></tr>
<tr class="separator:ga18316e4657502e9203931f924d93732f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga99dc703d840310177ecf7aedf72b4409"><td class="memItemLeft" align="right" valign="top">typedef struct <a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s">vrna_heat_capacity_s</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a></td></tr>
<tr class="memdesc:ga99dc703d840310177ecf7aedf72b4409"><td class="mdescLeft">&#160;</td><td class="mdescRight">A single result from heat capacity computations.  <a href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">More...</a><br /></td></tr>
<tr class="separator:ga99dc703d840310177ecf7aedf72b4409"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Base pair probabilities and derived computations</h2></td></tr>
<tr class="memitem:ga55e52b8027cf50c081c452c4dec6135d"><td class="memItemLeft" align="right" valign="top"><a id="ga55e52b8027cf50c081c452c4dec6135d"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>vrna_pairing_probs</b> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, char *structure)</td></tr>
<tr class="separator:ga55e52b8027cf50c081c452c4dec6135d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gad3f0c240512e6d43e2e4d4c2076021f5"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gad3f0c240512e6d43e2e4d4c2076021f5">vrna_mean_bp_distance_pr</a> (int length, <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a> *<a class="el" href="fold__vars_8h.html#ac98ec419070aee6831b44e5c700f090f">pr</a>)</td></tr>
<tr class="memdesc:gad3f0c240512e6d43e2e4d4c2076021f5"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the mean base pair distance in the thermodynamic ensemble from a probability matrix.  <a href="group__thermodynamics.html#gad3f0c240512e6d43e2e4d4c2076021f5">More...</a><br /></td></tr>
<tr class="separator:gad3f0c240512e6d43e2e4d4c2076021f5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa6b8983b559b9ef4b2e1b31113ea317b"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gaa6b8983b559b9ef4b2e1b31113ea317b">vrna_mean_bp_distance</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc)</td></tr>
<tr class="memdesc:gaa6b8983b559b9ef4b2e1b31113ea317b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Get the mean base pair distance in the thermodynamic ensemble.  <a href="group__thermodynamics.html#gaa6b8983b559b9ef4b2e1b31113ea317b">More...</a><br /></td></tr>
<tr class="separator:gaa6b8983b559b9ef4b2e1b31113ea317b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga20b7a56fa9e451461805fd968c6cd909"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga20b7a56fa9e451461805fd968c6cd909">vrna_ensemble_defect_pt</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const short *pt)</td></tr>
<tr class="memdesc:ga20b7a56fa9e451461805fd968c6cd909"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the Ensemble Defect for a given target structure provided as a <b>vrna_ptable</b>.  <a href="group__thermodynamics.html#ga20b7a56fa9e451461805fd968c6cd909">More...</a><br /></td></tr>
<tr class="separator:ga20b7a56fa9e451461805fd968c6cd909"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaaf197722d1faa86af5e7b4240acafdee"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gaaf197722d1faa86af5e7b4240acafdee">vrna_ensemble_defect</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const char *structure)</td></tr>
<tr class="memdesc:gaaf197722d1faa86af5e7b4240acafdee"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the Ensemble Defect for a given target structure.  <a href="group__thermodynamics.html#gaaf197722d1faa86af5e7b4240acafdee">More...</a><br /></td></tr>
<tr class="separator:gaaf197722d1faa86af5e7b4240acafdee"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4ec346141028c7bcd15bc235c408b6e9"><td class="memItemLeft" align="right" valign="top">double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga4ec346141028c7bcd15bc235c408b6e9">vrna_positional_entropy</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc)</td></tr>
<tr class="memdesc:ga4ec346141028c7bcd15bc235c408b6e9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute a vector of positional entropies.  <a href="group__thermodynamics.html#ga4ec346141028c7bcd15bc235c408b6e9">More...</a><br /></td></tr>
<tr class="separator:ga4ec346141028c7bcd15bc235c408b6e9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga132664bf29fdc30bb5ea715491d1ab22"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga132664bf29fdc30bb5ea715491d1ab22">vrna_stack_prob</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, double cutoff)</td></tr>
<tr class="memdesc:ga132664bf29fdc30bb5ea715491d1ab22"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute stacking probabilities.  <a href="group__thermodynamics.html#ga132664bf29fdc30bb5ea715491d1ab22">More...</a><br /></td></tr>
<tr class="separator:ga132664bf29fdc30bb5ea715491d1ab22"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Multimer probabilities computations</h2></td></tr>
<tr class="memitem:gaa1e39e73afb51fbaf4ae38f0c066c46b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gaa1e39e73afb51fbaf4ae38f0c066c46b">vrna_pf_dimer_probs</a> (double FAB, double FA, double FB, <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *prAB, const <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *prA, const <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *prB, int Alength, const <a class="el" href="group__energy__parameters.html#ga01d8b92fe734df8d79a6169482c7d8d8">vrna_exp_param_t</a> *exp_params)</td></tr>
<tr class="memdesc:gaa1e39e73afb51fbaf4ae38f0c066c46b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute Boltzmann probabilities of dimerization without homodimers.  <a href="group__thermodynamics.html#gaa1e39e73afb51fbaf4ae38f0c066c46b">More...</a><br /></td></tr>
<tr class="separator:gaa1e39e73afb51fbaf4ae38f0c066c46b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Structure probability computations</h2></td></tr>
<tr class="memitem:ga882c35d9dd775c1275593b3b6a966bec"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga882c35d9dd775c1275593b3b6a966bec">vrna_pr_structure</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const char *structure)</td></tr>
<tr class="memdesc:ga882c35d9dd775c1275593b3b6a966bec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the equilibrium probability of a particular secondary structure.  <a href="group__thermodynamics.html#ga882c35d9dd775c1275593b3b6a966bec">More...</a><br /></td></tr>
<tr class="separator:ga882c35d9dd775c1275593b3b6a966bec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa88e17336daede0e6db4da1f3252c031"><td class="memItemLeft" align="right" valign="top">double&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gaa88e17336daede0e6db4da1f3252c031">vrna_pr_energy</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, double e)</td></tr>
<tr class="separator:gaa88e17336daede0e6db4da1f3252c031"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Basic heat capacity function interface</h2></td></tr>
<tr class="memitem:gafd95c2350d67d8f4374369c19cf111f1"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1">vrna_heat_capacity</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, float T_min, float T_max, float T_increment, unsigned int mpoints)</td></tr>
<tr class="memdesc:gafd95c2350d67d8f4374369c19cf111f1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the specific heat for an RNA.  <a href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1">More...</a><br /></td></tr>
<tr class="separator:gafd95c2350d67d8f4374369c19cf111f1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaca54de116562d187484e3fa884aeb52c"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#gaca54de116562d187484e3fa884aeb52c">vrna_heat_capacity_cb</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, float T_min, float T_max, float T_increment, unsigned int mpoints, <a class="el" href="group__thermodynamics.html#ga18316e4657502e9203931f924d93732f">vrna_heat_capacity_callback</a> *cb, void *data)</td></tr>
<tr class="memdesc:gaca54de116562d187484e3fa884aeb52c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the specific heat for an RNA (callback variant)  <a href="group__thermodynamics.html#gaca54de116562d187484e3fa884aeb52c">More...</a><br /></td></tr>
<tr class="separator:gaca54de116562d187484e3fa884aeb52c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader">Simplified heat capacity computation</h2></td></tr>
<tr class="memitem:ga0c2664e15dded704d805a9d2a625ff21"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#ga0c2664e15dded704d805a9d2a625ff21">vrna_heat_capacity_simple</a> (const char *sequence, float T_min, float T_max, float T_increment, unsigned int mpoints)</td></tr>
<tr class="memdesc:ga0c2664e15dded704d805a9d2a625ff21"><td class="mdescLeft">&#160;</td><td class="mdescRight">Compute the specific heat for an RNA (simplified variant)  <a href="group__thermodynamics.html#ga0c2664e15dded704d805a9d2a625ff21">More...</a><br /></td></tr>
<tr class="separator:ga0c2664e15dded704d805a9d2a625ff21"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<hr/><h2 class="groupheader">Data Structure Documentation</h2>
<a name="structvrna__heat__capacity__s" id="structvrna__heat__capacity__s"></a>
<h2 class="memtitle"><span class="permalink"><a href="#structvrna__heat__capacity__s">&#9670;&nbsp;</a></span>vrna_heat_capacity_s</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">struct vrna_heat_capacity_s</td>
        </tr>
      </table>
</div><div class="memdoc">
<div class="textblock"><p>A single result from heat capacity computations. </p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a> </dd></dl>
</div><table class="memberdecls">
<tr><td colspan="2"><h3>Data Fields</h3></td></tr>
<tr class="memitem:a62ed3466ce614f4f1e729463fac41d72"><td class="memItemLeft" align="right" valign="top"><a id="a62ed3466ce614f4f1e729463fac41d72"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#a62ed3466ce614f4f1e729463fac41d72">temperature</a></td></tr>
<tr class="memdesc:a62ed3466ce614f4f1e729463fac41d72"><td class="mdescLeft">&#160;</td><td class="mdescRight">The temperature in &deg;C. <br /></td></tr>
<tr class="separator:a62ed3466ce614f4f1e729463fac41d72"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2987d5e5ef974d14a108ec1c7062562f"><td class="memItemLeft" align="right" valign="top"><a id="a2987d5e5ef974d14a108ec1c7062562f"></a>
float&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__thermodynamics.html#a2987d5e5ef974d14a108ec1c7062562f">heat_capacity</a></td></tr>
<tr class="memdesc:a2987d5e5ef974d14a108ec1c7062562f"><td class="mdescLeft">&#160;</td><td class="mdescRight">The specific heat at this temperature in Kcal/(Mol * K) <br /></td></tr>
<tr class="separator:a2987d5e5ef974d14a108ec1c7062562f"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>

</div>
</div>
<h2 class="groupheader">Typedef Documentation</h2>
<a id="ga18316e4657502e9203931f924d93732f"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga18316e4657502e9203931f924d93732f">&#9670;&nbsp;</a></span>vrna_heat_capacity_callback</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef void() vrna_heat_capacity_callback(float temp, float heat_capacity, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="heat__capacity_8h.html">ViennaRNA/heat_capacity.h</a>&gt;</code></p>

<p>The callback for heat capacity predictions. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000004">Notes on Callback Functions:</a></b></dt><dd>This function will be called for each evaluated temperature in the heat capacity prediction. </dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__thermodynamics.html#gaca54de116562d187484e3fa884aeb52c" title="Compute the specific heat for an RNA (callback variant)">vrna_heat_capacity_cb()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">temp</td><td>The current temperature this results corresponds to in &deg;C </td></tr>
    <tr><td class="paramname">heat_capacity</td><td>The heat capacity in Kcal/(Mol * K) </td></tr>
    <tr><td class="paramname">data</td><td>Some arbitrary data pointer passed through by the function executing the callback </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga99dc703d840310177ecf7aedf72b4409"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga99dc703d840310177ecf7aedf72b4409">&#9670;&nbsp;</a></span>vrna_heat_capacity_t</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef struct <a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s">vrna_heat_capacity_s</a> <a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="heat__capacity_8h.html">ViennaRNA/heat_capacity.h</a>&gt;</code></p>

<p>A single result from heat capacity computations. </p>
<p>This is a convenience typedef for <a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s" title="A single result from heat capacity computations.">vrna_heat_capacity_s</a>, i.e. results as obtained from <a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a> </p>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="gad3f0c240512e6d43e2e4d4c2076021f5"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gad3f0c240512e6d43e2e4d4c2076021f5">&#9670;&nbsp;</a></span>vrna_mean_bp_distance_pr()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_mean_bp_distance_pr </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>length</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a> *&#160;</td>
          <td class="paramname"><em>pr</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Get the mean base pair distance in the thermodynamic ensemble from a probability matrix. </p>
<p class="formulaDsp">
<img class="formulaDsp" alt="\[ &lt;d&gt; = \sum_{a,b} p_a p_b d(S_a,S_b) \]" src="form_51.png" width="150" height="33"/>
</p>
<p> this can be computed from the pair probs <img class="formulaInl" alt="$ p_{ij} $" src="form_52.png" width="17" height="11"/> as </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ &lt;d&gt; = \sum_{ij} p_{ij}(1-p_{ij}) \]" src="form_53.png" width="139" height="32"/>
</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">length</td><td>The length of the sequence </td></tr>
    <tr><td class="paramname">pr</td><td>The matrix containing the base pair probabilities </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The mean pair distance of the structure ensemble </dd></dl>

</div>
</div>
<a id="gaa6b8983b559b9ef4b2e1b31113ea317b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa6b8983b559b9ef4b2e1b31113ea317b">&#9670;&nbsp;</a></span>vrna_mean_bp_distance()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_mean_bp_distance </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Get the mean base pair distance in the thermodynamic ensemble. </p>
<p class="formulaDsp">
<img class="formulaDsp" alt="\[ &lt;d&gt; = \sum_{a,b} p_a p_b d(S_a,S_b) \]" src="form_51.png" width="150" height="33"/>
</p>
<p> this can be computed from the pair probs <img class="formulaInl" alt="$p_{ij}$" src="form_54.png" width="17" height="11"/> as </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ &lt;d&gt; = \sum_{ij} p_{ij}(1-p_{ij}) \]" src="form_53.png" width="139" height="32"/>
</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The fold compound data structure </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The mean pair distance of the structure ensemble</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000115">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b><a class="el" href="group__part__func__global__deprecated.html#ga79cbc375af65f11609feb6b055269e7d" title="Get the mean base pair distance of the last partition function computation.">mean_bp_distance()</a></b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga20b7a56fa9e451461805fd968c6cd909"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga20b7a56fa9e451461805fd968c6cd909">&#9670;&nbsp;</a></span>vrna_ensemble_defect_pt()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_ensemble_defect_pt </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const short *&#160;</td>
          <td class="paramname"><em>pt</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute the Ensemble Defect for a given target structure provided as a <b>vrna_ptable</b>. </p>
<p>Given a target structure <img class="formulaInl" alt="$s$" src="form_24.png" width="6" height="7"/>, compute the average dissimilarity of a randomly drawn structure from the ensemble, i.e.: </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ ED(s) = 1 - \frac{1}{n} \sum_{ij, (i,j) \in s} p_{ij} - \frac{1}{n} \sum_{i}(1 - s_i)q_i \]" src="form_55.png" width="265" height="40"/>
</p>
<p> with sequence length <img class="formulaInl" alt="$n$" src="form_56.png" width="10" height="7"/>, the probability <img class="formulaInl" alt="$p_{ij}$" src="form_54.png" width="17" height="11"/> of a base pair <img class="formulaInl" alt="$(i,j)$" src="form_57.png" width="26" height="15"/>, the probability <img class="formulaInl" alt="$q_i = 1 - \sum_j p_{ij}$" src="form_58.png" width="91" height="18"/> of nucleotide <img class="formulaInl" alt="$i$" src="form_59.png" width="6" height="10"/> being unpaired, and the indicator variable <img class="formulaInl" alt="$s_i = 1$" src="form_60.png" width="36" height="12"/> if <img class="formulaInl" alt="$\exists (i,j) \in s$" src="form_61.png" width="58" height="15"/>, and <img class="formulaInl" alt="$s_i = 0$" src="form_62.png" width="36" height="12"/> otherwise.</p>
<dl class="section pre"><dt>Precondition</dt><dd>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> input parameter <code>fc</code> must contain a valid base pair probability matrix. This means that partition function and base pair probabilities must have been computed using <code>fc</code> before execution of this function!</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__part__func__global.html#ga29e256d688ad221b78d37f427e0e99bc" title="Compute the partition function  for a given RNA sequence, or sequence alignment.">vrna_pf()</a>, vrna_pairing_probs(), <a class="el" href="group__thermodynamics.html#gaaf197722d1faa86af5e7b4240acafdee" title="Compute the Ensemble Defect for a given target structure.">vrna_ensemble_defect()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>A fold_compound with pre-computed base pair probabilities </td></tr>
    <tr><td class="paramname">pt</td><td>A pair table representing a target structure </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The ensemble defect with respect to the target structure, or -1. upon failure, e.g. pre-conditions are not met</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000116">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as overloaded method <b>ensemble_defect()</b> to objects of type <em>fold_compound</em>. </dd></dl>

</div>
</div>
<a id="gaaf197722d1faa86af5e7b4240acafdee"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaaf197722d1faa86af5e7b4240acafdee">&#9670;&nbsp;</a></span>vrna_ensemble_defect()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_ensemble_defect </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute the Ensemble Defect for a given target structure. </p>
<p>This is a wrapper around <b><a class="el" href="group__thermodynamics.html#ga20b7a56fa9e451461805fd968c6cd909" title="Compute the Ensemble Defect for a given target structure provided as a vrna_ptable.">vrna_ensemble_defect_pt()</a></b>. Given a target structure <img class="formulaInl" alt="$s$" src="form_24.png" width="6" height="7"/>, compute the average dissimilarity of a randomly drawn structure from the ensemble, i.e.: </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ ED(s) = 1 - \frac{1}{n} \sum_{ij, (i,j) \in s} p_{ij} - \frac{1}{n} \sum_{i}(1 - s_i)q_i \]" src="form_55.png" width="265" height="40"/>
</p>
<p> with sequence length <img class="formulaInl" alt="$n$" src="form_56.png" width="10" height="7"/>, the probability <img class="formulaInl" alt="$p_{ij}$" src="form_54.png" width="17" height="11"/> of a base pair <img class="formulaInl" alt="$(i,j)$" src="form_57.png" width="26" height="15"/>, the probability <img class="formulaInl" alt="$q_i = 1 - \sum_j p_{ij}$" src="form_58.png" width="91" height="18"/> of nucleotide <img class="formulaInl" alt="$i$" src="form_59.png" width="6" height="10"/> being unpaired, and the indicator variable <img class="formulaInl" alt="$s_i = 1$" src="form_60.png" width="36" height="12"/> if <img class="formulaInl" alt="$\exists (i,j) \in s$" src="form_61.png" width="58" height="15"/>, and <img class="formulaInl" alt="$s_i = 0$" src="form_62.png" width="36" height="12"/> otherwise.</p>
<dl class="section pre"><dt>Precondition</dt><dd>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> input parameter <code>fc</code> must contain a valid base pair probability matrix. This means that partition function and base pair probabilities must have been computed using <code>fc</code> before execution of this function!</dd></dl>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__part__func__global.html#ga29e256d688ad221b78d37f427e0e99bc" title="Compute the partition function  for a given RNA sequence, or sequence alignment.">vrna_pf()</a>, vrna_pairing_probs(), <a class="el" href="group__thermodynamics.html#ga20b7a56fa9e451461805fd968c6cd909" title="Compute the Ensemble Defect for a given target structure provided as a vrna_ptable.">vrna_ensemble_defect_pt()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>A fold_compound with pre-computed base pair probabilities </td></tr>
    <tr><td class="paramname">structure</td><td>A target structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The ensemble defect with respect to the target structure, or -1. upon failure, e.g. pre-conditions are not met</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000117">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>ensemble_defect()</b> to objects of type <em>fold_compound</em>. Note that the SWIG wrapper takes a structure in dot-bracket notation and converts it into a pair table using <a class="el" href="group__struct__utils__pair__table.html#ga2c4f27a8e277f21f0d2cc10bc9a99420" title="Create a pair table for a secondary structure string.">vrna_ptable_from_string()</a>. The resulting pair table is then internally passed to <a class="el" href="group__thermodynamics.html#ga20b7a56fa9e451461805fd968c6cd909" title="Compute the Ensemble Defect for a given target structure provided as a vrna_ptable.">vrna_ensemble_defect_pt()</a>. To control which kind of matching brackets will be used during conversion, the optional argument <code>options</code> can be used. See also the description of <a class="el" href="group__struct__utils__pair__table.html#ga2c4f27a8e277f21f0d2cc10bc9a99420" title="Create a pair table for a secondary structure string.">vrna_ptable_from_string()</a> for available options. (default: <b>VRNA_BRACKETS_RND</b>). </dd></dl>

</div>
</div>
<a id="ga4ec346141028c7bcd15bc235c408b6e9"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga4ec346141028c7bcd15bc235c408b6e9">&#9670;&nbsp;</a></span>vrna_positional_entropy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double * vrna_positional_entropy </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute a vector of positional entropies. </p>
<p>This function computes the positional entropies from base pair probabilities as </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ S(i) = - \sum_j p_{ij} \log(p_{ij}) - q_i \log(q_i) \]" src="form_63.png" width="210" height="32"/>
</p>
<p> with unpaired probabilities <img class="formulaInl" alt="$ q_i = 1 - \sum_j p_{ij} $" src="form_64.png" width="91" height="18"/>.</p>
<p>Low entropy regions have little structural flexibility and the reliability of the predicted structure is high. High entropy implies many structural alternatives. While these alternatives may be functionally important, they make structure prediction more difficult and thus less reliable.</p>
<dl class="section pre"><dt>Precondition</dt><dd>This function requires pre-computed base pair probabilities! Thus, <a class="el" href="group__part__func__global.html#ga29e256d688ad221b78d37f427e0e99bc" title="Compute the partition function  for a given RNA sequence, or sequence alignment.">vrna_pf()</a> must be called beforehand.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>A fold_compound with pre-computed base pair probabilities </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A 1-based vector of positional entropies <img class="formulaInl" alt="$ S(i) $" src="form_65.png" width="23" height="15"/>. (position 0 contains the sequence length)</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000118">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>positional_entropy()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga132664bf29fdc30bb5ea715491d1ab22"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga132664bf29fdc30bb5ea715491d1ab22">&#9670;&nbsp;</a></span>vrna_stack_prob()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a>* vrna_stack_prob </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>cutoff</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute stacking probabilities. </p>
<p>For each possible base pair <img class="formulaInl" alt="$(i,j)$" src="form_57.png" width="26" height="15"/>, compute the probability of a stack <img class="formulaInl" alt="$(i,j)$" src="form_57.png" width="26" height="15"/>, <img class="formulaInl" alt="$(i+1, j-1)$" src="form_66.png" width="74" height="15"/>.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The fold compound data structure with precomputed base pair probabilities </td></tr>
    <tr><td class="paramname">cutoff</td><td>A cutoff value that limits the output to stacks with <img class="formulaInl" alt="$ p &gt; \textrm{cutoff} $" src="form_67.png" width="62" height="14"/>. </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of stacks with enclosing base pair <img class="formulaInl" alt="$(i,j)$" src="form_57.png" width="26" height="15"/> and probabiltiy <img class="formulaInl" alt="$ p $" src="form_68.png" width="8" height="10"/> </dd></dl>

</div>
</div>
<a id="gaa1e39e73afb51fbaf4ae38f0c066c46b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa1e39e73afb51fbaf4ae38f0c066c46b">&#9670;&nbsp;</a></span>vrna_pf_dimer_probs()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_pf_dimer_probs </td>
          <td>(</td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>FAB</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>FA</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>FB</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *&#160;</td>
          <td class="paramname"><em>prAB</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *&#160;</td>
          <td class="paramname"><em>prA</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *&#160;</td>
          <td class="paramname"><em>prB</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>Alength</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="group__energy__parameters.html#ga01d8b92fe734df8d79a6169482c7d8d8">vrna_exp_param_t</a> *&#160;</td>
          <td class="paramname"><em>exp_params</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute Boltzmann probabilities of dimerization without homodimers. </p>
<p>Given the pair probabilities and free energies (in the null model) for a dimer AB and the two constituent monomers A and B, compute the conditional pair probabilities given that a dimer AB actually forms. Null model pair probabilities are given as a list as produced by <a class="el" href="group__part__func__global.html#ga94f6efc0b8d8712b023452794a0a5bd2" title="Create a vrna_ep_t from base pair probability matrix.">vrna_plist_from_probs()</a>, the dimer probabilities 'prAB' are modified in place.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">FAB</td><td>free energy of dimer AB </td></tr>
    <tr><td class="paramname">FA</td><td>free energy of monomer A </td></tr>
    <tr><td class="paramname">FB</td><td>free energy of monomer B </td></tr>
    <tr><td class="paramname">prAB</td><td>pair probabilities for dimer </td></tr>
    <tr><td class="paramname">prA</td><td>pair probabilities monomer </td></tr>
    <tr><td class="paramname">prB</td><td>pair probabilities monomer </td></tr>
    <tr><td class="paramname">Alength</td><td>Length of molecule A </td></tr>
    <tr><td class="paramname">exp_params</td><td>The precomputed Boltzmann factors </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="ga882c35d9dd775c1275593b3b6a966bec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga882c35d9dd775c1275593b3b6a966bec">&#9670;&nbsp;</a></span>vrna_pr_structure()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_pr_structure </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>

<p>Compute the equilibrium probability of a particular secondary structure. </p>
<p>The probability <img class="formulaInl" alt="$p(s)$" src="form_23.png" width="25" height="15"/> of a particular secondary structure <img class="formulaInl" alt="$s$" src="form_24.png" width="6" height="7"/> can be computed as </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ p(s) = \frac{exp(-\beta E(s)}{Z} \]" src="form_69.png" width="120" height="30"/>
</p>
<p> from the structures free energy <img class="formulaInl" alt="$E(s)$" src="form_25.png" width="26" height="15"/> and the partition function </p><p class="formulaDsp">
<img class="formulaDsp" alt="\[ Z = \sum_s exp(-\beta E(s)),\quad\mathrm{with}\quad\beta = \frac{1}{RT} \]" src="form_70.png" width="239" height="36"/>
</p>
<p> where <img class="formulaInl" alt="$R$" src="form_71.png" width="11" height="11"/> is the gas constant and <img class="formulaInl" alt="$T$" src="form_28.png" width="11" height="10"/> the thermodynamic temperature.</p>
<dl class="section pre"><dt>Precondition</dt><dd>The fold compound <code>fc</code> must have went through a call to <a class="el" href="group__part__func__global.html#ga29e256d688ad221b78d37f427e0e99bc" title="Compute the partition function  for a given RNA sequence, or sequence alignment.">vrna_pf()</a> to fill the dynamic programming matrices with the corresponding partition function.</dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The fold compound data structure with precomputed partition function </td></tr>
    <tr><td class="paramname">structure</td><td>The secondary structure to compute the probability for in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The probability of the input structure (range <img class="formulaInl" alt="$[0:1]$" src="form_72.png" width="32" height="15"/>)</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000119">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>pr_structure()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="gaa88e17336daede0e6db4da1f3252c031"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa88e17336daede0e6db4da1f3252c031">&#9670;&nbsp;</a></span>vrna_pr_energy()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">double vrna_pr_energy </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>e</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="equilibrium__probs_8h.html">ViennaRNA/equilibrium_probs.h</a>&gt;</code></p>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000120">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>pr_energy()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="gafd95c2350d67d8f4374369c19cf111f1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gafd95c2350d67d8f4374369c19cf111f1">&#9670;&nbsp;</a></span>vrna_heat_capacity()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a> * vrna_heat_capacity </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_min</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_max</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_increment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>mpoints</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="heat__capacity_8h.html">ViennaRNA/heat_capacity.h</a>&gt;</code></p>

<p>Compute the specific heat for an RNA. </p>
<p>This function computes an RNAs specific heat in a given temperature range from the partition function by numeric differentiation. The result is returned as a list of pairs of temperature in &deg;C and specific heat in Kcal/(Mol*K).</p>
<p>Users can specify the temperature range for the computation from <code>T_min</code> to <code>T_max</code>, as well as the increment step size <code>T_increment</code>. The latter also determines how many times the partition function is computed. Finally, the parameter <code>mpoints</code> determines how smooth the curve should be. The algorithm itself fits a parabola to <img class="formulaInl" alt="$ 2 \cdot mpoints + 1 $" src="form_81.png" width="89" height="12"/> data points to calculate 2nd derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__thermodynamics.html#gaca54de116562d187484e3fa884aeb52c" title="Compute the specific heat for an RNA (callback variant)">vrna_heat_capacity_cb()</a>, <a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409" title="A single result from heat capacity computations.">vrna_heat_capacity_t</a>, <a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s" title="A single result from heat capacity computations.">vrna_heat_capacity_s</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> with the RNA sequence to analyze </td></tr>
    <tr><td class="paramname">T_min</td><td>Lowest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_max</td><td>Highest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_increment</td><td>Stepsize for temperature incrementation in &deg;C (a reasonable choice might be 1&deg;C) </td></tr>
    <tr><td class="paramname">mpoints</td><td>The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2, min: 1, max: 100) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of pairs of temperatures and corresponding heat capacity or <em>NULL</em> upon any failure. The last entry of the list is indicated by a <b>temperature</b> field set to a value smaller than <code>T_min</code> </dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000088">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as overloaded method <b>heat_capacity()</b> to objects of type <em>fold_compound</em>. If the optional function arguments <code>T_min</code>, <code>T_max</code>, <code>T_increment</code>, and <code>mpoints</code> are omitted, they default to 0.0, 100.0, 1.0 and 2, respectively. </dd></dl>

</div>
</div>
<a id="gaca54de116562d187484e3fa884aeb52c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaca54de116562d187484e3fa884aeb52c">&#9670;&nbsp;</a></span>vrna_heat_capacity_cb()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int vrna_heat_capacity_cb </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_min</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_max</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_increment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>mpoints</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__thermodynamics.html#ga18316e4657502e9203931f924d93732f">vrna_heat_capacity_callback</a> *&#160;</td>
          <td class="paramname"><em>cb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>data</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="heat__capacity_8h.html">ViennaRNA/heat_capacity.h</a>&gt;</code></p>

<p>Compute the specific heat for an RNA (callback variant) </p>
<p>Similar to <a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a>, this function computes an RNAs specific heat in a given temperature range from the partition function by numeric differentiation. Instead of returning a list of temperature/specific heat pairs, however, this function returns the individual results through a callback mechanism. The provided function will be called for each result and passed the corresponding temperature and specific heat values along with the arbitrary data as provided through the <code>data</code> pointer argument.</p>
<p>Users can specify the temperature range for the computation from <code>T_min</code> to <code>T_max</code>, as well as the increment step size <code>T_increment</code>. The latter also determines how many times the partition function is computed. Finally, the parameter <code>mpoints</code> determines how smooth the curve should be. The algorithm itself fits a parabola to <img class="formulaInl" alt="$ 2 \cdot mpoints + 1 $" src="form_81.png" width="89" height="12"/> data points to calculate 2nd derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a>, <a class="el" href="group__thermodynamics.html#ga18316e4657502e9203931f924d93732f" title="The callback for heat capacity predictions.">vrna_heat_capacity_callback</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> with the RNA sequence to analyze </td></tr>
    <tr><td class="paramname">T_min</td><td>Lowest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_max</td><td>Highest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_increment</td><td>Stepsize for temperature incrementation in &deg;C (a reasonable choice might be 1&deg;C) </td></tr>
    <tr><td class="paramname">mpoints</td><td>The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2, min: 1, max: 100) </td></tr>
    <tr><td class="paramname">cb</td><td>The user-defined callback function that receives the individual results </td></tr>
    <tr><td class="paramname">data</td><td>An arbitrary data structure that will be passed to the callback in conjunction with the results </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>Returns 0 upon failure, and non-zero otherwise</dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000089">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>heat_capacity_cb()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga0c2664e15dded704d805a9d2a625ff21"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga0c2664e15dded704d805a9d2a625ff21">&#9670;&nbsp;</a></span>vrna_heat_capacity_simple()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409">vrna_heat_capacity_t</a> * vrna_heat_capacity_simple </td>
          <td>(</td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>sequence</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_min</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_max</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">float&#160;</td>
          <td class="paramname"><em>T_increment</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>mpoints</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="heat__capacity_8h.html">ViennaRNA/heat_capacity.h</a>&gt;</code></p>

<p>Compute the specific heat for an RNA (simplified variant) </p>
<p>Similar to <a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a>, this function computes an RNAs specific heat in a given temperature range from the partition function by numeric differentiation. This simplified version, however, only requires the RNA sequence as input instead of a vrna_fold_compound_t data structure. The result is returned as a list of pairs of temperature in &deg;C and specific heat in Kcal/(Mol*K).</p>
<p>Users can specify the temperature range for the computation from <code>T_min</code> to <code>T_max</code>, as well as the increment step size <code>T_increment</code>. The latter also determines how many times the partition function is computed. Finally, the parameter <code>mpoints</code> determines how smooth the curve should be. The algorithm itself fits a parabola to <img class="formulaInl" alt="$ 2 \cdot mpoints + 1 $" src="form_81.png" width="89" height="12"/> data points to calculate 2nd derivatives. Increasing this parameter produces a smoother curve.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__thermodynamics.html#gafd95c2350d67d8f4374369c19cf111f1" title="Compute the specific heat for an RNA.">vrna_heat_capacity()</a>, <a class="el" href="group__thermodynamics.html#gaca54de116562d187484e3fa884aeb52c" title="Compute the specific heat for an RNA (callback variant)">vrna_heat_capacity_cb()</a>, <a class="el" href="group__thermodynamics.html#ga99dc703d840310177ecf7aedf72b4409" title="A single result from heat capacity computations.">vrna_heat_capacity_t</a>, <a class="el" href="group__thermodynamics.html#structvrna__heat__capacity__s" title="A single result from heat capacity computations.">vrna_heat_capacity_s</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">sequence</td><td>The RNA sequence input (must be uppercase) </td></tr>
    <tr><td class="paramname">T_min</td><td>Lowest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_max</td><td>Highest temperature in &deg;C </td></tr>
    <tr><td class="paramname">T_increment</td><td>Stepsize for temperature incrementation in &deg;C (a reasonable choice might be 1&deg;C) </td></tr>
    <tr><td class="paramname">mpoints</td><td>The number of interpolation points to calculate 2nd derivative (a reasonable choice might be 2, min: 1, max: 100) </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of pairs of temperatures and corresponding heat capacity or <em>NULL</em> upon any failure. The last entry of the list is indicated by a <b>temperature</b> field set to a value smaller than <code>T_min</code> </dd></dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000090">SWIG Wrapper Notes:</a></b></dt><dd>This function is available as overloaded function <b>heat_capacity()</b>. If the optional function arguments <code>T_min</code>, <code>T_max</code>, <code>T_increment</code>, and <code>mpoints</code> are omitted, they default to 0.0, 100.0, 1.0 and 2, respectively. </dd></dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:33 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
