<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: Unstructured Domains</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('group__domains__up.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#files">Files</a> &#124;
<a href="#nested-classes">Data Structures</a> &#124;
<a href="#define-members">Macros</a> &#124;
<a href="#typedef-members">Typedefs</a> &#124;
<a href="#func-members">Functions</a>  </div>
  <div class="headertitle">
<div class="title">Unstructured Domains<div class="ingroups"><a class="el" href="group__grammar.html">The RNA Folding Grammar</a> &raquo; <a class="el" href="group__domains.html">Extending the Folding Grammar with Additional Domains</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p>Add and modify unstructured domains to the RNA folding grammar.  
<a href="#details">More...</a></p>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<p>Add and modify unstructured domains to the RNA folding grammar. </p>
<p>This module provides the tools to add and modify unstructured domains to the production rules of the RNA folding grammar. Usually this functionality is utilized for incorporating ligand binding to unpaired stretches of an RNA.</p>
<dl class="bug"><dt><b><a class="el" href="bug.html#_bug000003">Bug:</a></b></dt><dd>Although the additional production rule(s) for unstructured domains as descibed in <a class="el" href="folding_grammar.html#sec_domains_up">Unstructured Domains</a> are always treated as 'segments possibly bound to one or more ligands', the current implementation requires that at least one ligand is bound. The default implementation already takes care of the required changes, however, upon using callback functions other than the default ones, one has to take care of this fact. Please also note, that this behavior might change in one of the next releases, such that the decomposition schemes as shown above comply with the actual implementation.</dd></dl>
<p>A default implementation allows one to readily use this feature by simply adding sequence motifs and corresponding binding free energies with the function <a class="el" href="group__domains__up.html#ga55f7de5ef5b7472b0eeab9296b57f671" title="Add an unstructured domain motif, e.g. for ligand binding.">vrna_ud_add_motif()</a> (see also <a class="el" href="group__ligands__up.html">Ligands Binding to Unstructured Domains</a>).</p>
<p>The grammar extension is realized using a callback function that</p><ul>
<li>evaluates the binding free energy of a ligand to its target sequence segment (white boxes in the figures above), or</li>
<li>returns the free energy of an unpaired stretch possibly bound by a ligand, stored in the additional <em>U</em> DP matrix.</li>
</ul>
<p>The callback is passed the segment positions, the loop context, and which of the two above mentioned evaluations are required. A second callback implements the pre-processing step that prepares the <em>U</em> DP matrix by evaluating all possible cases of the additional production rule. Both callbacks have a default implementation in <em>RNAlib</em>, but may be over-written by a user-implementation, making it fully user-customizable.</p>
<p>For equilibrium probability computations, two additional callbacks exist. One to store/add and one to retrieve the probability of unstructured domains at particular positions. Our implementation already takes care of computing the probabilities, but users of the unstructured domain feature are required to provide a mechanism to efficiently store/add the corresponding values into some external data structure. </p>
<div id="dynsection-0" onclick="return toggleVisibility(this)" class="dynheader closed" style="cursor:pointer;">
  <img id="dynsection-0-trigger" src="closed.png" alt="+"/> Collaboration diagram for Unstructured Domains:</div>
<div id="dynsection-0-summary" class="dynsummary" style="display:block;">
</div>
<div id="dynsection-0-content" class="dyncontent" style="display:none;">
<div class="center"><iframe scrolling="no" frameborder="0" src="group__domains__up.svg" width="380" height="66"><p><b>This browser is not able to show SVG: try Firefox, Chrome, Safari, or Opera instead.</b></p></iframe>
</div>
</div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="files"></a>
Files</h2></td></tr>
<tr class="memitem:unstructured__domains_8h"><td class="memItemLeft" align="right" valign="top">file &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="unstructured__domains_8h.html">unstructured_domains.h</a></td></tr>
<tr class="memdesc:unstructured__domains_8h"><td class="mdescLeft">&#160;</td><td class="mdescRight">Functions to modify unstructured domains, e.g. to incorporate ligands binding to unpaired stretches. <br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Data Structures</h2></td></tr>
<tr class="memitem:structvrna__unstructured__domain__s"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#structvrna__unstructured__domain__s">vrna_unstructured_domain_s</a></td></tr>
<tr class="memdesc:structvrna__unstructured__domain__s"><td class="mdescLeft">&#160;</td><td class="mdescRight">Data structure to store all functionality for ligand binding.  <a href="group__domains__up.html#structvrna__unstructured__domain__s">More...</a><br /></td></tr>
<tr class="separator:structvrna__unstructured__domain__s"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="define-members"></a>
Macros</h2></td></tr>
<tr class="memitem:gaac911374e86236a51bfd42e1f098eaba"><td class="memItemLeft" align="right" valign="top"><a id="gaac911374e86236a51bfd42e1f098eaba"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gaac911374e86236a51bfd42e1f098eaba">VRNA_UNSTRUCTURED_DOMAIN_EXT_LOOP</a>&#160;&#160;&#160;1U</td></tr>
<tr class="memdesc:gaac911374e86236a51bfd42e1f098eaba"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand bound to unpiared stretch in the exterior loop. <br /></td></tr>
<tr class="separator:gaac911374e86236a51bfd42e1f098eaba"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga23b610ea9564346c45cc1e2bbb62adf7"><td class="memItemLeft" align="right" valign="top"><a id="ga23b610ea9564346c45cc1e2bbb62adf7"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga23b610ea9564346c45cc1e2bbb62adf7">VRNA_UNSTRUCTURED_DOMAIN_HP_LOOP</a>&#160;&#160;&#160;2U</td></tr>
<tr class="memdesc:ga23b610ea9564346c45cc1e2bbb62adf7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand bound to unpaired stretch in a hairpin loop. <br /></td></tr>
<tr class="separator:ga23b610ea9564346c45cc1e2bbb62adf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac4a0feccd9654c149203200248c2716b"><td class="memItemLeft" align="right" valign="top"><a id="gac4a0feccd9654c149203200248c2716b"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gac4a0feccd9654c149203200248c2716b">VRNA_UNSTRUCTURED_DOMAIN_INT_LOOP</a>&#160;&#160;&#160;4U</td></tr>
<tr class="memdesc:gac4a0feccd9654c149203200248c2716b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand bound to unpiared stretch in an interior loop. <br /></td></tr>
<tr class="separator:gac4a0feccd9654c149203200248c2716b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga67b80796655e5227a4ed662bfbe398b0"><td class="memItemLeft" align="right" valign="top"><a id="ga67b80796655e5227a4ed662bfbe398b0"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga67b80796655e5227a4ed662bfbe398b0">VRNA_UNSTRUCTURED_DOMAIN_MB_LOOP</a>&#160;&#160;&#160;8U</td></tr>
<tr class="memdesc:ga67b80796655e5227a4ed662bfbe398b0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand bound to unpiared stretch in a multibranch loop. <br /></td></tr>
<tr class="separator:ga67b80796655e5227a4ed662bfbe398b0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaab12b58d59be76446a9f76fad2fe624c"><td class="memItemLeft" align="right" valign="top"><a id="gaab12b58d59be76446a9f76fad2fe624c"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gaab12b58d59be76446a9f76fad2fe624c">VRNA_UNSTRUCTURED_DOMAIN_MOTIF</a>&#160;&#160;&#160;16U</td></tr>
<tr class="memdesc:gaab12b58d59be76446a9f76fad2fe624c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand binding without additional unbound nucleotides (motif-only) <br /></td></tr>
<tr class="separator:gaab12b58d59be76446a9f76fad2fe624c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga3c6be4cce70f1af9e885788856101699"><td class="memItemLeft" align="right" valign="top"><a id="ga3c6be4cce70f1af9e885788856101699"></a>
#define&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga3c6be4cce70f1af9e885788856101699">VRNA_UNSTRUCTURED_DOMAIN_ALL_LOOPS</a></td></tr>
<tr class="memdesc:ga3c6be4cce70f1af9e885788856101699"><td class="mdescLeft">&#160;</td><td class="mdescRight">Flag to indicate ligand bound to unpiared stretch in any loop (convenience macro) <br /></td></tr>
<tr class="separator:ga3c6be4cce70f1af9e885788856101699"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="typedef-members"></a>
Typedefs</h2></td></tr>
<tr class="memitem:ga0009117b14d29143e8b18ab891f48c2d"><td class="memItemLeft" align="right" valign="top"><a id="ga0009117b14d29143e8b18ab891f48c2d"></a>
typedef struct <a class="el" href="group__domains__up.html#structvrna__unstructured__domain__s">vrna_unstructured_domain_s</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga0009117b14d29143e8b18ab891f48c2d">vrna_ud_t</a></td></tr>
<tr class="memdesc:ga0009117b14d29143e8b18ab891f48c2d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Typename for the ligand binding extension data structure <a class="el" href="group__domains__up.html#structvrna__unstructured__domain__s" title="Data structure to store all functionality for ligand binding.">vrna_unstructured_domain_s</a>. <br /></td></tr>
<tr class="separator:ga0009117b14d29143e8b18ab891f48c2d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga75825c57d0bfde4ae4f95c044260c5c3"><td class="memItemLeft" align="right" valign="top">typedef int()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga75825c57d0bfde4ae4f95c044260c5c3">vrna_callback_ud_energy</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, void *data)</td></tr>
<tr class="memdesc:ga75825c57d0bfde4ae4f95c044260c5c3"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to retrieve binding free energy of a ligand bound to an unpaired sequence segment.  <a href="group__domains__up.html#ga75825c57d0bfde4ae4f95c044260c5c3">More...</a><br /></td></tr>
<tr class="separator:ga75825c57d0bfde4ae4f95c044260c5c3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga861706f257ba993753464b823e65b86e"><td class="memItemLeft" align="right" valign="top">typedef <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a>()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga861706f257ba993753464b823e65b86e">vrna_callback_ud_exp_energy</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, void *data)</td></tr>
<tr class="memdesc:ga861706f257ba993753464b823e65b86e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to retrieve Boltzmann factor of the binding free energy of a ligand bound to an unpaired sequence segment.  <a href="group__domains__up.html#ga861706f257ba993753464b823e65b86e">More...</a><br /></td></tr>
<tr class="separator:ga861706f257ba993753464b823e65b86e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga4fdfc02c1b660c07f2d887772f02a0a1"><td class="memItemLeft" align="right" valign="top">typedef void()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">vrna_callback_ud_production</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, void *data)</td></tr>
<tr class="memdesc:ga4fdfc02c1b660c07f2d887772f02a0a1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback for pre-processing the production rule of the ligand binding to unpaired stretches feature.  <a href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">More...</a><br /></td></tr>
<tr class="separator:ga4fdfc02c1b660c07f2d887772f02a0a1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga33d78327dcd04c1ca5ab2887edc18c7b"><td class="memItemLeft" align="right" valign="top">typedef void()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga33d78327dcd04c1ca5ab2887edc18c7b">vrna_callback_ud_exp_production</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, void *data)</td></tr>
<tr class="memdesc:ga33d78327dcd04c1ca5ab2887edc18c7b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback for pre-processing the production rule of the ligand binding to unpaired stretches feature (partition function variant)  <a href="group__domains__up.html#ga33d78327dcd04c1ca5ab2887edc18c7b">More...</a><br /></td></tr>
<tr class="separator:ga33d78327dcd04c1ca5ab2887edc18c7b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gab10498abc84fcaf336aca8f8d7d42eb2"><td class="memItemLeft" align="right" valign="top">typedef void()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gab10498abc84fcaf336aca8f8d7d42eb2">vrna_callback_ud_probs_add</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a> exp_energy, void *data)</td></tr>
<tr class="memdesc:gab10498abc84fcaf336aca8f8d7d42eb2"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to store/add equilibrium probability for a ligand bound to an unpaired sequence segment.  <a href="group__domains__up.html#gab10498abc84fcaf336aca8f8d7d42eb2">More...</a><br /></td></tr>
<tr class="separator:gab10498abc84fcaf336aca8f8d7d42eb2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gaa10ba1b6f1e179ea84c5caca9cdaae67"><td class="memItemLeft" align="right" valign="top">typedef <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a>()&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gaa10ba1b6f1e179ea84c5caca9cdaae67">vrna_callback_ud_probs_get</a>(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, int motif, void *data)</td></tr>
<tr class="memdesc:gaa10ba1b6f1e179ea84c5caca9cdaae67"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to retrieve equilibrium probability for a ligand bound to an unpaired sequence segment.  <a href="group__domains__up.html#gaa10ba1b6f1e179ea84c5caca9cdaae67">More...</a><br /></td></tr>
<tr class="separator:gaa10ba1b6f1e179ea84c5caca9cdaae67"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:ga2039caedf194c5edec794866986d95ec"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga2039caedf194c5edec794866986d95ec">vrna_ud_motifs_centroid</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const char *structure)</td></tr>
<tr class="memdesc:ga2039caedf194c5edec794866986d95ec"><td class="mdescLeft">&#160;</td><td class="mdescRight">Detect unstructured domains in centroid structure.  <a href="group__domains__up.html#ga2039caedf194c5edec794866986d95ec">More...</a><br /></td></tr>
<tr class="separator:ga2039caedf194c5edec794866986d95ec"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga980126e9f350b64474b35f20fce2782c"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga980126e9f350b64474b35f20fce2782c">vrna_ud_motifs_MEA</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const char *structure, <a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *probability_list)</td></tr>
<tr class="memdesc:ga980126e9f350b64474b35f20fce2782c"><td class="mdescLeft">&#160;</td><td class="mdescRight">Detect unstructured domains in MEA structure.  <a href="group__domains__up.html#ga980126e9f350b64474b35f20fce2782c">More...</a><br /></td></tr>
<tr class="separator:ga980126e9f350b64474b35f20fce2782c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga464d086264dd6f45089a65acec4e8c21"><td class="memItemLeft" align="right" valign="top"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga464d086264dd6f45089a65acec4e8c21">vrna_ud_motifs_MFE</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *fc, const char *structure)</td></tr>
<tr class="memdesc:ga464d086264dd6f45089a65acec4e8c21"><td class="mdescLeft">&#160;</td><td class="mdescRight">Detect unstructured domains in MFE structure.  <a href="group__domains__up.html#ga464d086264dd6f45089a65acec4e8c21">More...</a><br /></td></tr>
<tr class="separator:ga464d086264dd6f45089a65acec4e8c21"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga55f7de5ef5b7472b0eeab9296b57f671"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga55f7de5ef5b7472b0eeab9296b57f671">vrna_ud_add_motif</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, const char *motif, double motif_en, const char *motif_name, unsigned int loop_type)</td></tr>
<tr class="memdesc:ga55f7de5ef5b7472b0eeab9296b57f671"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add an unstructured domain motif, e.g. for ligand binding.  <a href="group__domains__up.html#ga55f7de5ef5b7472b0eeab9296b57f671">More...</a><br /></td></tr>
<tr class="separator:ga55f7de5ef5b7472b0eeab9296b57f671"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gada59cb0c498b812eadd010811af3f2d4"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gada59cb0c498b812eadd010811af3f2d4">vrna_ud_remove</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc)</td></tr>
<tr class="memdesc:gada59cb0c498b812eadd010811af3f2d4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove ligand binding to unpaired stretches.  <a href="group__domains__up.html#gada59cb0c498b812eadd010811af3f2d4">More...</a><br /></td></tr>
<tr class="separator:gada59cb0c498b812eadd010811af3f2d4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:gac1f18c312b91d80089534a87d956e58b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#gac1f18c312b91d80089534a87d956e58b">vrna_ud_set_data</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, void *data, <a class="el" href="group__fold__compound.html#ga7806651f51b195013839a218b3bbd5a3">vrna_callback_free_auxdata</a> *free_cb)</td></tr>
<tr class="memdesc:gac1f18c312b91d80089534a87d956e58b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Attach an auxiliary data structure.  <a href="group__domains__up.html#gac1f18c312b91d80089534a87d956e58b">More...</a><br /></td></tr>
<tr class="separator:gac1f18c312b91d80089534a87d956e58b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga745a99f0bc72898d54de16f6e538828a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga745a99f0bc72898d54de16f6e538828a">vrna_ud_set_prod_rule_cb</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, <a class="el" href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">vrna_callback_ud_production</a> *pre_cb, <a class="el" href="group__domains__up.html#ga75825c57d0bfde4ae4f95c044260c5c3">vrna_callback_ud_energy</a> *e_cb)</td></tr>
<tr class="memdesc:ga745a99f0bc72898d54de16f6e538828a"><td class="mdescLeft">&#160;</td><td class="mdescRight">Attach production rule callbacks for free energies computations.  <a href="group__domains__up.html#ga745a99f0bc72898d54de16f6e538828a">More...</a><br /></td></tr>
<tr class="separator:ga745a99f0bc72898d54de16f6e538828a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ga2fb1db2099da26c76247e1209ad4aa09"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ga2fb1db2099da26c76247e1209ad4aa09">vrna_ud_set_exp_prod_rule_cb</a> (<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, <a class="el" href="group__domains__up.html#ga33d78327dcd04c1ca5ab2887edc18c7b">vrna_callback_ud_exp_production</a> *pre_cb, <a class="el" href="group__domains__up.html#ga861706f257ba993753464b823e65b86e">vrna_callback_ud_exp_energy</a> *exp_e_cb)</td></tr>
<tr class="memdesc:ga2fb1db2099da26c76247e1209ad4aa09"><td class="mdescLeft">&#160;</td><td class="mdescRight">Attach production rule for partition function.  <a href="group__domains__up.html#ga2fb1db2099da26c76247e1209ad4aa09">More...</a><br /></td></tr>
<tr class="separator:ga2fb1db2099da26c76247e1209ad4aa09"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<hr/><h2 class="groupheader">Data Structure Documentation</h2>
<a name="structvrna__unstructured__domain__s" id="structvrna__unstructured__domain__s"></a>
<h2 class="memtitle"><span class="permalink"><a href="#structvrna__unstructured__domain__s">&#9670;&nbsp;</a></span>vrna_unstructured_domain_s</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">struct vrna_unstructured_domain_s</td>
        </tr>
      </table>
</div><div class="memdoc">
<div class="textblock"><p>Data structure to store all functionality for ligand binding. </p>
</div><table class="memberdecls">
<tr><td colspan="2"><h3>Data Fields</h3></td></tr>
<tr class="memitem:ae17005ef8043aca2fc3864804cd5def6"><td class="memItemLeft" align="right" valign="top"><a id="ae17005ef8043aca2fc3864804cd5def6"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ae17005ef8043aca2fc3864804cd5def6">uniq_motif_count</a></td></tr>
<tr class="memdesc:ae17005ef8043aca2fc3864804cd5def6"><td class="mdescLeft">&#160;</td><td class="mdescRight">The unique number of motifs of different lengths. <br /></td></tr>
<tr class="separator:ae17005ef8043aca2fc3864804cd5def6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2b484b0e19a47145db7055ada8b14159"><td class="memItemLeft" align="right" valign="top"><a id="a2b484b0e19a47145db7055ada8b14159"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a2b484b0e19a47145db7055ada8b14159">uniq_motif_size</a></td></tr>
<tr class="memdesc:a2b484b0e19a47145db7055ada8b14159"><td class="mdescLeft">&#160;</td><td class="mdescRight">An array storing a unique list of motif lengths. <br /></td></tr>
<tr class="separator:a2b484b0e19a47145db7055ada8b14159"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ad34148ea1d2e501f3e02029449546e"><td class="memItemLeft" align="right" valign="top"><a id="a5ad34148ea1d2e501f3e02029449546e"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a5ad34148ea1d2e501f3e02029449546e">motif_count</a></td></tr>
<tr class="memdesc:a5ad34148ea1d2e501f3e02029449546e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Total number of distinguished motifs. <br /></td></tr>
<tr class="separator:a5ad34148ea1d2e501f3e02029449546e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af285436bbdea4436ad2cedec65d48c75"><td class="memItemLeft" align="right" valign="top"><a id="af285436bbdea4436ad2cedec65d48c75"></a>
char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#af285436bbdea4436ad2cedec65d48c75">motif</a></td></tr>
<tr class="memdesc:af285436bbdea4436ad2cedec65d48c75"><td class="mdescLeft">&#160;</td><td class="mdescRight">Motif sequences. <br /></td></tr>
<tr class="separator:af285436bbdea4436ad2cedec65d48c75"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad061d55ef0b51d304c8dce88ebdbbf78"><td class="memItemLeft" align="right" valign="top"><a id="ad061d55ef0b51d304c8dce88ebdbbf78"></a>
char **&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ad061d55ef0b51d304c8dce88ebdbbf78">motif_name</a></td></tr>
<tr class="memdesc:ad061d55ef0b51d304c8dce88ebdbbf78"><td class="mdescLeft">&#160;</td><td class="mdescRight">Motif identifier/name. <br /></td></tr>
<tr class="separator:ad061d55ef0b51d304c8dce88ebdbbf78"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6a9f89ab7d147eb0ffd6cf8d77d178c0"><td class="memItemLeft" align="right" valign="top"><a id="a6a9f89ab7d147eb0ffd6cf8d77d178c0"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a6a9f89ab7d147eb0ffd6cf8d77d178c0">motif_size</a></td></tr>
<tr class="memdesc:a6a9f89ab7d147eb0ffd6cf8d77d178c0"><td class="mdescLeft">&#160;</td><td class="mdescRight">Motif lengths. <br /></td></tr>
<tr class="separator:a6a9f89ab7d147eb0ffd6cf8d77d178c0"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aae4f0a45419784fb7c244bce2781403d"><td class="memItemLeft" align="right" valign="top"><a id="aae4f0a45419784fb7c244bce2781403d"></a>
double *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#aae4f0a45419784fb7c244bce2781403d">motif_en</a></td></tr>
<tr class="memdesc:aae4f0a45419784fb7c244bce2781403d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Ligand binding free energy contribution. <br /></td></tr>
<tr class="separator:aae4f0a45419784fb7c244bce2781403d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3f0c026cced369ff176a66b2820f2057"><td class="memItemLeft" align="right" valign="top"><a id="a3f0c026cced369ff176a66b2820f2057"></a>
unsigned int *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a3f0c026cced369ff176a66b2820f2057">motif_type</a></td></tr>
<tr class="memdesc:a3f0c026cced369ff176a66b2820f2057"><td class="mdescLeft">&#160;</td><td class="mdescRight">Type of motif, i.e. loop type the ligand binds to. <br /></td></tr>
<tr class="separator:a3f0c026cced369ff176a66b2820f2057"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac2656d57130ac56e85212836482cbe80"><td class="memItemLeft" align="right" valign="top"><a class="el" href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">vrna_callback_ud_production</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ac2656d57130ac56e85212836482cbe80">prod_cb</a></td></tr>
<tr class="memdesc:ac2656d57130ac56e85212836482cbe80"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to ligand binding production rule, i.e. create/fill DP free energy matrices.  <a href="group__domains__up.html#ac2656d57130ac56e85212836482cbe80">More...</a><br /></td></tr>
<tr class="separator:ac2656d57130ac56e85212836482cbe80"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a181ae6ec67f77e68899ec6a0780f3e01"><td class="memItemLeft" align="right" valign="top"><a id="a181ae6ec67f77e68899ec6a0780f3e01"></a>
<a class="el" href="group__domains__up.html#ga33d78327dcd04c1ca5ab2887edc18c7b">vrna_callback_ud_exp_production</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a181ae6ec67f77e68899ec6a0780f3e01">exp_prod_cb</a></td></tr>
<tr class="memdesc:a181ae6ec67f77e68899ec6a0780f3e01"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to ligand binding production rule, i.e. create/fill DP partition function matrices. <br /></td></tr>
<tr class="separator:a181ae6ec67f77e68899ec6a0780f3e01"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1efba765dd4ea893dc8bd7362761f99b"><td class="memItemLeft" align="right" valign="top"><a id="a1efba765dd4ea893dc8bd7362761f99b"></a>
<a class="el" href="group__domains__up.html#ga75825c57d0bfde4ae4f95c044260c5c3">vrna_callback_ud_energy</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a1efba765dd4ea893dc8bd7362761f99b">energy_cb</a></td></tr>
<tr class="memdesc:a1efba765dd4ea893dc8bd7362761f99b"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to evaluate free energy of ligand binding to a particular unpaired stretch. <br /></td></tr>
<tr class="separator:a1efba765dd4ea893dc8bd7362761f99b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad5aaa2530557880ae89e297f8f09aa55"><td class="memItemLeft" align="right" valign="top"><a id="ad5aaa2530557880ae89e297f8f09aa55"></a>
<a class="el" href="group__domains__up.html#ga861706f257ba993753464b823e65b86e">vrna_callback_ud_exp_energy</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#ad5aaa2530557880ae89e297f8f09aa55">exp_energy_cb</a></td></tr>
<tr class="memdesc:ad5aaa2530557880ae89e297f8f09aa55"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to evaluate Boltzmann factor of ligand binding to a particular unpaired stretch. <br /></td></tr>
<tr class="separator:ad5aaa2530557880ae89e297f8f09aa55"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8802b1b0512999a9f35202031811ce17"><td class="memItemLeft" align="right" valign="top"><a id="a8802b1b0512999a9f35202031811ce17"></a>
void *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a8802b1b0512999a9f35202031811ce17">data</a></td></tr>
<tr class="memdesc:a8802b1b0512999a9f35202031811ce17"><td class="mdescLeft">&#160;</td><td class="mdescRight">Auxiliary data structure passed to energy evaluation callbacks. <br /></td></tr>
<tr class="separator:a8802b1b0512999a9f35202031811ce17"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a21b3084846902172858bc53f113d05a4"><td class="memItemLeft" align="right" valign="top"><a id="a21b3084846902172858bc53f113d05a4"></a>
<a class="el" href="group__fold__compound.html#ga7806651f51b195013839a218b3bbd5a3">vrna_callback_free_auxdata</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a21b3084846902172858bc53f113d05a4">free_data</a></td></tr>
<tr class="memdesc:a21b3084846902172858bc53f113d05a4"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to free auxiliary data structure. <br /></td></tr>
<tr class="separator:a21b3084846902172858bc53f113d05a4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a457b43dfab7f4321b5d7a84e5deea5d7"><td class="memItemLeft" align="right" valign="top"><a id="a457b43dfab7f4321b5d7a84e5deea5d7"></a>
<a class="el" href="group__domains__up.html#gab10498abc84fcaf336aca8f8d7d42eb2">vrna_callback_ud_probs_add</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a457b43dfab7f4321b5d7a84e5deea5d7">probs_add</a></td></tr>
<tr class="memdesc:a457b43dfab7f4321b5d7a84e5deea5d7"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to store/add outside partition function. <br /></td></tr>
<tr class="separator:a457b43dfab7f4321b5d7a84e5deea5d7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0b155283bccf65c5253f31e0211ae8ff"><td class="memItemLeft" align="right" valign="top"><a id="a0b155283bccf65c5253f31e0211ae8ff"></a>
<a class="el" href="group__domains__up.html#gaa10ba1b6f1e179ea84c5caca9cdaae67">vrna_callback_ud_probs_get</a> *&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="group__domains__up.html#a0b155283bccf65c5253f31e0211ae8ff">probs_get</a></td></tr>
<tr class="memdesc:a0b155283bccf65c5253f31e0211ae8ff"><td class="mdescLeft">&#160;</td><td class="mdescRight">Callback to retrieve outside partition function. <br /></td></tr>
<tr class="separator:a0b155283bccf65c5253f31e0211ae8ff"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h4 class="groupheader">Field Documentation</h4>
<a id="ac2656d57130ac56e85212836482cbe80"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ac2656d57130ac56e85212836482cbe80">&#9670;&nbsp;</a></span>prod_cb</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">vrna_callback_ud_production</a>* vrna_unstructured_domain_s::prod_cb</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Callback to ligand binding production rule, i.e. create/fill DP free energy matrices. </p>
<p>This callback will be executed right before the actual secondary structure decompositions, and, therefore, any implementation must not interleave with the regular DP matrices. </p>

</div>
</div>

</div>
</div>
<h2 class="groupheader">Typedef Documentation</h2>
<a id="ga75825c57d0bfde4ae4f95c044260c5c3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga75825c57d0bfde4ae4f95c044260c5c3">&#9670;&nbsp;</a></span>vrna_callback_ud_energy</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef int() vrna_callback_ud_energy(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback to retrieve binding free energy of a ligand bound to an unpaired sequence segment. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000008">Notes on Callback Functions:</a></b></dt><dd>This function will be called to determine the additional energy contribution of a specific unstructured domain, e.g. the binding free energy of some ligand. </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The current <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> </td></tr>
    <tr><td class="paramname">i</td><td>The start of the unstructured domain (5' end) </td></tr>
    <tr><td class="paramname">j</td><td>The end of the unstructured domain (3' end) </td></tr>
    <tr><td class="paramname">loop_type</td><td>The loop context of the unstructured domain </td></tr>
    <tr><td class="paramname">data</td><td>Auxiliary data </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The auxiliary energy contribution in deka-cal/mol </dd></dl>

</div>
</div>
<a id="ga861706f257ba993753464b823e65b86e"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga861706f257ba993753464b823e65b86e">&#9670;&nbsp;</a></span>vrna_callback_ud_exp_energy</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a>() vrna_callback_ud_exp_energy(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback to retrieve Boltzmann factor of the binding free energy of a ligand bound to an unpaired sequence segment. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000009">Notes on Callback Functions:</a></b></dt><dd>This function will be called to determine the additional energy contribution of a specific unstructured domain, e.g. the binding free energy of some ligand (Partition function variant, i.e. the Boltzmann factors instead of actual free energies). </dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The current <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> </td></tr>
    <tr><td class="paramname">i</td><td>The start of the unstructured domain (5' end) </td></tr>
    <tr><td class="paramname">j</td><td>The end of the unstructured domain (3' end) </td></tr>
    <tr><td class="paramname">loop_type</td><td>The loop context of the unstructured domain </td></tr>
    <tr><td class="paramname">data</td><td>Auxiliary data </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>The auxiliary energy contribution as Boltzmann factor </dd></dl>

</div>
</div>
<a id="ga4fdfc02c1b660c07f2d887772f02a0a1"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga4fdfc02c1b660c07f2d887772f02a0a1">&#9670;&nbsp;</a></span>vrna_callback_ud_production</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef void() vrna_callback_ud_production(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback for pre-processing the production rule of the ligand binding to unpaired stretches feature. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000010">Notes on Callback Functions:</a></b></dt><dd>The production rule for the unstructured domain grammar extension </dd></dl>

</div>
</div>
<a id="ga33d78327dcd04c1ca5ab2887edc18c7b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga33d78327dcd04c1ca5ab2887edc18c7b">&#9670;&nbsp;</a></span>vrna_callback_ud_exp_production</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef void() vrna_callback_ud_exp_production(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback for pre-processing the production rule of the ligand binding to unpaired stretches feature (partition function variant) </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000011">Notes on Callback Functions:</a></b></dt><dd>The production rule for the unstructured domain grammar extension (Partition function variant) </dd></dl>

</div>
</div>
<a id="gab10498abc84fcaf336aca8f8d7d42eb2"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gab10498abc84fcaf336aca8f8d7d42eb2">&#9670;&nbsp;</a></span>vrna_callback_ud_probs_add</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef void() vrna_callback_ud_probs_add(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a> exp_energy, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback to store/add equilibrium probability for a ligand bound to an unpaired sequence segment. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000012">Notes on Callback Functions:</a></b></dt><dd>A callback function to store equilibrium probabilities for the unstructured domain feature </dd></dl>

</div>
</div>
<a id="gaa10ba1b6f1e179ea84c5caca9cdaae67"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gaa10ba1b6f1e179ea84c5caca9cdaae67">&#9670;&nbsp;</a></span>vrna_callback_ud_probs_get</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">typedef <a class="el" href="group__data__structures.html#ga31125aeace516926bf7f251f759b6126">FLT_OR_DBL</a>() vrna_callback_ud_probs_get(<a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *vc, int i, int j, unsigned int loop_type, int motif, void *data)</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Callback to retrieve equilibrium probability for a ligand bound to an unpaired sequence segment. </p>
<dl class="callbacks"><dt><b><a class="el" href="callbacks.html#_callbacks000013">Notes on Callback Functions:</a></b></dt><dd>A callback function to retrieve equilibrium probabilities for the unstructured domain feature </dd></dl>

</div>
</div>
<h2 class="groupheader">Function Documentation</h2>
<a id="ga2039caedf194c5edec794866986d95ec"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2039caedf194c5edec794866986d95ec">&#9670;&nbsp;</a></span>vrna_ud_motifs_centroid()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a>* vrna_ud_motifs_centroid </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Detect unstructured domains in centroid structure. </p>
<p>Given a centroid structure and a set of unstructured domains compute the list of unstructured domain motifs present in the centroid. Since we do not explicitly annotate unstructured domain motifs in dot-bracket strings, this function can be used to check for the presence and location of unstructured domain motifs under the assumption that the dot-bracket string is the centroid structure of the equiibrium ensemble.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__centroid__fold.html#ga0e64bb67e51963dc71cbd4d30b80a018" title="Get the centroid structure of the ensemble.">vrna_centroid()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The fold_compound data structure with pre-computed equilibrium probabilities and model settings </td></tr>
    <tr><td class="paramname">structure</td><td>The centroid structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of unstructured domain motifs (possibly NULL). The last element terminates the list with <code>start=0</code>, <code>number=-1</code> </dd></dl>

</div>
</div>
<a id="ga980126e9f350b64474b35f20fce2782c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga980126e9f350b64474b35f20fce2782c">&#9670;&nbsp;</a></span>vrna_ud_motifs_MEA()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a>* vrna_ud_motifs_MEA </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__struct__utils__plist.html#gab9ac98ab55ded9fb90043b024b915aca">vrna_ep_t</a> *&#160;</td>
          <td class="paramname"><em>probability_list</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Detect unstructured domains in MEA structure. </p>
<p>Given an MEA structure and a set of unstructured domains compute the list of unstructured domain motifs present in the MEA structure. Since we do not explicitly annotate unstructured domain motifs in dot-bracket strings, this function can be used to check for the presence and location of unstructured domain motifs under the assumption that the dot-bracket string is the MEA structure of the equiibrium ensemble.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__mea__fold.html#ga396ec6144c6a74fcbab4cea6b42d76c3" title="Computes a MEA (maximum expected accuracy) structure.">MEA()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The fold_compound data structure with pre-computed equilibrium probabilities and model settings </td></tr>
    <tr><td class="paramname">structure</td><td>The MEA structure in dot-bracket notation </td></tr>
    <tr><td class="paramname">probability_list</td><td>The list of probabilities to extract the MEA structure from </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of unstructured domain motifs (possibly NULL). The last element terminates the list with <code>start=0</code>, <code>number=-1</code> </dd></dl>

</div>
</div>
<a id="ga464d086264dd6f45089a65acec4e8c21"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga464d086264dd6f45089a65acec4e8c21">&#9670;&nbsp;</a></span>vrna_ud_motifs_MFE()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="structvrna__unstructured__domain__motif__s.html">vrna_ud_motif_t</a>* vrna_ud_motifs_MFE </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>fc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>structure</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Detect unstructured domains in MFE structure. </p>
<p>Given an MFE structure and a set of unstructured domains compute the list of unstructured domain motifs present in the MFE structure. Since we do not explicitly annotate unstructured domain motifs in dot-bracket strings, this function can be used to check for the presence and location of unstructured domain motifs under the assumption that the dot-bracket string is the MFE structure of the equiibrium ensemble.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__mfe__global.html#gabd3b147371ccf25c577f88bbbaf159fd" title="Compute minimum free energy and an appropriate secondary structure of an RNA sequence,...">vrna_mfe()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">fc</td><td>The fold_compound data structure with model settings </td></tr>
    <tr><td class="paramname">structure</td><td>The MFE structure in dot-bracket notation </td></tr>
  </table>
  </dd>
</dl>
<dl class="section return"><dt>Returns</dt><dd>A list of unstructured domain motifs (possibly NULL). The last element terminates the list with <code>start=0</code>, <code>number=-1</code> </dd></dl>

</div>
</div>
<a id="ga55f7de5ef5b7472b0eeab9296b57f671"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga55f7de5ef5b7472b0eeab9296b57f671">&#9670;&nbsp;</a></span>vrna_ud_add_motif()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_ud_add_motif </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>motif</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">double&#160;</td>
          <td class="paramname"><em>motif_en</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const char *&#160;</td>
          <td class="paramname"><em>motif_name</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">unsigned int&#160;</td>
          <td class="paramname"><em>loop_type</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Add an unstructured domain motif, e.g. for ligand binding. </p>
<p>This function adds a ligand binding motif and the associated binding free energy to the <a class="el" href="group__domains__up.html#ga0009117b14d29143e8b18ab891f48c2d" title="Typename for the ligand binding extension data structure vrna_unstructured_domain_s.">vrna_ud_t</a> attribute of a <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a>. The motif data will then be used in subsequent secondary structure predictions. Multiple calls to this function with different motifs append all additional data to a list of ligands, which all will be evaluated. Ligand motif data can be removed from the <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> again using the <a class="el" href="group__domains__up.html#gada59cb0c498b812eadd010811af3f2d4" title="Remove ligand binding to unpaired stretches.">vrna_ud_remove()</a> function. The loop type parameter allows one to limit the ligand binding to particular loop type, such as the exterior loop, hairpin loops, interior loops, or multibranch loops.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__domains__up.html#gaac911374e86236a51bfd42e1f098eaba" title="Flag to indicate ligand bound to unpiared stretch in the exterior loop.">VRNA_UNSTRUCTURED_DOMAIN_EXT_LOOP</a>, <a class="el" href="group__domains__up.html#ga23b610ea9564346c45cc1e2bbb62adf7" title="Flag to indicate ligand bound to unpaired stretch in a hairpin loop.">VRNA_UNSTRUCTURED_DOMAIN_HP_LOOP</a>, <a class="el" href="group__domains__up.html#gac4a0feccd9654c149203200248c2716b" title="Flag to indicate ligand bound to unpiared stretch in an interior loop.">VRNA_UNSTRUCTURED_DOMAIN_INT_LOOP</a>, <a class="el" href="group__domains__up.html#ga67b80796655e5227a4ed662bfbe398b0" title="Flag to indicate ligand bound to unpiared stretch in a multibranch loop.">VRNA_UNSTRUCTURED_DOMAIN_MB_LOOP</a>, <a class="el" href="group__domains__up.html#ga3c6be4cce70f1af9e885788856101699" title="Flag to indicate ligand bound to unpiared stretch in any loop (convenience macro)">VRNA_UNSTRUCTURED_DOMAIN_ALL_LOOPS</a>, <a class="el" href="group__domains__up.html#gada59cb0c498b812eadd010811af3f2d4" title="Remove ligand binding to unpaired stretches.">vrna_ud_remove()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure the ligand motif should be bound to </td></tr>
    <tr><td class="paramname">motif</td><td>The sequence motif the ligand binds to </td></tr>
    <tr><td class="paramname">motif_en</td><td>The binding free energy of the ligand in kcal/mol </td></tr>
    <tr><td class="paramname">motif_name</td><td>The name/id of the motif (may be <code>NULL</code>) </td></tr>
    <tr><td class="paramname">loop_type</td><td>The loop type the ligand binds to </td></tr>
  </table>
  </dd>
</dl>

</div>
</div>
<a id="gada59cb0c498b812eadd010811af3f2d4"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gada59cb0c498b812eadd010811af3f2d4">&#9670;&nbsp;</a></span>vrna_ud_remove()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_ud_remove </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Remove ligand binding to unpaired stretches. </p>
<p>This function removes all ligand motifs that were bound to a <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> using the <a class="el" href="group__domains__up.html#ga55f7de5ef5b7472b0eeab9296b57f671" title="Add an unstructured domain motif, e.g. for ligand binding.">vrna_ud_add_motif()</a> function.</p>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure the ligand motif data should be removed from</td></tr>
  </table>
  </dd>
</dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000083">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>ud_remove()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="gac1f18c312b91d80089534a87d956e58b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#gac1f18c312b91d80089534a87d956e58b">&#9670;&nbsp;</a></span>vrna_ud_set_data()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_ud_set_data </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">void *&#160;</td>
          <td class="paramname"><em>data</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga7806651f51b195013839a218b3bbd5a3">vrna_callback_free_auxdata</a> *&#160;</td>
          <td class="paramname"><em>free_cb</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Attach an auxiliary data structure. </p>
<p>This function binds an arbitrary, auxiliary data structure for user-implemented ligand binding. The optional callback <code>free_cb</code> will be passed the bound data structure whenever the <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> is removed from memory to avoid memory leaks.</p>
<dl class="section see"><dt>See also</dt><dd><a class="el" href="group__domains__up.html#ga745a99f0bc72898d54de16f6e538828a" title="Attach production rule callbacks for free energies computations.">vrna_ud_set_prod_rule_cb()</a>, <a class="el" href="group__domains__up.html#ga2fb1db2099da26c76247e1209ad4aa09" title="Attach production rule for partition function.">vrna_ud_set_exp_prod_rule_cb()</a>, <a class="el" href="group__domains__up.html#gada59cb0c498b812eadd010811af3f2d4" title="Remove ligand binding to unpaired stretches.">vrna_ud_remove()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure the auxiliary data structure should be bound to </td></tr>
    <tr><td class="paramname">data</td><td>A pointer to the auxiliary data structure </td></tr>
    <tr><td class="paramname">free_cb</td><td>A pointer to a callback function that free's memory occupied by <code>data</code> </td></tr>
  </table>
  </dd>
</dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000084">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>ud_set_data()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga745a99f0bc72898d54de16f6e538828a"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga745a99f0bc72898d54de16f6e538828a">&#9670;&nbsp;</a></span>vrna_ud_set_prod_rule_cb()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_ud_set_prod_rule_cb </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__domains__up.html#ga4fdfc02c1b660c07f2d887772f02a0a1">vrna_callback_ud_production</a> *&#160;</td>
          <td class="paramname"><em>pre_cb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__domains__up.html#ga75825c57d0bfde4ae4f95c044260c5c3">vrna_callback_ud_energy</a> *&#160;</td>
          <td class="paramname"><em>e_cb</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Attach production rule callbacks for free energies computations. </p>
<p>Use this function to bind a user-implemented grammar extension for unstructured domains.</p>
<p>The callback <code>e_cb</code> needs to evaluate the free energy contribution <img class="formulaInl" alt="$f(i,j)$" src="form_102.png" width="34" height="15"/> of the unpaired segment <img class="formulaInl" alt="$[i,j]$" src="form_103.png" width="23" height="15"/>. It will be executed in each of the regular secondary structure production rules. Whenever the callback is passed the <a class="el" href="group__domains__up.html#gaab12b58d59be76446a9f76fad2fe624c" title="Flag to indicate ligand binding without additional unbound nucleotides (motif-only)">VRNA_UNSTRUCTURED_DOMAIN_MOTIF</a> flag via its <code>loop_type</code> parameter the contribution of any ligand that consecutively binds from position <img class="formulaInl" alt="$i$" src="form_59.png" width="6" height="10"/> to <img class="formulaInl" alt="$j$" src="form_104.png" width="7" height="12"/> (the white box) is requested. Otherwise, the callback usually performs a lookup in the precomputed <code>B</code> matrices. Which <code>B</code> matrix is addressed will be indicated by the flags <a class="el" href="group__domains__up.html#gaac911374e86236a51bfd42e1f098eaba" title="Flag to indicate ligand bound to unpiared stretch in the exterior loop.">VRNA_UNSTRUCTURED_DOMAIN_EXT_LOOP</a>, <a class="el" href="group__domains__up.html#ga23b610ea9564346c45cc1e2bbb62adf7" title="Flag to indicate ligand bound to unpaired stretch in a hairpin loop.">VRNA_UNSTRUCTURED_DOMAIN_HP_LOOP</a> <a class="el" href="group__domains__up.html#gac4a0feccd9654c149203200248c2716b" title="Flag to indicate ligand bound to unpiared stretch in an interior loop.">VRNA_UNSTRUCTURED_DOMAIN_INT_LOOP</a>, and <a class="el" href="group__domains__up.html#ga67b80796655e5227a4ed662bfbe398b0" title="Flag to indicate ligand bound to unpiared stretch in a multibranch loop.">VRNA_UNSTRUCTURED_DOMAIN_MB_LOOP</a>. As their names already imply, they specify exterior loops (<code>F</code> production rule), hairpin loops and interior loops (<code>C</code> production rule), and multibranch loops (<code>M</code> and <code>M1</code> production rule).</p>
<div class="image">
<object type="image/svg+xml" data="ligands_up_callback.svg" style="pointer-events: none;"></object>
</div>
 <p>The <code>pre_cb</code> callback will be executed as a pre-processing step right before the regular secondary structure rules. Usually one would use this callback to fill the dynamic programming matrices <code>U</code> and preparations of the auxiliary data structure <a class="el" href="group__domains__up.html#a8802b1b0512999a9f35202031811ce17" title="Auxiliary data structure passed to energy evaluation callbacks.">vrna_unstructured_domain_s.data</a></p>
<div class="image">
<object type="image/svg+xml" data="B_prod_rule.svg" style="pointer-events: none;"></object>
</div>
 <dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure the callback will be bound to </td></tr>
    <tr><td class="paramname">pre_cb</td><td>A pointer to a callback function for the <code>B</code> production rule </td></tr>
    <tr><td class="paramname">e_cb</td><td>A pointer to a callback function for free energy evaluation</td></tr>
  </table>
  </dd>
</dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000085">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>ud_set_prod_rule_cb()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
<a id="ga2fb1db2099da26c76247e1209ad4aa09"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ga2fb1db2099da26c76247e1209ad4aa09">&#9670;&nbsp;</a></span>vrna_ud_set_exp_prod_rule_cb()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vrna_ud_set_exp_prod_rule_cb </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166">vrna_fold_compound_t</a> *&#160;</td>
          <td class="paramname"><em>vc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__domains__up.html#ga33d78327dcd04c1ca5ab2887edc18c7b">vrna_callback_ud_exp_production</a> *&#160;</td>
          <td class="paramname"><em>pre_cb</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="group__domains__up.html#ga861706f257ba993753464b823e65b86e">vrna_callback_ud_exp_energy</a> *&#160;</td>
          <td class="paramname"><em>exp_e_cb</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><code>#include &lt;<a class="el" href="unstructured__domains_8h.html">ViennaRNA/unstructured_domains.h</a>&gt;</code></p>

<p>Attach production rule for partition function. </p>
<p>This function is the partition function companion of <a class="el" href="group__domains__up.html#ga745a99f0bc72898d54de16f6e538828a" title="Attach production rule callbacks for free energies computations.">vrna_ud_set_prod_rule_cb()</a>.</p>
<p>Use it to bind callbacks to (i) fill the <code>U</code> production rule dynamic programming matrices and/or prepare the <a class="el" href="group__domains__up.html#a8802b1b0512999a9f35202031811ce17" title="Auxiliary data structure passed to energy evaluation callbacks.">vrna_unstructured_domain_s.data</a>, and (ii) provide a callback to retrieve partition functions for subsegments <img class="formulaInl" alt="$ [i,j] $" src="form_105.png" width="23" height="15"/>.</p>
<div class="image">
<object type="image/svg+xml" data="B_prod_rule.svg" style="pointer-events: none;"></object>
</div>
 <div class="image">
<object type="image/svg+xml" data="ligands_up_callback.svg" style="pointer-events: none;"></object>
</div>
 <dl class="section see"><dt>See also</dt><dd><a class="el" href="group__domains__up.html#ga745a99f0bc72898d54de16f6e538828a" title="Attach production rule callbacks for free energies computations.">vrna_ud_set_prod_rule_cb()</a></dd></dl>
<dl class="params"><dt>Parameters</dt><dd>
  <table class="params">
    <tr><td class="paramname">vc</td><td>The <a class="el" href="group__fold__compound.html#ga1b0cef17fd40466cef5968eaeeff6166" title="Typename for the fold_compound data structure vrna_fc_s.">vrna_fold_compound_t</a> data structure the callback will be bound to </td></tr>
    <tr><td class="paramname">pre_cb</td><td>A pointer to a callback function for the <code>B</code> production rule </td></tr>
    <tr><td class="paramname">exp_e_cb</td><td>A pointer to a callback function that retrieves the partition function for a segment <img class="formulaInl" alt="$[i,j]$" src="form_103.png" width="23" height="15"/> that may be bound by one or more ligands.</td></tr>
  </table>
  </dd>
</dl>
<dl class="wrappers"><dt><b><a class="el" href="wrappers.html#_wrappers000086">SWIG Wrapper Notes:</a></b></dt><dd>This function is attached as method <b>ud_set_exp_prod_rule_cb()</b> to objects of type <em>fold_compound</em> </dd></dl>

</div>
</div>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:31 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
