<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>RNAlib-2.4.18: The RNAlib API</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtreedata.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
  $(document).ready(function() {
    if ($('.searchresults').length > 0) { searchBox.DOMSearchField().focus(); }
  });
  /* @license-end */
</script>
<link rel="search" href="search_opensearch.php?v=opensearch.xml" type="application/opensearchdescription+xml" title="RNAlib-2.4.18"/>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">RNAlib-2.4.18
   </div>
  </td>
   <td>        <div id="MSearchBox" class="MSearchBoxInactive">
          <div class="left">
            <form id="FSearchBox" action="search.html" method="get">
              <img id="MSearchSelect" src="search/mag.svg" alt=""/>
              <input type="text" id="MSearchField" name="query" value="Search" size="20" accesskey="S" 
                     onfocus="searchBox.OnSearchFieldFocus(true)" 
                     onblur="searchBox.OnSearchFieldFocus(false)"/>
            </form>
          </div><div class="right"></div>
        </div>
</td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search','.html');
/* @license-end */
</script>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(document).ready(function(){initNavTree('modules.html',''); initResizable(); });
/* @license-end */
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">The RNAlib API</div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock">Our library is grouped into several modules, each addressing different aspects of RNA secondary structure related problems. You can find an overview of the different groups below.</div><div class="directory">
<div class="levels">[detail level <span onclick="javascript:toggleLevel(1);">1</span><span onclick="javascript:toggleLevel(2);">2</span><span onclick="javascript:toggleLevel(3);">3</span><span onclick="javascript:toggleLevel(4);">4</span>]</div><table class="directory">
<tr id="row_0_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_0_" class="arrow" onclick="toggleFolder('0_')">&#9658;</span><a class="el" href="group__eval.html" target="_self">Free Energy Evaluation</a></td><td class="desc">Functions and variables related to free energy evaluation of sequence/structure pairs </td></tr>
<tr id="row_0_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_0_0_" class="arrow" onclick="toggleFolder('0_0_')">&#9658;</span><a class="el" href="group__eval__loops.html" target="_self">Energy Evaluation for Individual Loops</a></td><td class="desc">Functions to evaluate the free energy of particular types of loops </td></tr>
<tr id="row_0_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__eval__loops__ext.html" target="_self">Exterior Loops</a></td><td class="desc">Functions to evaluate the free energy contributions for exterior loops </td></tr>
<tr id="row_0_0_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__eval__loops__hp.html" target="_self">Hairpin Loops</a></td><td class="desc">Functions to evaluate the free energy contributions for hairpin loops </td></tr>
<tr id="row_0_0_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__eval__loops__int.html" target="_self">Internal Loops</a></td><td class="desc">Functions to evaluate the free energy contributions for internal loops </td></tr>
<tr id="row_0_0_3_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__eval__loops__mb.html" target="_self">Multibranch Loops</a></td><td class="desc">Functions to evaluate the free energy contributions for mutlibranch loops </td></tr>
<tr id="row_0_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__eval__move.html" target="_self">Energy Evaluation for Atomic Moves</a></td><td class="desc">Functions to evaluate the free energy change of a structure after application of (a set of) atomic moves </td></tr>
<tr id="row_0_2_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__eval__deprecated.html" target="_self">Deprecated Interface for Free Energy Evaluation</a></td><td class="desc">Deprecated Energy Evaluation functions </td></tr>
<tr id="row_1_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_1_" class="arrow" onclick="toggleFolder('1_')">&#9658;</span><a class="el" href="group__grammar.html" target="_self">The RNA Folding Grammar</a></td><td class="desc">The RNA folding grammar as implemented in RNAlib </td></tr>
<tr id="row_1_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__model__details.html" target="_self">Fine-tuning of the Implemented Models</a></td><td class="desc">Functions and data structures to fine-tune the implemented secondary structure evaluation model </td></tr>
<tr id="row_1_1_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_1_1_" class="arrow" onclick="toggleFolder('1_1_')">&#9658;</span><a class="el" href="group__energy__parameters.html" target="_self">Energy Parameters</a></td><td class="desc">All relevant functions to retrieve and copy pre-calculated energy parameter sets as well as reading/writing the energy parameter set from/to file(s) </td></tr>
<tr id="row_1_1_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><span id="arr_1_1_0_" class="arrow" onclick="toggleFolder('1_1_0_')">&#9658;</span><a class="el" href="group__energy__parameters__rw.html" target="_self">Reading/Writing Energy Parameter Sets from/to File</a></td><td class="desc">Read and Write energy parameter sets from and to files or strings </td></tr>
<tr id="row_1_1_0_0_" class="even" style="display:none;"><td class="entry"><span style="width:64px;display:inline-block;">&#160;</span><a class="el" href="group__energy__parameters__convert.html" target="_self">Converting Energy Parameter Files</a></td><td class="desc">Convert energy parameter files into the latest format </td></tr>
<tr id="row_1_2_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_1_2_" class="arrow" onclick="toggleFolder('1_2_')">&#9658;</span><a class="el" href="group__domains.html" target="_self">Extending the Folding Grammar with Additional Domains</a></td><td class="desc">This module covers simple and straight-forward extensions to the RNA folding grammar </td></tr>
<tr id="row_1_2_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__domains__up.html" target="_self">Unstructured Domains</a></td><td class="desc">Add and modify unstructured domains to the RNA folding grammar </td></tr>
<tr id="row_1_2_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__domains__struc.html" target="_self">Structured Domains</a></td><td class="desc">Add and modify structured domains to the RNA folding grammar </td></tr>
<tr id="row_1_3_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_1_3_" class="arrow" onclick="toggleFolder('1_3_')">&#9658;</span><a class="el" href="group__constraints.html" target="_self">Constraining the RNA Folding Grammar</a></td><td class="desc">This module provides general functions that allow for an easy control of constrained secondary structure prediction and evaluation </td></tr>
<tr id="row_1_3_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__hard__constraints.html" target="_self">Hard Constraints</a></td><td class="desc">This module covers all functionality for hard constraints in secondary structure prediction </td></tr>
<tr id="row_1_3_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__soft__constraints.html" target="_self">Soft Constraints</a></td><td class="desc">Functions and data structures for secondary structure soft constraints </td></tr>
<tr id="row_2_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_2_" class="arrow" onclick="toggleFolder('2_')">&#9658;</span><a class="el" href="group__landscape.html" target="_self">The RNA Secondary Structure Landscape</a></td><td class="desc"></td></tr>
<tr id="row_2_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__neighbors.html" target="_self">Neighborhood Relation and Move Sets for Secondary Structures</a></td><td class="desc">Different functions to generate structural neighbors of a secondary structure according to a particular Move Set </td></tr>
<tr id="row_2_1_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_2_1_" class="arrow" onclick="toggleFolder('2_1_')">&#9658;</span><a class="el" href="group__paths.html" target="_self">(Re-)folding Paths, Saddle Points, Energy Barriers, and Local Minima</a></td><td class="desc">API for various RNA folding path algorithms </td></tr>
<tr id="row_2_1_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__paths__direct.html" target="_self">Direct Refolding Paths between two Secondary Structures</a></td><td class="desc">Heuristics to explore direct, optimal (re-)folding paths between two secondary structures </td></tr>
<tr id="row_2_1_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__paths__walk.html" target="_self">Folding Paths that start at a single Secondary Structure</a></td><td class="desc">Implementation of gradient- and random walks starting from a single secondary structure </td></tr>
<tr id="row_2_1_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__paths__deprecated.html" target="_self">Deprecated Interface for (Re-)folding Paths, Saddle Points, and Energy Barriers</a></td><td class="desc"></td></tr>
<tr id="row_3_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_3_" class="arrow" onclick="toggleFolder('3_')">&#9658;</span><a class="el" href="group__mfe.html" target="_self">Minimum Free Energy (MFE) Algorithms</a></td><td class="desc">Predicting the Minimum Free Energy (MFE) and a corresponding (consensus) secondary structure </td></tr>
<tr id="row_3_0_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_3_0_" class="arrow" onclick="toggleFolder('3_0_')">&#9658;</span><a class="el" href="group__mfe__global.html" target="_self">Global MFE Prediction</a></td><td class="desc">Variations of the global Minimum Free Energy (MFE) prediction algorithm </td></tr>
<tr id="row_3_0_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__mfe.html" target="_self">Computing MFE representatives of a Distance Based Partitioning</a></td><td class="desc">Compute the minimum free energy (MFE) and secondary structures for a partitioning of the secondary structure space according to the base pair distance to two fixed reference structures basepair distance to two fixed reference structures </td></tr>
<tr id="row_3_0_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__mfe__global__deprecated.html" target="_self">Deprecated Interface for Global MFE Prediction</a></td><td class="desc"></td></tr>
<tr id="row_3_1_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_3_1_" class="arrow" onclick="toggleFolder('3_1_')">&#9658;</span><a class="el" href="group__mfe__window.html" target="_self">Local (sliding window) MFE Prediction</a></td><td class="desc">Variations of the local (sliding window) Minimum Free Energy (MFE) prediction algorithm </td></tr>
<tr id="row_3_1_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__mfe__window__deprecated.html" target="_self">Deprecated Interface for Local (Sliding Window) MFE Prediction</a></td><td class="desc"></td></tr>
<tr id="row_3_2_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__mfe__backtracking.html" target="_self">Backtracking MFE structures</a></td><td class="desc">Backtracking related interfaces </td></tr>
<tr id="row_4_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_4_" class="arrow" onclick="toggleFolder('4_')">&#9658;</span><a class="el" href="group__pf__fold.html" target="_self">Partition Function and Equilibrium Properties</a></td><td class="desc">Compute the partition function to assess various equilibrium properties </td></tr>
<tr id="row_4_0_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_4_0_" class="arrow" onclick="toggleFolder('4_0_')">&#9658;</span><a class="el" href="group__part__func__global.html" target="_self">Global Partition Function and Equilibrium Probabilities</a></td><td class="desc">Variations of the global partition function algorithm </td></tr>
<tr id="row_4_0_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__pf.html" target="_self">Computing Partition Functions of a Distance Based Partitioning</a></td><td class="desc">Compute the partition function and stochastically sample secondary structures for a partitioning of the secondary structure space according to the base pair distance to two fixed reference structures </td></tr>
<tr id="row_4_0_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__thermodynamics.html" target="_self">Predicting various thermodynamic properties</a></td><td class="desc">Compute various thermodynamic properties using the partition function </td></tr>
<tr id="row_4_0_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__part__func__global__deprecated.html" target="_self">Deprecated Interface for Global Partition Function Computation</a></td><td class="desc"></td></tr>
<tr id="row_4_1_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_4_1_" class="arrow" onclick="toggleFolder('4_1_')">&#9658;</span><a class="el" href="group__part__func__window.html" target="_self">Local (sliding window) Partition Function and Equilibrium Probabilities</a></td><td class="desc">Scanning version using a sliding window approach to compute equilibrium probabilities </td></tr>
<tr id="row_4_1_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__part__func__window__deprecated.html" target="_self">Deprecated Interface for Local (Sliding Window) Partition Function Computation</a></td><td class="desc"></td></tr>
<tr id="row_5_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_5_" class="arrow" onclick="toggleFolder('5_')">&#9658;</span><a class="el" href="group__subopt__and__representatives.html" target="_self">Suboptimals and Representative Structures</a></td><td class="desc">Sample and enumerate suboptimal secondary structures from RNA sequence data </td></tr>
<tr id="row_5_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__subopt__zuker.html" target="_self">Suboptimal Structures sensu Stiegler et al. 1984 / Zuker et al. 1989</a></td><td class="desc"></td></tr>
<tr id="row_5_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__subopt__wuchty.html" target="_self">Suboptimal Structures within an Energy Band around the MFE</a></td><td class="desc"></td></tr>
<tr id="row_5_2_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_5_2_" class="arrow" onclick="toggleFolder('5_2_')">&#9658;</span><a class="el" href="group__subopt__stochbt.html" target="_self">Random Structure Samples from the Ensemble</a></td><td class="desc">Functions to draw random structure samples from the ensemble according to their equilibrium probability </td></tr>
<tr id="row_5_2_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__stochbt.html" target="_self">Stochastic Backtracking of Structures from Distance Based Partitioning</a></td><td class="desc">Contains functions related to stochastic backtracking from a specified distance class </td></tr>
<tr id="row_5_2_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__subopt__stochbt__deprecated.html" target="_self">Deprecated Interface for Stochastic Backtracking</a></td><td class="desc"></td></tr>
<tr id="row_5_3_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__mea__fold.html" target="_self">Compute the Structure with Maximum Expected Accuracy (MEA)</a></td><td class="desc"></td></tr>
<tr id="row_5_4_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__centroid__fold.html" target="_self">Compute the Centroid Structure</a></td><td class="desc"></td></tr>
<tr id="row_6_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_6_" class="arrow" onclick="toggleFolder('6_')">&#9658;</span><a class="el" href="group__cofold.html" target="_self">RNA-RNA Interaction</a></td><td class="desc"></td></tr>
<tr id="row_6_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__pf__cofold.html" target="_self">Partition Function for Two Hybridized Sequences</a></td><td class="desc">Partition Function Cofolding </td></tr>
<tr id="row_6_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__up__cofold.html" target="_self">Partition Function for two Hybridized Sequences as a Stepwise Process</a></td><td class="desc">RNA-RNA interaction as a stepwise process </td></tr>
<tr id="row_7_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_7_" class="arrow" onclick="toggleFolder('7_')">&#9658;</span><a class="el" href="group__class__fold.html" target="_self">Classified Dynamic Programming Variants</a></td><td class="desc"></td></tr>
<tr id="row_7_0_" class="even" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_7_0_" class="arrow" onclick="toggleFolder('7_0_')">&#9658;</span><a class="el" href="group__kl__neighborhood.html" target="_self">Distance Based Partitioning of the Secondary Structure Space</a></td><td class="desc"></td></tr>
<tr id="row_7_0_0_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__mfe.html" target="_self">Computing MFE representatives of a Distance Based Partitioning</a></td><td class="desc">Compute the minimum free energy (MFE) and secondary structures for a partitioning of the secondary structure space according to the base pair distance to two fixed reference structures basepair distance to two fixed reference structures </td></tr>
<tr id="row_7_0_1_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__pf.html" target="_self">Computing Partition Functions of a Distance Based Partitioning</a></td><td class="desc">Compute the partition function and stochastically sample secondary structures for a partitioning of the secondary structure space according to the base pair distance to two fixed reference structures </td></tr>
<tr id="row_7_0_2_" class="even" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__kl__neighborhood__stochbt.html" target="_self">Stochastic Backtracking of Structures from Distance Based Partitioning</a></td><td class="desc">Contains functions related to stochastic backtracking from a specified distance class </td></tr>
<tr id="row_7_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__dos.html" target="_self">Compute the Density of States</a></td><td class="desc"></td></tr>
<tr id="row_8_" class="even"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><a class="el" href="group__inverse__fold.html" target="_self">Inverse Folding (Design)</a></td><td class="desc">RNA sequence design </td></tr>
<tr id="row_9_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_9_" class="arrow" onclick="toggleFolder('9_')">&#9658;</span><a class="el" href="group__probing__data.html" target="_self">Experimental Structure Probing Data</a></td><td class="desc">Include Experimental Structure Probing Data to Guide Structure Predictions </td></tr>
<tr id="row_9_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__SHAPE__reactivities.html" target="_self">SHAPE Reactivity Data</a></td><td class="desc">Incorporate SHAPE reactivity structure probing data into the folding recursions by means of soft constraints </td></tr>
<tr id="row_9_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__perturbation.html" target="_self">Generate Soft Constraints from Data</a></td><td class="desc">Find a vector of perturbation energies that minimizes the discripancies between predicted and observed pairing probabilities and the amount of neccessary adjustments </td></tr>
<tr id="row_10_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_10_" class="arrow" onclick="toggleFolder('10_')">&#9658;</span><a class="el" href="group__ligand__binding.html" target="_self">Ligands Binding to RNA Structures</a></td><td class="desc">Simple Extensions to Model Ligand Binding to RNA Structures </td></tr>
<tr id="row_10_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__ligands__up.html" target="_self">Ligands Binding to Unstructured Domains</a></td><td class="desc">Add ligand binding to loop regions using the <a class="el" href="group__domains__up.html">Unstructured Domains</a> feature </td></tr>
<tr id="row_10_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__constraints__ligand.html" target="_self">Incorporating Ligands Binding to Specific Sequence/Structure Motifs using Soft Constraints</a></td><td class="desc"></td></tr>
<tr id="row_11_"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_11_" class="arrow" onclick="toggleFolder('11_')">&#9658;</span><a class="el" href="group__paired__modules.html" target="_self">Structure Modules and Pseudoknots</a></td><td class="desc"></td></tr>
<tr id="row_11_0_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__pseudoknots.html" target="_self">Pseudoknots</a></td><td class="desc">Implementations to predict pseudoknotted structures </td></tr>
<tr id="row_11_1_" class="even" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__gquads.html" target="_self">G-Quadruplexes</a></td><td class="desc">Various functions related to G-quadruplex computations </td></tr>
<tr id="row_12_" class="even"><td class="entry"><span style="width:0px;display:inline-block;">&#160;</span><span id="arr_12_" class="arrow" onclick="toggleFolder('12_')">&#9658;</span><a class="el" href="group__utils.html" target="_self">Utilities</a></td><td class="desc"></td></tr>
<tr id="row_12_0_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__alphabet__utils.html" target="_self">Utilities to deal with Nucleotide Alphabets</a></td><td class="desc">Functions to cope with various aspects related to the nucleotide sequence alphabet </td></tr>
<tr id="row_12_1_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__string__utils.html" target="_self">(Nucleic Acid Sequence) String Utilitites</a></td><td class="desc">Functions to parse, convert, manipulate, create, and compare (nucleic acid sequence) strings </td></tr>
<tr id="row_12_2_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_2_" class="arrow" onclick="toggleFolder('12_2_')">&#9658;</span><a class="el" href="group__struct__utils.html" target="_self">Secondary Structure Utilities</a></td><td class="desc">Functions to create, parse, convert, manipulate, and compare secondary structure representations </td></tr>
<tr id="row_12_2_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__dot__bracket.html" target="_self">Dot-Bracket Notation of Secondary Structures</a></td><td class="desc">The Dot-Bracket notation as introduced already in the early times of the ViennaRNA Package denotes base pairs by matching pairs of parenthesis <code>()</code> and unpaired nucleotides by dots <code>.</code> </td></tr>
<tr id="row_12_2_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__wuss.html" target="_self">Washington University Secondary Structure (WUSS) notation</a></td><td class="desc">The WUSS notation, as frequently used for consensus secondary structures in <a class="el" href="file_formats.html#msa-formats-stockholm">Stockholm 1.0 format</a> </td></tr>
<tr id="row_12_2_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__pair__table.html" target="_self">Pair Table Representation of Secondary Structures</a></td><td class="desc"></td></tr>
<tr id="row_12_2_3_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__plist.html" target="_self">Pair List Representation of Secondary Structures</a></td><td class="desc"></td></tr>
<tr id="row_12_2_4_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__abstract__shapes.html" target="_self">Abstract Shapes Representation of Secondary Structures</a></td><td class="desc">Abstract Shapes, introduced by Giegerich et al. in (2004) <a class="el" href="citelist.html#CITEREF_giegerich:2004">[10]</a>, collapse the secondary structure while retaining the nestedness of helices and hairpin loops </td></tr>
<tr id="row_12_2_5_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__helix__list.html" target="_self">Helix List Representation of Secondary Structures</a></td><td class="desc"></td></tr>
<tr id="row_12_2_6_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__tree.html" target="_self">Tree Representation of Secondary Structures</a></td><td class="desc">Secondary structures can be readily represented as trees, where internal nodes represent base pairs, and leaves represent unpaired nucleotides. The dot-bracket structure string already is a tree represented by a string of parenthesis (base pairs) and dots for the leaf nodes (unpaired nucleotides) </td></tr>
<tr id="row_12_2_7_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__struct__utils__deprecated.html" target="_self">Deprecated Interface for Secondary Structure Utilities</a></td><td class="desc"></td></tr>
<tr id="row_12_3_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_3_" class="arrow" onclick="toggleFolder('12_3_')">&#9658;</span><a class="el" href="group__aln__utils.html" target="_self">Multiple Sequence Alignment Utilities</a></td><td class="desc">Functions to extract features from and to manipulate multiple sequence alignments </td></tr>
<tr id="row_12_3_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__aln__utils__deprecated.html" target="_self">Deprecated Interface for Multiple Sequence Alignment Utilities</a></td><td class="desc"></td></tr>
<tr id="row_12_4_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_4_" class="arrow" onclick="toggleFolder('12_4_')">&#9658;</span><a class="el" href="group__file__utils.html" target="_self">Files and I/O</a></td><td class="desc">Functions to parse, write, and convert various file formats and to deal with file system related issues </td></tr>
<tr id="row_12_4_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__file__formats.html" target="_self">Nucleic Acid Sequences and Structures</a></td><td class="desc">Functions to read/write different file formats for nucleic acid sequences and secondary structures </td></tr>
<tr id="row_12_4_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__file__formats__msa.html" target="_self">Multiple Sequence Alignments</a></td><td class="desc">Functions to read/write multiple sequence alignments (MSA) in various file formats </td></tr>
<tr id="row_12_4_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__command__files.html" target="_self">Command Files</a></td><td class="desc">Functions to parse and interpret the content of <a class="el" href="file_formats.html#constraint-formats-file">Command Files</a> </td></tr>
<tr id="row_12_5_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_5_" class="arrow" onclick="toggleFolder('12_5_')">&#9658;</span><a class="el" href="group__plotting__utils.html" target="_self">Plotting</a></td><td class="desc">Functions for Creating Secondary Structure Plots, Dot-Plots, and More </td></tr>
<tr id="row_12_5_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__plot__layout__utils.html" target="_self">Layouts and Coordinates</a></td><td class="desc">Functions to compute coordinate layouts for secondary structure plots </td></tr>
<tr id="row_12_5_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__annotation__utils.html" target="_self">Annotation</a></td><td class="desc">Functions to generate annotations for Secondary Structure Plots, Dot-Plots, and Others </td></tr>
<tr id="row_12_5_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__alignment__plots.html" target="_self">Alignment Plots</a></td><td class="desc">Functions to generate Alignment plots with annotated consensus structure </td></tr>
<tr id="row_12_5_3_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__plotting__utils__deprecated.html" target="_self">Deprecated Interface for Plotting Utilities</a></td><td class="desc"></td></tr>
<tr id="row_12_6_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__search__utils.html" target="_self">Search Algorithms</a></td><td class="desc">Implementations of various search algorithms to detect strings of objects within other strings of objects </td></tr>
<tr id="row_12_7_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__combinatorics__utils.html" target="_self">Combinatorics Algorithms</a></td><td class="desc">Implementations to solve various combinatorial aspects for strings of objects </td></tr>
<tr id="row_12_8_" style="display:none;"><td class="entry"><span style="width:16px;display:inline-block;">&#160;</span><span id="arr_12_8_" class="arrow" onclick="toggleFolder('12_8_')">&#9658;</span><a class="el" href="group__data__structures.html" target="_self">(Abstract) Data Structures</a></td><td class="desc">All datastructures and typedefs shared among the ViennaRNA Package can be found here </td></tr>
<tr id="row_12_8_0_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__fold__compound.html" target="_self">The Fold Compound</a></td><td class="desc">This module provides interfaces that deal with the most basic data structure used in structure predicting and energy evaluating function of the RNAlib </td></tr>
<tr id="row_12_8_1_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__dp__matrices.html" target="_self">The Dynamic Programming Matrices</a></td><td class="desc">This module provides interfaces that deal with creation and destruction of dynamic programming matrices used within the RNAlib </td></tr>
<tr id="row_12_8_2_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__hash__table__utils.html" target="_self">Hash Tables</a></td><td class="desc">Various implementations of hash table functions </td></tr>
<tr id="row_12_8_3_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__heap__utils.html" target="_self">Heaps</a></td><td class="desc">Interface for an abstract implementation of a heap data structure </td></tr>
<tr id="row_12_8_4_" style="display:none;"><td class="entry"><span style="width:48px;display:inline-block;">&#160;</span><a class="el" href="group__buffer__utils.html" target="_self">Buffers</a></td><td class="desc">Functions that provide dynamically buffered stream-like data structures </td></tr>
<tr id="row_12_9_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__message__utils.html" target="_self">Messages</a></td><td class="desc">Functions to print various kind of messages </td></tr>
<tr id="row_12_10_" style="display:none;"><td class="entry"><span style="width:32px;display:inline-block;">&#160;</span><a class="el" href="group__units.html" target="_self">Unit Conversion</a></td><td class="desc">Functions to convert between various physical units </td></tr>
</table>
</div><!-- directory -->
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="footer">Generated on Fri Apr 23 2021 00:39:34 for RNAlib-2.4.18 by <a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1 </li>
  </ul>
</div>
</body>
</html>
