\doxysection{File List}
Here is a list of all documented files with brief descriptions\+:\begin{DoxyCompactList}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{2Dfold_8h}{2\+Dfold.\+h}} \\*MFE structures for base pair distance classes }{\pageref{2Dfold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{2Dpfold_8h}{2\+Dpfold.\+h}} \\*Partition function implementations for base pair distance classes }{\pageref{2Dpfold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries ali\+\_\+plex.\+h} }{\pageref{ali__plex_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{alifold_8h}{alifold.\+h}} \\*Functions for comparative structure prediction using RNA sequence alignments }{\pageref{alifold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{aln__util_8h}{aln\+\_\+util.\+h}} \\*Use \mbox{\hyperlink{utils_2alignments_8h}{Vienna\+RNA/utils/alignments.\+h}} instead }{\pageref{aln__util_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{alphabet_8h}{alphabet.\+h}} \\*Functions to process, convert, and generally handle different nucleotide and/or base pair alphabets }{\pageref{alphabet_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{boltzmann__sampling_8h}{boltzmann\+\_\+sampling.\+h}} \\*Boltzmann Sampling of secondary structures from the ensemble }{\pageref{boltzmann__sampling_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{centroid_8h}{centroid.\+h}} \\*Centroid structure computation }{\pageref{centroid_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{char__stream_8h}{char\+\_\+stream.\+h}} \\*Use \mbox{\hyperlink{datastructures_2char__stream_8h}{Vienna\+RNA/datastructures/char\+\_\+stream.\+h}} instead }{\pageref{char__stream_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{cofold_8h}{cofold.\+h}} \\*MFE implementations for RNA-\/\+RNA interaction }{\pageref{cofold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{combinatorics_8h}{combinatorics.\+h}} \\*Various implementations that deal with combinatorial aspects of objects }{\pageref{combinatorics_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{commands_8h}{commands.\+h}} \\*Parse and apply different commands that alter the behavior of secondary structure prediction and evaluation }{\pageref{commands_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{concentrations_8h}{concentrations.\+h}} \\*Concentration computations for RNA-\/\+RNA interactions }{\pageref{concentrations_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{constraints_8h}{constraints.\+h}} \\*Use \mbox{\hyperlink{constraints_2basic_8h}{Vienna\+RNA/constraints/basic.\+h}} instead }{\pageref{constraints_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{constraints__hard_8h}{constraints\+\_\+hard.\+h}} \\*Use \mbox{\hyperlink{hard_8h}{Vienna\+RNA/constraints/hard.\+h}} instead }{\pageref{constraints__hard_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{constraints__ligand_8h}{constraints\+\_\+ligand.\+h}} \\*Use \mbox{\hyperlink{ligand_8h}{Vienna\+RNA/constraints/ligand.\+h}} instead }{\pageref{constraints__ligand_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{constraints__SHAPE_8h}{constraints\+\_\+\+SHAPE.\+h}} \\*Use \mbox{\hyperlink{SHAPE_8h}{Vienna\+RNA/constraints/\+SHAPE.\+h}} instead }{\pageref{constraints__SHAPE_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{constraints__soft_8h}{constraints\+\_\+soft.\+h}} \\*Use \mbox{\hyperlink{soft_8h}{Vienna\+RNA/constraints/soft.\+h}} instead }{\pageref{constraints__soft_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{convert__epars_8h}{convert\+\_\+epars.\+h}} \\*Use \mbox{\hyperlink{convert_8h}{Vienna\+RNA/params/convert.\+h}} instead }{\pageref{convert__epars_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{data__structures_8h}{data\+\_\+structures.\+h}} \\*Use \mbox{\hyperlink{datastructures_2basic_8h}{Vienna\+RNA/datastructures/basic.\+h}} instead }{\pageref{data__structures_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{dist__vars_8h}{dist\+\_\+vars.\+h}} \\*Global variables for Distance-\/\+Package }{\pageref{dist__vars_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{dp__matrices_8h}{dp\+\_\+matrices.\+h}} \\*Functions to deal with standard dynamic programming (DP) matrices }{\pageref{dp__matrices_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{duplex_8h}{duplex.\+h}} \\*Functions for simple RNA-\/\+RNA duplex interactions }{\pageref{duplex_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{edit__cost_8h}{edit\+\_\+cost.\+h}} \\*Global variables for Edit Costs included by treedist.\+c and stringdist.\+c }{\pageref{edit__cost_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{energy__const_8h}{energy\+\_\+const.\+h}} \\*Use \mbox{\hyperlink{constants_8h}{Vienna\+RNA/params/constants.\+h}} instead }{\pageref{energy__const_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{energy__par_8h}{energy\+\_\+par.\+h}} \\*Use \mbox{\hyperlink{default_8h_source}{Vienna\+RNA/params/default.\+h}} instead }{\pageref{energy__par_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{equilibrium__probs_8h}{equilibrium\+\_\+probs.\+h}} \\*Equilibrium Probability implementations }{\pageref{equilibrium__probs_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{eval_8h}{eval.\+h}} \\*Functions and variables related to energy evaluation of sequence/structure pairs }{\pageref{eval_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{exterior__loops_8h}{exterior\+\_\+loops.\+h}} \\*Use \mbox{\hyperlink{external_8h}{Vienna\+RNA/loops/external.\+h}} instead }{\pageref{exterior__loops_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{file__formats_8h}{file\+\_\+formats.\+h}} \\*Use \mbox{\hyperlink{io_2file__formats_8h}{Vienna\+RNA/io/file\+\_\+formats.\+h}} instead }{\pageref{file__formats_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{file__formats__msa_8h}{file\+\_\+formats\+\_\+msa.\+h}} \\*Use \mbox{\hyperlink{io_2file__formats__msa_8h}{Vienna\+RNA/io/file\+\_\+formats\+\_\+msa.\+h}} instead }{\pageref{file__formats__msa_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{file__utils_8h}{file\+\_\+utils.\+h}} \\*Use \mbox{\hyperlink{io_2utils_8h}{Vienna\+RNA/io/utils.\+h}} instead }{\pageref{file__utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{findpath_8h}{findpath.\+h}} \\*Use \mbox{\hyperlink{landscape_2findpath_8h}{Vienna\+RNA/landscape/findpath.\+h}} instead }{\pageref{findpath_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{fold_8h}{fold.\+h}} \\*MFE calculations for single RNA sequences }{\pageref{fold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{fold__compound_8h}{fold\+\_\+compound.\+h}} \\*The Basic Fold Compound API }{\pageref{fold__compound_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{fold__vars_8h}{fold\+\_\+vars.\+h}} \\*Here all all declarations of the global variables used throughout RNAlib }{\pageref{fold__vars_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{gquad_8h}{gquad.\+h}} \\*G-\/quadruplexes }{\pageref{gquad_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{grammar_8h}{grammar.\+h}} \\*Implementations for the RNA folding grammar }{\pageref{grammar_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{hairpin__loops_8h}{hairpin\+\_\+loops.\+h}} \\*Use \mbox{\hyperlink{hairpin_8h}{Vienna\+RNA/loops/hairpin.\+h}} instead }{\pageref{hairpin__loops_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{heat__capacity_8h}{heat\+\_\+capacity.\+h}} \\*Compute heat capacity for an RNA }{\pageref{heat__capacity_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{interior__loops_8h}{interior\+\_\+loops.\+h}} \\*Use \mbox{\hyperlink{internal_8h}{Vienna\+RNA/loops/internal.\+h}} instead }{\pageref{interior__loops_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{inverse_8h}{inverse.\+h}} \\*Inverse folding routines }{\pageref{inverse_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{Lfold_8h}{Lfold.\+h}} \\*Functions for locally optimal MFE structure prediction }{\pageref{Lfold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{loop__energies_8h}{loop\+\_\+energies.\+h}} \\*Use \mbox{\hyperlink{all_8h}{Vienna\+RNA/loops/all.\+h}} instead }{\pageref{loop__energies_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{LPfold_8h}{LPfold.\+h}} \\*Partition function and equilibrium probability implementation for the sliding window algorithm }{\pageref{LPfold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{MEA_8h}{MEA.\+h}} \\*Computes a MEA (maximum expected accuracy) structure }{\pageref{MEA_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{mfe_8h}{mfe.\+h}} \\*Compute Minimum Free energy (MFE) and backtrace corresponding secondary structures from RNA sequence data }{\pageref{mfe_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{mfe__window_8h}{mfe\+\_\+window.\+h}} \\*Compute local Minimum Free Energy (MFE) using a sliding window approach and backtrace corresponding secondary structures }{\pageref{mfe__window_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{mm_8h}{mm.\+h}} \\*Several Maximum Matching implementations }{\pageref{mm_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{model_8h}{model.\+h}} \\*The model details data structure and its corresponding modifiers }{\pageref{model_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries move\+\_\+set.\+h} }{\pageref{move__set_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{multibranch__loops_8h}{multibranch\+\_\+loops.\+h}} \\*Use \mbox{\hyperlink{multibranch_8h}{Vienna\+RNA/loops/multibranch.\+h}} instead }{\pageref{multibranch__loops_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{naview_8h}{naview.\+h}} \\*Use \mbox{\hyperlink{plotting_2naview_8h}{Vienna\+RNA/plotting/naview.\+h}} instead }{\pageref{naview_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{neighbor_8h}{neighbor.\+h}} \\*Use \mbox{\hyperlink{landscape_2neighbor_8h}{Vienna\+RNA/landscape/neighbor.\+h}} instead }{\pageref{neighbor_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries pair\+\_\+mat.\+h} }{\pageref{pair__mat_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{params_8h}{params.\+h}} \\*Use \mbox{\hyperlink{params_2basic_8h}{Vienna\+RNA/params/basic.\+h}} instead }{\pageref{params_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{part__func_8h}{part\+\_\+func.\+h}} \\*Partition function implementations }{\pageref{part__func_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{part__func__co_8h}{part\+\_\+func\+\_\+co.\+h}} \\*Partition function for two RNA sequences }{\pageref{part__func__co_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{part__func__up_8h}{part\+\_\+func\+\_\+up.\+h}} \\*Implementations for accessibility and RNA-\/\+RNA interaction as a stepwise process }{\pageref{part__func__up_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{part__func__window_8h}{part\+\_\+func\+\_\+window.\+h}} \\*Partition function and equilibrium probability implementation for the sliding window algorithm }{\pageref{part__func__window_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{perturbation__fold_8h}{perturbation\+\_\+fold.\+h}} \\*Find a vector of perturbation energies that minimizes the discripancies between predicted and observed pairing probabilities and the amount of neccessary adjustments }{\pageref{perturbation__fold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{pk__plex_8h}{pk\+\_\+plex.\+h}} \\*Heuristics for two-\/step pseudoknot forming interaction predictions }{\pageref{pk__plex_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries PKplex.\+h} }{\pageref{PKplex_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries plex.\+h} }{\pageref{plex_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{plot__aln_8h}{plot\+\_\+aln.\+h}} \\*Use \mbox{\hyperlink{plotting_2alignments_8h}{Vienna\+RNA/plotting/alignments.\+h}} instead }{\pageref{plot__aln_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{plot__layouts_8h}{plot\+\_\+layouts.\+h}} \\*Use \mbox{\hyperlink{layouts_8h}{Vienna\+RNA/plotting/layouts.\+h}} instead }{\pageref{plot__layouts_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{plot__structure_8h}{plot\+\_\+structure.\+h}} \\*Use \mbox{\hyperlink{plotting_2structures_8h}{Vienna\+RNA/plotting/structures.\+h}} instead }{\pageref{plot__structure_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{plot__utils_8h}{plot\+\_\+utils.\+h}} \\*Use \mbox{\hyperlink{plotting_2utils_8h}{Vienna\+RNA/plotting/utils.\+h}} instead }{\pageref{plot__utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries Profile\+Aln.\+h} }{\pageref{ProfileAln_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{profiledist_8h}{profiledist.\+h}} }{\pageref{profiledist_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{PS__dot_8h}{PS\+\_\+dot.\+h}} \\*Use \mbox{\hyperlink{probabilities_8h}{Vienna\+RNA/plotting/probabilities.\+h}} instead }{\pageref{PS__dot_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{read__epars_8h}{read\+\_\+epars.\+h}} \\*Use \mbox{\hyperlink{io_8h}{Vienna\+RNA/params/io.\+h}} instead }{\pageref{read__epars_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{ribo_8h}{ribo.\+h}} \\*Parse Ribo\+Sum Scoring Matrices for Covariance Scoring of Alignments }{\pageref{ribo_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{RNAstruct_8h}{RNAstruct.\+h}} \\*Parsing and Coarse Graining of Structures }{\pageref{RNAstruct_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{sequence_8h}{sequence.\+h}} \\*Functions and data structures related to sequence representations , }{\pageref{sequence_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries snofold.\+h} }{\pageref{snofold_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries snoop.\+h} }{\pageref{snoop_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries special\+\_\+const.\+h} }{\pageref{special__const_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{stream__output_8h}{stream\+\_\+output.\+h}} \\*Use \mbox{\hyperlink{datastructures_2stream__output_8h}{Vienna\+RNA/datastructures/stream\+\_\+output.\+h}} instead }{\pageref{stream__output_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{string__utils_8h}{string\+\_\+utils.\+h}} \\*Use \mbox{\hyperlink{strings_8h}{Vienna\+RNA/utils/strings.\+h}} instead }{\pageref{string__utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{stringdist_8h}{stringdist.\+h}} \\*Functions for String Alignment }{\pageref{stringdist_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{structure__utils_8h}{structure\+\_\+utils.\+h}} \\*Use \mbox{\hyperlink{utils_2structures_8h}{Vienna\+RNA/utils/structures.\+h}} instead }{\pageref{structure__utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{structured__domains_8h}{structured\+\_\+domains.\+h}} \\*This module provides interfaces that deal with additional structured domains in the folding grammar }{\pageref{structured__domains_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{subopt_8h}{subopt.\+h}} \\*RNAsubopt and density of states declarations }{\pageref{subopt_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{svm__utils_8h}{svm\+\_\+utils.\+h}} \\*Use \mbox{\hyperlink{svm_8h_source}{Vienna\+RNA/utils/svm.\+h}} instead }{\pageref{svm__utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{treedist_8h}{treedist.\+h}} \\*Functions for \mbox{\hyperlink{structTree}{Tree}} Edit Distances }{\pageref{treedist_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries ugly\+\_\+bt.\+h} }{\pageref{ugly__bt_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{units_8h}{units.\+h}} \\*Use \mbox{\hyperlink{utils_2units_8h}{Vienna\+RNA/utils/units.\+h}} instead }{\pageref{units_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{unstructured__domains_8h}{unstructured\+\_\+domains.\+h}} \\*Functions to modify unstructured domains, e.\+g. to incorporate ligands binding to unpaired stretches }{\pageref{unstructured__domains_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{utils_8h}{utils.\+h}} \\*Use \mbox{\hyperlink{utils_2basic_8h}{Vienna\+RNA/utils/basic.\+h}} instead }{\pageref{utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries vrna\+\_\+config.\+h} }{\pageref{vrna__config_8h}}{}
\item\contentsline{section}{Vienna\+RNA/\mbox{\hyperlink{walk_8h}{walk.\+h}} \\*Use \mbox{\hyperlink{landscape_2walk_8h}{Vienna\+RNA/landscape/walk.\+h}} instead }{\pageref{walk_8h}}{}
\item\contentsline{section}{Vienna\+RNA/{\bfseries zscore.\+h} }{\pageref{zscore_8h}}{}
\item\contentsline{section}{Vienna\+RNA/constraints/\mbox{\hyperlink{constraints_2basic_8h}{basic.\+h}} \\*Functions and data structures for constraining secondary structure predictions and evaluation }{\pageref{constraints_2basic_8h}}{}
\item\contentsline{section}{Vienna\+RNA/constraints/\mbox{\hyperlink{hard_8h}{hard.\+h}} \\*Functions and data structures for handling of secondary structure hard constraints }{\pageref{hard_8h}}{}
\item\contentsline{section}{Vienna\+RNA/constraints/\mbox{\hyperlink{ligand_8h}{ligand.\+h}} \\*Functions for incorporation of ligands binding to hairpin and interior loop motifs using the soft constraints framework }{\pageref{ligand_8h}}{}
\item\contentsline{section}{Vienna\+RNA/constraints/\mbox{\hyperlink{SHAPE_8h}{SHAPE.\+h}} \\*This module provides function to incorporate SHAPE reactivity data into the folding recursions by means of soft constraints }{\pageref{SHAPE_8h}}{}
\item\contentsline{section}{Vienna\+RNA/constraints/\mbox{\hyperlink{soft_8h}{soft.\+h}} \\*Functions and data structures for secondary structure soft constraints }{\pageref{soft_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/\mbox{\hyperlink{datastructures_2basic_8h}{basic.\+h}} \\*Various data structures and pre-\/processor macros }{\pageref{datastructures_2basic_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/\mbox{\hyperlink{datastructures_2char__stream_8h}{char\+\_\+stream.\+h}} \\*Implementation of a dynamic, buffered character stream }{\pageref{datastructures_2char__stream_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/\mbox{\hyperlink{hash__tables_8h}{hash\+\_\+tables.\+h}} \\*Implementations of hash table functions }{\pageref{hash__tables_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/\mbox{\hyperlink{heap_8h}{heap.\+h}} \\*Implementation of an abstract heap data structure }{\pageref{heap_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/{\bfseries lists.\+h} }{\pageref{lists_8h}}{}
\item\contentsline{section}{Vienna\+RNA/datastructures/\mbox{\hyperlink{datastructures_2stream__output_8h}{stream\+\_\+output.\+h}} \\*An implementation of a buffered, ordered stream output data structure }{\pageref{datastructures_2stream__output_8h}}{}
\item\contentsline{section}{Vienna\+RNA/io/\mbox{\hyperlink{io_2file__formats_8h}{file\+\_\+formats.\+h}} \\*Read and write different file formats for RNA sequences, structures }{\pageref{io_2file__formats_8h}}{}
\item\contentsline{section}{Vienna\+RNA/io/\mbox{\hyperlink{io_2file__formats__msa_8h}{file\+\_\+formats\+\_\+msa.\+h}} \\*Functions dealing with file formats for Multiple Sequence Alignments (MSA) }{\pageref{io_2file__formats__msa_8h}}{}
\item\contentsline{section}{Vienna\+RNA/io/\mbox{\hyperlink{io_2utils_8h}{utils.\+h}} \\*Several utilities for file handling }{\pageref{io_2utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/landscape/\mbox{\hyperlink{landscape_2findpath_8h}{findpath.\+h}} \\*A breadth-\/first search heuristic for optimal direct folding paths }{\pageref{landscape_2findpath_8h}}{}
\item\contentsline{section}{Vienna\+RNA/landscape/\mbox{\hyperlink{move_8h}{move.\+h}} \\*Methods to operate with structural neighbors of RNA secondary structures }{\pageref{move_8h}}{}
\item\contentsline{section}{Vienna\+RNA/landscape/\mbox{\hyperlink{landscape_2neighbor_8h}{neighbor.\+h}} \\*Methods to compute the neighbors of an RNA secondary structure }{\pageref{landscape_2neighbor_8h}}{}
\item\contentsline{section}{Vienna\+RNA/landscape/\mbox{\hyperlink{paths_8h}{paths.\+h}} \\*API for computing (optimal) (re-\/)folding paths between secondary structures }{\pageref{paths_8h}}{}
\item\contentsline{section}{Vienna\+RNA/landscape/\mbox{\hyperlink{landscape_2walk_8h}{walk.\+h}} \\*Methods to generate particular paths such as gradient or random walks through the energy landscape of an RNA sequence }{\pageref{landscape_2walk_8h}}{}
\item\contentsline{section}{Vienna\+RNA/loops/\mbox{\hyperlink{all_8h}{all.\+h}} \\*Energy evaluation for MFE and partition function calculations }{\pageref{all_8h}}{}
\item\contentsline{section}{Vienna\+RNA/loops/\mbox{\hyperlink{external_8h}{external.\+h}} \\*Energy evaluation of exterior loops for MFE and partition function calculations }{\pageref{external_8h}}{}
\item\contentsline{section}{Vienna\+RNA/loops/\mbox{\hyperlink{hairpin_8h}{hairpin.\+h}} \\*Energy evaluation of hairpin loops for MFE and partition function calculations }{\pageref{hairpin_8h}}{}
\item\contentsline{section}{Vienna\+RNA/loops/\mbox{\hyperlink{internal_8h}{internal.\+h}} \\*Energy evaluation of interior loops for MFE and partition function calculations }{\pageref{internal_8h}}{}
\item\contentsline{section}{Vienna\+RNA/loops/\mbox{\hyperlink{multibranch_8h}{multibranch.\+h}} \\*Energy evaluation of multibranch loops for MFE and partition function calculations }{\pageref{multibranch_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{1_88_84__epars_8h}{1.\+8.\+4\+\_\+epars.\+h}} \\*Free energy parameters for parameter file conversion }{\pageref{1_88_84__epars_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{1_88_84__intloops_8h}{1.\+8.\+4\+\_\+intloops.\+h}} \\*Free energy parameters for interior loop contributions needed by the parameter file conversion functions }{\pageref{1_88_84__intloops_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{params_2basic_8h}{basic.\+h}} \\*Functions to deal with sets of energy parameters }{\pageref{params_2basic_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{constants_8h}{constants.\+h}} \\*Energy parameter constants }{\pageref{constants_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{convert_8h}{convert.\+h}} \\*Functions and definitions for energy parameter file format conversion }{\pageref{convert_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries default.\+h} }{\pageref{default_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl11.\+h} }{\pageref{intl11_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl11d\+H.\+h} }{\pageref{intl11dH_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl21.\+h} }{\pageref{intl21_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl21d\+H.\+h} }{\pageref{intl21dH_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl22.\+h} }{\pageref{intl22_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/{\bfseries intl22d\+H.\+h} }{\pageref{intl22dH_8h}}{}
\item\contentsline{section}{Vienna\+RNA/params/\mbox{\hyperlink{io_8h}{io.\+h}} \\*Read and write energy parameter files }{\pageref{io_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{plotting_2alignments_8h}{alignments.\+h}} \\*Various functions for plotting Sequence / Structure Alignments }{\pageref{plotting_2alignments_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{layouts_8h}{layouts.\+h}} \\*Secondary structure plot layout algorithms }{\pageref{layouts_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{plotting_2naview_8h}{naview.\+h}} \\*Implementation of the Naview RNA secondary structure layout algorithm \cite{bruccoleri:1988} }{\pageref{plotting_2naview_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{probabilities_8h}{probabilities.\+h}} \\*Various functions for plotting RNA secondary structures, dot-\/plots and other visualizations }{\pageref{probabilities_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{plotting_2structures_8h}{structures.\+h}} \\*Various functions for plotting RNA secondary structures }{\pageref{plotting_2structures_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\mbox{\hyperlink{plotting_2utils_8h}{utils.\+h}} \\*Various utilities to assist in plotting secondary structures and consensus structures }{\pageref{plotting_2utils_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\+RNApuzzler/\mbox{\hyperlink{RNApuzzler_8h}{RNApuzzler.\+h}} \\*Implementation of the RNApuzzler RNA secondary structure layout algorithm \cite{wiegreffe:2018} }{\pageref{RNApuzzler_8h}}{}
\item\contentsline{section}{Vienna\+RNA/plotting/\+RNApuzzler/\mbox{\hyperlink{RNAturtle_8h}{RNAturtle.\+h}} \\*Implementation of the RNAturtle RNA secondary structure layout algorithm \cite{wiegreffe:2018} }{\pageref{RNAturtle_8h}}{}
\item\contentsline{section}{Vienna\+RNA/search/\mbox{\hyperlink{BoyerMoore_8h}{Boyer\+Moore.\+h}} \\*Variants of the Boyer-\/\+Moore string search algorithm }{\pageref{BoyerMoore_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/\mbox{\hyperlink{utils_2alignments_8h}{alignments.\+h}} \\*Various utility-\/ and helper-\/functions for sequence alignments and comparative structure prediction }{\pageref{utils_2alignments_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/\mbox{\hyperlink{utils_2basic_8h}{basic.\+h}} \\*General utility-\/ and helper-\/functions used throughout the {\itshape Vienna\+RNA} {\itshape Package} }{\pageref{utils_2basic_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/{\bfseries cpu.\+h} }{\pageref{cpu_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/{\bfseries higher\+\_\+order\+\_\+functions.\+h} }{\pageref{higher__order__functions_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/\mbox{\hyperlink{strings_8h}{strings.\+h}} \\*General utility-\/ and helper-\/functions for RNA sequence and structure strings used throughout the Vienna\+RNA Package }{\pageref{strings_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/\mbox{\hyperlink{utils_2structures_8h}{structures.\+h}} \\*Various utility-\/ and helper-\/functions for secondary structure parsing, converting, etc }{\pageref{utils_2structures_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/{\bfseries svm.\+h} }{\pageref{svm_8h}}{}
\item\contentsline{section}{Vienna\+RNA/utils/\mbox{\hyperlink{utils_2units_8h}{units.\+h}} \\*Physical Units and Functions to convert them into each other }{\pageref{utils_2units_8h}}{}
\end{DoxyCompactList}
