# Acknowledgements
This work was funded by the National Aeronautics and Space Administration grant 80NSSC18K1139 awarded to LDW and ASP.
We would like to thank the following collaborators, whose input has been invaluable to our work:

- Prof. Khanh Dao Duc & Mr. Artem Kushner at University of British Columbia (https://kdaoduc.com/)

- Mr. Rohan Gupta at Georgia Institute of Technology

- The group of George Fox at University of Houston

- Dr. Burak Gulen, Postdoctoral Scientist, University Medical Center Hamburg Eppendorf

- Ms. Yulia Dumov, Sr. UX/UI Designer, Tufts Technology Services (TTS)

Our webserver would not be possible without these free resources:

- Sequences and taxonomies were retrieved from the [National Center for Biotechnology Information](https://www.ncbi.nlm.nih.gov/) and [UniProt](https://www.uniprot.org/)

- PDB identifiers are filtered through the API provided by Dr. Khanh Dao Duc and Mr. Artem Kushner through their project for ribosomal structure analysis [riboXYZ](https://ribosome.xyz/).

- rProtein structural annotation was retrieved from [ECOD](http://prodata.swmed.edu/ecod/) (10.1371/journal.pcbi.1003926; 10.1002/prot.24818)

- Alignment viewer – modified version of the react implementation for the [MSAViewer](https://github.com/plotly/react-msa-viewer) (10.1093/bioinformatics/btw474)

- Topology Viewer – modified version of the [PDBe topology viewer](https://github.com/PDBeurope/pdb-topology-viewer)

- PDB topologies provided through the [EBI PDBe REST API](https://www.ebi.ac.uk/pdbe/api/doc/topology.html)

- PDB topologies for custom structures are generated with [Pro-Origami](http://munk.cis.unimelb.edu.au/pro-origami/)

- [MolStar Viewer](https://molstar.org/viewer/) ; [Github](https://github.com/molstar/molstar)

- Parsing of sequences and alignments and calculation of frequencies was possible thanks to the [Biopython package](https://biopython.org/)
