/*
  File autogenerated by gengetopt version 2.23
  generated with the following command:
  /usr/bin/gengetopt -i RNAplex.ggo --file-name=RNAplex_cmdl --include-getopt --default-optional --func-name=RNAplex_cmdline_parser --arg-struct-name=RNAplex_args_info

  The developers of gengetopt consider the fixed text that goes in all
  gengetopt output files to be in the public domain:
  we make no copyright claims on it.
*/

/* If we use autoconf.  */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifndef FIX_UNUSED
#define FIX_UNUSED(X) (void) (X) /* avoid warnings for unused params */
#endif


#include "RNAplex_cmdl.h"

const char *RNAplex_args_info_purpose = "Find targets of a query RNA";

const char *RNAplex_args_info_usage = "Usage: RNAplex [options]\n";

const char *RNAplex_args_info_versiontext = "";

const char *RNAplex_args_info_description = "reads two RNA sequences from stdin or <filename> and computes optimal and\nsuboptimal secondary structures for their hybridization. The calculation is\nsimplified by allowing only inter-molecular base pairs. Accessibility effects\ncan be estimated by RNAplex if a RNAplfold accessibility profile is provided.\nThe computed optimal and suboptimal structure are written to stdout, one\nstructure per line. Each line consist of: The structure in dot bracket format\nwith a \"&\" separating the two strands. The range of the structure in the two\nsequences in the format  \"from,to : from,to\"; the energy of duplex structure\nin kcal/mol.\nThe format is especially useful for computing the hybrid structure between a\nsmall probe sequence and a long target sequence.\n";

const char *RNAplex_args_info_detailed_help[] = {
  "  -h, --help                    Print help and exit",
  "      --detailed-help           Print help, including all details and hidden\n                                  options, and exit",
  "      --version                 Print version and exit",
  "\nInput Options:",
  "  Below are command line options which alter the general input behavior of this\n  program\n",
  "  -q, --query=STRING            File containing the query sequence.\n",
  "  Input sequences can be given piped to RNAplex or given in a query file with\n  the -q option. Note that the -q option implies that the -t option is also\n  used\n\n",
  "  -t, --target=STRING           File containing the target sequence.\n",
  "  Input sequences can be given piped to RNAplex or given in a target file with\n  the -t option. Note that the -t option implies that the -q option is also\n  used\n\n",
  "  -a, --accessibility-dir=STRING\n                                Location of the accessibility profiles.\n",
  "  This option switches the accessibility modes on and indicates in which\n  directory accessibility profiles as generated by RNAplfold can be found\n\n",
  "  -b, --binary                  Allow the reading and parsing of memory dumped\n                                  opening energy file\n                                    (default=off)",
  "  The -b option allows one to read and process opening energy files which are\n  saved in binary format\n   This can reduce by a factor of 500x-1000x the time needed to process those\n  files. RNAplex recognizes the corresponding opening energy files by looking\n  for files named after the sequence and containing the suffix _openen_bin.\n  Please look at the man page of RNAplfold if you need more information on how\n  to produce binary opening energy files.\n\n",
  "  -P, --paramFile=paramfile     Read energy parameters from paramfile, instead\n                                  of using the default parameter set.\n",
  "  Different sets of energy parameters for RNA and DNA should accompany your\n  distribution.\n  See the RNAlib documentation for details on the file format. When passing the\n  placeholder file name \"DNA\", DNA parameters are loaded without the need to\n  actually specify any input file.\n\n",
  "\nAlgorithms:",
  "  Options which alter the computing behaviour of RNAplex.\n\n",
  "  -T, --temp=DOUBLE             Rescale energy parameters to a temperature T.\n                                  Default is 37C.\n\n",
  "  -l, --interaction-length=INT  Maximal length of an interaction\n                                    (default=`40')",
  "  Maximal allowed length of an interaction\n\n",
  "  -c, --extension-cost=INT      Cost to add to each nucleotide in a duplex\n                                    (default=`0')",
  "  Cost of extending a duplex by one nucleotide. Allows one to find compact\n  duplexes, having few/small bulges or interior loops Only useful when no\n  accessibility profiles are available. This option is disabled if\n  accessibility profiles are used (-a option)\n\n",
  "  -p, --probe-mode              Compute Tm for probes  (default=off)",
  "  Use this option if you want to compute the melting temperature of your probes\n\n",
  "  -Q, --probe-concentration=DOUBLE\n                                Set the probe concentration for the Tm\n                                  computation\n\n                                    (default=`0.1')",
  "  -N, --na-concentration=DOUBLE Set the Na+ concentration for the Tm\n                                  computation\n\n                                    (default=`1.0')",
  "  -M, --mg-concentration=DOUBLE Set the Mg2+ concentration for the Tm\n                                  computation\n\n                                    (default=`1.0')",
  "  -K, --k-concentration=DOUBLE  Set the K+ concentration for the Tm computation\n\n                                    (default=`1.0')",
  "  -U, --tris-concentration=DOUBLE\n                                Set the tris+ concentration for the Tm\n                                  computation\n\n                                    (default=`1.0')",
  "  -f, --fast-folding=INT        Speedup of the target search\n                                    (default=`0')",
  "  This option allows one to decide if the backtracking has to be done (-f 0, -f\n  2) or not (-f 1). For -f 0 the structure is computed based on the standard\n  energy model. This is the slowest and most precise mode of RNAplex. With -f\n  2, the structure is computed based on the approximated plex model. If a lot\n  of targets are returned this is can greatly improve the runtime of RNAplex.\n  -f 1 is the fastest mode, as no structure are recomputed \n\n",
  "  -V, --scale-accessibility=DOUBLE\n                                Rescale all opening energy by a factor V\n                                    (default=`1.0')",
  "  Scale-factor for the accessibility. If V is set to 1 then the scaling has no\n  effect on the accessibility.\n\n",
  "  -C, --constraint              Calculate structures subject to constraints.\n                                    (default=off)",
  "  The program reads first the sequence, then a string containing constraints on\n  the structure for the query sequence encoded with the symbols:\n  . (no constraint for this base)\n  | (the corresponding base has to be paired)\n\n",
  "  -A, --alignment-mode          Tells RNAplex to compute interactions based on\n                                  alignments\n                                    (default=off)",
  "  If the A option is set RNAplex expects clustalw files as input for the -q and\n  -t option.\n\n",
  "  -k, --convert-to-bin          If set, RNAplex will convert all opening energy\n                                  file in a directory set by the -a option into\n                                  binary opening energy files\n                                    (default=off)",
  "  RNAplex can be used to convert existing text formatted opening energy files\n  into binary formatted files. In this mode RNAplex does not compute\n  interactions.\n\n",
  "\nOutput:",
  "  Options that modify the output\n\n",
  "  -z, --duplex-distance=INT     Distance between target 3' ends of two\n                                  consecutive duplexes\n                                    (default=`0')",
  "  Distance between the target 3'ends of two consecutive duplexes. Should be set\n  to the maximal length of interaction to get good results\n   Smaller z leads to larger overlaps between consecutive duplexes.\n\n",
  "  -e, --energy-threshold=DOUBLE Minimal energy for a duplex to be returned\n                                    (default=`-100000')",
  "  Energy threshold for a duplex to be returned. The threshold is set on the\n  total energy of interaction, i.e. the hybridization energy corrected for\n  opening energy if -a is set or the energy corrected by -c. If unset, only the\n  mfe will be returned\n\n",
  "  -I, --produce-ps=STRING       Draw an alignment annotated interaction from\n                                  RNAplex\n",
  "  This option allows one to produce interaction figures in PS-format a la\n  RNAalifold, where base-pair conservation is represented in color-coded\n  format. In this mode no interaction are computed, but the -I option indicates\n  the location of the file containing interactions between two RNA\n  (alignments/sequence) from a previous run. If the -A option is not set a\n  structure figure a la RNAfold with color-coded annotation of the\n  accessibilities is returned\n\n",
  "  -L, --WindowLength=INT        Tells how large the region around the target\n                                  site should be for redrawing the alignment\n                                  interaction\n                                    (default=`1')",
  "  This option allows one to specify how large the region surrounding the target\n  site should be set when generating the alignment figure of the interaction\n\n",
  "\nIf in doubt our program is right, nature is at fault.\nComments should be sent to rna@tbi.univie.ac.at.\n\n",
    0
};

static void
init_help_array(void)
{
  RNAplex_args_info_help[0] = RNAplex_args_info_detailed_help[0];
  RNAplex_args_info_help[1] = RNAplex_args_info_detailed_help[1];
  RNAplex_args_info_help[2] = RNAplex_args_info_detailed_help[2];
  RNAplex_args_info_help[3] = RNAplex_args_info_detailed_help[3];
  RNAplex_args_info_help[4] = RNAplex_args_info_detailed_help[4];
  RNAplex_args_info_help[5] = RNAplex_args_info_detailed_help[5];
  RNAplex_args_info_help[6] = RNAplex_args_info_detailed_help[7];
  RNAplex_args_info_help[7] = RNAplex_args_info_detailed_help[9];
  RNAplex_args_info_help[8] = RNAplex_args_info_detailed_help[11];
  RNAplex_args_info_help[9] = RNAplex_args_info_detailed_help[13];
  RNAplex_args_info_help[10] = RNAplex_args_info_detailed_help[15];
  RNAplex_args_info_help[11] = RNAplex_args_info_detailed_help[16];
  RNAplex_args_info_help[12] = RNAplex_args_info_detailed_help[17];
  RNAplex_args_info_help[13] = RNAplex_args_info_detailed_help[18];
  RNAplex_args_info_help[14] = RNAplex_args_info_detailed_help[20];
  RNAplex_args_info_help[15] = RNAplex_args_info_detailed_help[22];
  RNAplex_args_info_help[16] = RNAplex_args_info_detailed_help[24];
  RNAplex_args_info_help[17] = RNAplex_args_info_detailed_help[25];
  RNAplex_args_info_help[18] = RNAplex_args_info_detailed_help[26];
  RNAplex_args_info_help[19] = RNAplex_args_info_detailed_help[27];
  RNAplex_args_info_help[20] = RNAplex_args_info_detailed_help[28];
  RNAplex_args_info_help[21] = RNAplex_args_info_detailed_help[29];
  RNAplex_args_info_help[22] = RNAplex_args_info_detailed_help[31];
  RNAplex_args_info_help[23] = RNAplex_args_info_detailed_help[33];
  RNAplex_args_info_help[24] = RNAplex_args_info_detailed_help[35];
  RNAplex_args_info_help[25] = RNAplex_args_info_detailed_help[37];
  RNAplex_args_info_help[26] = RNAplex_args_info_detailed_help[39];
  RNAplex_args_info_help[27] = RNAplex_args_info_detailed_help[40];
  RNAplex_args_info_help[28] = RNAplex_args_info_detailed_help[41];
  RNAplex_args_info_help[29] = RNAplex_args_info_detailed_help[43];
  RNAplex_args_info_help[30] = RNAplex_args_info_detailed_help[45];
  RNAplex_args_info_help[31] = RNAplex_args_info_detailed_help[47];
  RNAplex_args_info_help[32] = RNAplex_args_info_detailed_help[49];
  RNAplex_args_info_help[33] = 0; 
  
}

const char *RNAplex_args_info_help[34];

typedef enum {ARG_NO
  , ARG_FLAG
  , ARG_STRING
  , ARG_INT
  , ARG_DOUBLE
} RNAplex_cmdline_parser_arg_type;

static
void clear_given (struct RNAplex_args_info *args_info);
static
void clear_args (struct RNAplex_args_info *args_info);

static int
RNAplex_cmdline_parser_internal (int argc, char **argv, struct RNAplex_args_info *args_info,
                        struct RNAplex_cmdline_parser_params *params, const char *additional_error);


static char *
gengetopt_strdup (const char *s);

static
void clear_given (struct RNAplex_args_info *args_info)
{
  args_info->help_given = 0 ;
  args_info->detailed_help_given = 0 ;
  args_info->version_given = 0 ;
  args_info->query_given = 0 ;
  args_info->target_given = 0 ;
  args_info->accessibility_dir_given = 0 ;
  args_info->binary_given = 0 ;
  args_info->paramFile_given = 0 ;
  args_info->temp_given = 0 ;
  args_info->interaction_length_given = 0 ;
  args_info->extension_cost_given = 0 ;
  args_info->probe_mode_given = 0 ;
  args_info->probe_concentration_given = 0 ;
  args_info->na_concentration_given = 0 ;
  args_info->mg_concentration_given = 0 ;
  args_info->k_concentration_given = 0 ;
  args_info->tris_concentration_given = 0 ;
  args_info->fast_folding_given = 0 ;
  args_info->scale_accessibility_given = 0 ;
  args_info->constraint_given = 0 ;
  args_info->alignment_mode_given = 0 ;
  args_info->convert_to_bin_given = 0 ;
  args_info->duplex_distance_given = 0 ;
  args_info->energy_threshold_given = 0 ;
  args_info->produce_ps_given = 0 ;
  args_info->WindowLength_given = 0 ;
}

static
void clear_args (struct RNAplex_args_info *args_info)
{
  FIX_UNUSED (args_info);
  args_info->query_arg = NULL;
  args_info->query_orig = NULL;
  args_info->target_arg = NULL;
  args_info->target_orig = NULL;
  args_info->accessibility_dir_arg = NULL;
  args_info->accessibility_dir_orig = NULL;
  args_info->binary_flag = 0;
  args_info->paramFile_arg = NULL;
  args_info->paramFile_orig = NULL;
  args_info->temp_orig = NULL;
  args_info->interaction_length_arg = 40;
  args_info->interaction_length_orig = NULL;
  args_info->extension_cost_arg = 0;
  args_info->extension_cost_orig = NULL;
  args_info->probe_mode_flag = 0;
  args_info->probe_concentration_arg = 0.1;
  args_info->probe_concentration_orig = NULL;
  args_info->na_concentration_arg = 1.0;
  args_info->na_concentration_orig = NULL;
  args_info->mg_concentration_arg = 1.0;
  args_info->mg_concentration_orig = NULL;
  args_info->k_concentration_arg = 1.0;
  args_info->k_concentration_orig = NULL;
  args_info->tris_concentration_arg = 1.0;
  args_info->tris_concentration_orig = NULL;
  args_info->fast_folding_arg = 0;
  args_info->fast_folding_orig = NULL;
  args_info->scale_accessibility_arg = 1.0;
  args_info->scale_accessibility_orig = NULL;
  args_info->constraint_flag = 0;
  args_info->alignment_mode_flag = 0;
  args_info->convert_to_bin_flag = 0;
  args_info->duplex_distance_arg = 0;
  args_info->duplex_distance_orig = NULL;
  args_info->energy_threshold_arg = -100000;
  args_info->energy_threshold_orig = NULL;
  args_info->produce_ps_arg = NULL;
  args_info->produce_ps_orig = NULL;
  args_info->WindowLength_arg = 1;
  args_info->WindowLength_orig = NULL;
  
}

static
void init_args_info(struct RNAplex_args_info *args_info)
{

  init_help_array(); 
  args_info->help_help = RNAplex_args_info_detailed_help[0] ;
  args_info->detailed_help_help = RNAplex_args_info_detailed_help[1] ;
  args_info->version_help = RNAplex_args_info_detailed_help[2] ;
  args_info->query_help = RNAplex_args_info_detailed_help[5] ;
  args_info->target_help = RNAplex_args_info_detailed_help[7] ;
  args_info->accessibility_dir_help = RNAplex_args_info_detailed_help[9] ;
  args_info->binary_help = RNAplex_args_info_detailed_help[11] ;
  args_info->paramFile_help = RNAplex_args_info_detailed_help[13] ;
  args_info->temp_help = RNAplex_args_info_detailed_help[17] ;
  args_info->interaction_length_help = RNAplex_args_info_detailed_help[18] ;
  args_info->extension_cost_help = RNAplex_args_info_detailed_help[20] ;
  args_info->probe_mode_help = RNAplex_args_info_detailed_help[22] ;
  args_info->probe_concentration_help = RNAplex_args_info_detailed_help[24] ;
  args_info->na_concentration_help = RNAplex_args_info_detailed_help[25] ;
  args_info->mg_concentration_help = RNAplex_args_info_detailed_help[26] ;
  args_info->k_concentration_help = RNAplex_args_info_detailed_help[27] ;
  args_info->tris_concentration_help = RNAplex_args_info_detailed_help[28] ;
  args_info->fast_folding_help = RNAplex_args_info_detailed_help[29] ;
  args_info->scale_accessibility_help = RNAplex_args_info_detailed_help[31] ;
  args_info->constraint_help = RNAplex_args_info_detailed_help[33] ;
  args_info->alignment_mode_help = RNAplex_args_info_detailed_help[35] ;
  args_info->convert_to_bin_help = RNAplex_args_info_detailed_help[37] ;
  args_info->duplex_distance_help = RNAplex_args_info_detailed_help[41] ;
  args_info->energy_threshold_help = RNAplex_args_info_detailed_help[43] ;
  args_info->produce_ps_help = RNAplex_args_info_detailed_help[45] ;
  args_info->WindowLength_help = RNAplex_args_info_detailed_help[47] ;
  
}

void
RNAplex_cmdline_parser_print_version (void)
{
  printf ("%s %s\n",
     (strlen(RNAPLEX_CMDLINE_PARSER_PACKAGE_NAME) ? RNAPLEX_CMDLINE_PARSER_PACKAGE_NAME : RNAPLEX_CMDLINE_PARSER_PACKAGE),
     RNAPLEX_CMDLINE_PARSER_VERSION);

  if (strlen(RNAplex_args_info_versiontext) > 0)
    printf("\n%s\n", RNAplex_args_info_versiontext);
}

static void print_help_common(void)
{
	size_t len_purpose = strlen(RNAplex_args_info_purpose);
	size_t len_usage = strlen(RNAplex_args_info_usage);

	if (len_usage > 0) {
		printf("%s\n", RNAplex_args_info_usage);
	}
	if (len_purpose > 0) {
		printf("%s\n", RNAplex_args_info_purpose);
	}

	if (len_usage || len_purpose) {
		printf("\n");
	}

	if (strlen(RNAplex_args_info_description) > 0) {
		printf("%s\n\n", RNAplex_args_info_description);
	}
}

void
RNAplex_cmdline_parser_print_help (void)
{
  int i = 0;
  print_help_common();
  while (RNAplex_args_info_help[i])
    printf("%s\n", RNAplex_args_info_help[i++]);
}

void
RNAplex_cmdline_parser_print_detailed_help (void)
{
  int i = 0;
  print_help_common();
  while (RNAplex_args_info_detailed_help[i])
    printf("%s\n", RNAplex_args_info_detailed_help[i++]);
}

void
RNAplex_cmdline_parser_init (struct RNAplex_args_info *args_info)
{
  clear_given (args_info);
  clear_args (args_info);
  init_args_info (args_info);
}

void
RNAplex_cmdline_parser_params_init(struct RNAplex_cmdline_parser_params *params)
{
  if (params)
    { 
      params->override = 0;
      params->initialize = 1;
      params->check_required = 1;
      params->check_ambiguity = 0;
      params->print_errors = 1;
    }
}

struct RNAplex_cmdline_parser_params *
RNAplex_cmdline_parser_params_create(void)
{
  struct RNAplex_cmdline_parser_params *params = 
    (struct RNAplex_cmdline_parser_params *)malloc(sizeof(struct RNAplex_cmdline_parser_params));
  RNAplex_cmdline_parser_params_init(params);  
  return params;
}

static void
free_string_field (char **s)
{
  if (*s)
    {
      free (*s);
      *s = 0;
    }
}


static void
RNAplex_cmdline_parser_release (struct RNAplex_args_info *args_info)
{

  free_string_field (&(args_info->query_arg));
  free_string_field (&(args_info->query_orig));
  free_string_field (&(args_info->target_arg));
  free_string_field (&(args_info->target_orig));
  free_string_field (&(args_info->accessibility_dir_arg));
  free_string_field (&(args_info->accessibility_dir_orig));
  free_string_field (&(args_info->paramFile_arg));
  free_string_field (&(args_info->paramFile_orig));
  free_string_field (&(args_info->temp_orig));
  free_string_field (&(args_info->interaction_length_orig));
  free_string_field (&(args_info->extension_cost_orig));
  free_string_field (&(args_info->probe_concentration_orig));
  free_string_field (&(args_info->na_concentration_orig));
  free_string_field (&(args_info->mg_concentration_orig));
  free_string_field (&(args_info->k_concentration_orig));
  free_string_field (&(args_info->tris_concentration_orig));
  free_string_field (&(args_info->fast_folding_orig));
  free_string_field (&(args_info->scale_accessibility_orig));
  free_string_field (&(args_info->duplex_distance_orig));
  free_string_field (&(args_info->energy_threshold_orig));
  free_string_field (&(args_info->produce_ps_arg));
  free_string_field (&(args_info->produce_ps_orig));
  free_string_field (&(args_info->WindowLength_orig));
  
  

  clear_given (args_info);
}


static void
write_into_file(FILE *outfile, const char *opt, const char *arg, const char *values[])
{
  FIX_UNUSED (values);
  if (arg) {
    fprintf(outfile, "%s=\"%s\"\n", opt, arg);
  } else {
    fprintf(outfile, "%s\n", opt);
  }
}


int
RNAplex_cmdline_parser_dump(FILE *outfile, struct RNAplex_args_info *args_info)
{
  int i = 0;

  if (!outfile)
    {
      fprintf (stderr, "%s: cannot dump options to stream\n", RNAPLEX_CMDLINE_PARSER_PACKAGE);
      return EXIT_FAILURE;
    }

  if (args_info->help_given)
    write_into_file(outfile, "help", 0, 0 );
  if (args_info->detailed_help_given)
    write_into_file(outfile, "detailed-help", 0, 0 );
  if (args_info->version_given)
    write_into_file(outfile, "version", 0, 0 );
  if (args_info->query_given)
    write_into_file(outfile, "query", args_info->query_orig, 0);
  if (args_info->target_given)
    write_into_file(outfile, "target", args_info->target_orig, 0);
  if (args_info->accessibility_dir_given)
    write_into_file(outfile, "accessibility-dir", args_info->accessibility_dir_orig, 0);
  if (args_info->binary_given)
    write_into_file(outfile, "binary", 0, 0 );
  if (args_info->paramFile_given)
    write_into_file(outfile, "paramFile", args_info->paramFile_orig, 0);
  if (args_info->temp_given)
    write_into_file(outfile, "temp", args_info->temp_orig, 0);
  if (args_info->interaction_length_given)
    write_into_file(outfile, "interaction-length", args_info->interaction_length_orig, 0);
  if (args_info->extension_cost_given)
    write_into_file(outfile, "extension-cost", args_info->extension_cost_orig, 0);
  if (args_info->probe_mode_given)
    write_into_file(outfile, "probe-mode", 0, 0 );
  if (args_info->probe_concentration_given)
    write_into_file(outfile, "probe-concentration", args_info->probe_concentration_orig, 0);
  if (args_info->na_concentration_given)
    write_into_file(outfile, "na-concentration", args_info->na_concentration_orig, 0);
  if (args_info->mg_concentration_given)
    write_into_file(outfile, "mg-concentration", args_info->mg_concentration_orig, 0);
  if (args_info->k_concentration_given)
    write_into_file(outfile, "k-concentration", args_info->k_concentration_orig, 0);
  if (args_info->tris_concentration_given)
    write_into_file(outfile, "tris-concentration", args_info->tris_concentration_orig, 0);
  if (args_info->fast_folding_given)
    write_into_file(outfile, "fast-folding", args_info->fast_folding_orig, 0);
  if (args_info->scale_accessibility_given)
    write_into_file(outfile, "scale-accessibility", args_info->scale_accessibility_orig, 0);
  if (args_info->constraint_given)
    write_into_file(outfile, "constraint", 0, 0 );
  if (args_info->alignment_mode_given)
    write_into_file(outfile, "alignment-mode", 0, 0 );
  if (args_info->convert_to_bin_given)
    write_into_file(outfile, "convert-to-bin", 0, 0 );
  if (args_info->duplex_distance_given)
    write_into_file(outfile, "duplex-distance", args_info->duplex_distance_orig, 0);
  if (args_info->energy_threshold_given)
    write_into_file(outfile, "energy-threshold", args_info->energy_threshold_orig, 0);
  if (args_info->produce_ps_given)
    write_into_file(outfile, "produce-ps", args_info->produce_ps_orig, 0);
  if (args_info->WindowLength_given)
    write_into_file(outfile, "WindowLength", args_info->WindowLength_orig, 0);
  

  i = EXIT_SUCCESS;
  return i;
}

int
RNAplex_cmdline_parser_file_save(const char *filename, struct RNAplex_args_info *args_info)
{
  FILE *outfile;
  int i = 0;

  outfile = fopen(filename, "w");

  if (!outfile)
    {
      fprintf (stderr, "%s: cannot open file for writing: %s\n", RNAPLEX_CMDLINE_PARSER_PACKAGE, filename);
      return EXIT_FAILURE;
    }

  i = RNAplex_cmdline_parser_dump(outfile, args_info);
  fclose (outfile);

  return i;
}

void
RNAplex_cmdline_parser_free (struct RNAplex_args_info *args_info)
{
  RNAplex_cmdline_parser_release (args_info);
}

/** @brief replacement of strdup, which is not standard */
char *
gengetopt_strdup (const char *s)
{
  char *result = 0;
  if (!s)
    return result;

  result = (char*)malloc(strlen(s) + 1);
  if (result == (char*)0)
    return (char*)0;
  strcpy(result, s);
  return result;
}

int
RNAplex_cmdline_parser (int argc, char **argv, struct RNAplex_args_info *args_info)
{
  return RNAplex_cmdline_parser2 (argc, argv, args_info, 0, 1, 1);
}

int
RNAplex_cmdline_parser_ext (int argc, char **argv, struct RNAplex_args_info *args_info,
                   struct RNAplex_cmdline_parser_params *params)
{
  int result;
  result = RNAplex_cmdline_parser_internal (argc, argv, args_info, params, 0);

  if (result == EXIT_FAILURE)
    {
      RNAplex_cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}

int
RNAplex_cmdline_parser2 (int argc, char **argv, struct RNAplex_args_info *args_info, int override, int initialize, int check_required)
{
  int result;
  struct RNAplex_cmdline_parser_params params;
  
  params.override = override;
  params.initialize = initialize;
  params.check_required = check_required;
  params.check_ambiguity = 0;
  params.print_errors = 1;

  result = RNAplex_cmdline_parser_internal (argc, argv, args_info, &params, 0);

  if (result == EXIT_FAILURE)
    {
      RNAplex_cmdline_parser_free (args_info);
      exit (EXIT_FAILURE);
    }
  
  return result;
}

int
RNAplex_cmdline_parser_required (struct RNAplex_args_info *args_info, const char *prog_name)
{
  FIX_UNUSED (args_info);
  FIX_UNUSED (prog_name);
  return EXIT_SUCCESS;
}

/*
 * Extracted from the glibc source tree, version 2.3.6
 *
 * Licensed under the GPL as per the whole glibc source tree.
 *
 * This file was modified so that getopt_long can be called
 * many times without risking previous memory to be spoiled.
 *
 * Modified by Andre Noll and Lorenzo Bettini for use in
 * GNU gengetopt generated files.
 *
 */

/* 
 * we must include anything we need since this file is not thought to be
 * inserted in a file already using getopt.h
 *
 * Lorenzo
 */

struct option
{
  const char *name;
  /* has_arg can't be an enum because some compilers complain about
     type mismatches in all the code that assumes it is an int.  */
  int has_arg;
  int *flag;
  int val;
};

/* This version of `getopt' appears to the caller like standard Unix `getopt'
   but it behaves differently for the user, since it allows the user
   to intersperse the options with the other arguments.

   As `getopt' works, it permutes the elements of ARGV so that,
   when it is done, all the options precede everything else.  Thus
   all application programs are extended to handle flexible argument order.
*/
/*
   If the field `flag' is not NULL, it points to a variable that is set
   to the value given in the field `val' when the option is found, but
   left unchanged if the option is not found.

   To have a long-named option do something other than set an `int' to
   a compiled-in constant, such as set a value from `custom_optarg', set the
   option's `flag' field to zero and its `val' field to a nonzero
   value (the equivalent single-letter option character, if there is
   one).  For long options that have a zero `flag' field, `getopt'
   returns the contents of the `val' field.  */

/* Names for the values of the `has_arg' field of `struct option'.  */
#ifndef no_argument
#define no_argument		0
#endif

#ifndef required_argument
#define required_argument	1
#endif

#ifndef optional_argument
#define optional_argument	2
#endif

struct custom_getopt_data {
	/*
	 * These have exactly the same meaning as the corresponding global variables,
	 * except that they are used for the reentrant versions of getopt.
	 */
	int custom_optind;
	int custom_opterr;
	int custom_optopt;
	char *custom_optarg;

	/* True if the internal members have been initialized.  */
	int initialized;

	/*
	 * The next char to be scanned in the option-element in which the last option
	 * character we returned was found.  This allows us to pick up the scan where
	 * we left off.  If this is zero, or a null string, it means resume the scan by
	 * advancing to the next ARGV-element.
	 */
	char *nextchar;

	/*
	 * Describe the part of ARGV that contains non-options that have been skipped.
	 * `first_nonopt' is the index in ARGV of the first of them; `last_nonopt' is
	 * the index after the last of them.
	 */
	int first_nonopt;
	int last_nonopt;
};

/*
 * the variables optarg, optind, opterr and optopt are renamed with
 * the custom_ prefix so that they don't interfere with getopt ones.
 *
 * Moreover they're static so they are visible only from within the
 * file where this very file will be included.
 */

/*
 * For communication from `custom_getopt' to the caller.  When `custom_getopt' finds an
 * option that takes an argument, the argument value is returned here.
 */
static char *custom_optarg;

/*
 * Index in ARGV of the next element to be scanned.  This is used for
 * communication to and from the caller and for communication between
 * successive calls to `custom_getopt'.
 *
 * On entry to `custom_getopt', 1 means this is the first call; initialize.
 *
 * When `custom_getopt' returns -1, this is the index of the first of the non-option
 * elements that the caller should itself scan.
 *
 * Otherwise, `custom_optind' communicates from one call to the next how much of ARGV
 * has been scanned so far.
 *
 * 1003.2 says this must be 1 before any call.
 */
static int custom_optind = 1;

/*
 * Callers store zero here to inhibit the error message for unrecognized
 * options.
 */
static int custom_opterr = 1;

/*
 * Set to an option character which was unrecognized.  This must be initialized
 * on some systems to avoid linking in the system's own getopt implementation.
 */
static int custom_optopt = '?';

/*
 * Exchange two adjacent subsequences of ARGV.  One subsequence is elements
 * [first_nonopt,last_nonopt) which contains all the non-options that have been
 * skipped so far.  The other is elements [last_nonopt,custom_optind), which contains
 * all the options processed since those non-options were skipped.
 * `first_nonopt' and `last_nonopt' are relocated so that they describe the new
 * indices of the non-options in ARGV after they are moved.
 */
static void exchange(char **argv, struct custom_getopt_data *d)
{
	int bottom = d->first_nonopt;
	int middle = d->last_nonopt;
	int top = d->custom_optind;
	char *tem;

	/*
	 * Exchange the shorter segment with the far end of the longer segment.
	 * That puts the shorter segment into the right place.  It leaves the
	 * longer segment in the right place overall, but it consists of two
	 * parts that need to be swapped next.
	 */
	while (top > middle && middle > bottom) {
		if (top - middle > middle - bottom) {
			/* Bottom segment is the short one.  */
			int len = middle - bottom;
			int i;

			/* Swap it with the top part of the top segment.  */
			for (i = 0; i < len; i++) {
				tem = argv[bottom + i];
				argv[bottom + i] =
					argv[top - (middle - bottom) + i];
				argv[top - (middle - bottom) + i] = tem;
			}
			/* Exclude the moved bottom segment from further swapping.  */
			top -= len;
		} else {
			/* Top segment is the short one.  */
			int len = top - middle;
			int i;

			/* Swap it with the bottom part of the bottom segment.  */
			for (i = 0; i < len; i++) {
				tem = argv[bottom + i];
				argv[bottom + i] = argv[middle + i];
				argv[middle + i] = tem;
			}
			/* Exclude the moved top segment from further swapping.  */
			bottom += len;
		}
	}
	/* Update records for the slots the non-options now occupy.  */
	d->first_nonopt += (d->custom_optind - d->last_nonopt);
	d->last_nonopt = d->custom_optind;
}

/* Initialize the internal data when the first call is made.  */
static void custom_getopt_initialize(struct custom_getopt_data *d)
{
	/*
	 * Start processing options with ARGV-element 1 (since ARGV-element 0
	 * is the program name); the sequence of previously skipped non-option
	 * ARGV-elements is empty.
	 */
	d->first_nonopt = d->last_nonopt = d->custom_optind;
	d->nextchar = NULL;
	d->initialized = 1;
}

#define NONOPTION_P (argv[d->custom_optind][0] != '-' || argv[d->custom_optind][1] == '\0')

/* return: zero: continue, nonzero: return given value to user */
static int shuffle_argv(int argc, char *const *argv,const struct option *longopts,
	struct custom_getopt_data *d)
{
	/*
	 * Give FIRST_NONOPT & LAST_NONOPT rational values if CUSTOM_OPTIND has been
	 * moved back by the user (who may also have changed the arguments).
	 */
	if (d->last_nonopt > d->custom_optind)
		d->last_nonopt = d->custom_optind;
	if (d->first_nonopt > d->custom_optind)
		d->first_nonopt = d->custom_optind;
	/*
	 * If we have just processed some options following some
	 * non-options, exchange them so that the options come first.
	 */
	if (d->first_nonopt != d->last_nonopt &&
			d->last_nonopt != d->custom_optind)
		exchange((char **) argv, d);
	else if (d->last_nonopt != d->custom_optind)
		d->first_nonopt = d->custom_optind;
	/*
	 * Skip any additional non-options and extend the range of
	 * non-options previously skipped.
	 */
	while (d->custom_optind < argc && NONOPTION_P)
		d->custom_optind++;
	d->last_nonopt = d->custom_optind;
	/*
	 * The special ARGV-element `--' means premature end of options.  Skip
	 * it like a null option, then exchange with previous non-options as if
	 * it were an option, then skip everything else like a non-option.
	 */
	if (d->custom_optind != argc && !strcmp(argv[d->custom_optind], "--")) {
		d->custom_optind++;
		if (d->first_nonopt != d->last_nonopt
				&& d->last_nonopt != d->custom_optind)
			exchange((char **) argv, d);
		else if (d->first_nonopt == d->last_nonopt)
			d->first_nonopt = d->custom_optind;
		d->last_nonopt = argc;
		d->custom_optind = argc;
	}
	/*
	 * If we have done all the ARGV-elements, stop the scan and back over
	 * any non-options that we skipped and permuted.
	 */
	if (d->custom_optind == argc) {
		/*
		 * Set the next-arg-index to point at the non-options that we
		 * previously skipped, so the caller will digest them.
		 */
		if (d->first_nonopt != d->last_nonopt)
			d->custom_optind = d->first_nonopt;
		return -1;
	}
	/*
	 * If we have come to a non-option and did not permute it, either stop
	 * the scan or describe it to the caller and pass it by.
	 */
	if (NONOPTION_P) {
		d->custom_optarg = argv[d->custom_optind++];
		return 1;
	}
	/*
	 * We have found another option-ARGV-element. Skip the initial
	 * punctuation.
	 */
	d->nextchar = (argv[d->custom_optind] + 1 + (longopts != NULL && argv[d->custom_optind][1] == '-'));
	return 0;
}

/*
 * Check whether the ARGV-element is a long option.
 *
 * If there's a long option "fubar" and the ARGV-element is "-fu", consider
 * that an abbreviation of the long option, just like "--fu", and not "-f" with
 * arg "u".
 *
 * This distinction seems to be the most useful approach.
 *
 */
static int check_long_opt(int argc, char *const *argv, const char *optstring,
		const struct option *longopts, int *longind,
		int print_errors, struct custom_getopt_data *d)
{
	char *nameend;
	const struct option *p;
	const struct option *pfound = NULL;
	int exact = 0;
	int ambig = 0;
	int indfound = -1;
	int option_index;

	for (nameend = d->nextchar; *nameend && *nameend != '='; nameend++)
		/* Do nothing.  */ ;

	/* Test all long options for either exact match or abbreviated matches */
	for (p = longopts, option_index = 0; p->name; p++, option_index++)
		if (!strncmp(p->name, d->nextchar, nameend - d->nextchar)) {
			if ((unsigned int) (nameend - d->nextchar)
					== (unsigned int) strlen(p->name)) {
				/* Exact match found.  */
				pfound = p;
				indfound = option_index;
				exact = 1;
				break;
			} else if (pfound == NULL) {
				/* First nonexact match found.  */
				pfound = p;
				indfound = option_index;
			} else if (pfound->has_arg != p->has_arg
					|| pfound->flag != p->flag
					|| pfound->val != p->val)
				/* Second or later nonexact match found.  */
				ambig = 1;
		}
	if (ambig && !exact) {
		if (print_errors) {
			fprintf(stderr,
				"%s: option `%s' is ambiguous\n",
				argv[0], argv[d->custom_optind]);
		}
		d->nextchar += strlen(d->nextchar);
		d->custom_optind++;
		d->custom_optopt = 0;
		return '?';
	}
	if (pfound) {
		option_index = indfound;
		d->custom_optind++;
		if (*nameend) {
			if (pfound->has_arg != no_argument)
				d->custom_optarg = nameend + 1;
			else {
				if (print_errors) {
					if (argv[d->custom_optind - 1][1] == '-') {
						/* --option */
						fprintf(stderr, "%s: option `--%s' doesn't allow an argument\n",
							argv[0], pfound->name);
					} else {
						/* +option or -option */
						fprintf(stderr, "%s: option `%c%s' doesn't allow an argument\n",
							argv[0], argv[d->custom_optind - 1][0], pfound->name);
					}

				}
				d->nextchar += strlen(d->nextchar);
				d->custom_optopt = pfound->val;
				return '?';
			}
		} else if (pfound->has_arg == required_argument) {
			if (d->custom_optind < argc)
				d->custom_optarg = argv[d->custom_optind++];
			else {
				if (print_errors) {
					fprintf(stderr,
						"%s: option `%s' requires an argument\n",
						argv[0],
						argv[d->custom_optind - 1]);
				}
				d->nextchar += strlen(d->nextchar);
				d->custom_optopt = pfound->val;
				return optstring[0] == ':' ? ':' : '?';
			}
		}
		d->nextchar += strlen(d->nextchar);
		if (longind != NULL)
			*longind = option_index;
		if (pfound->flag) {
			*(pfound->flag) = pfound->val;
			return 0;
		}
		return pfound->val;
	}
	/*
	 * Can't find it as a long option.  If this is not getopt_long_only, or
	 * the option starts with '--' or is not a valid short option, then
	 * it's an error.  Otherwise interpret it as a short option.
	 */
	if (print_errors) {
		if (argv[d->custom_optind][1] == '-') {
			/* --option */
			fprintf(stderr,
				"%s: unrecognized option `--%s'\n",
				argv[0], d->nextchar);
		} else {
			/* +option or -option */
			fprintf(stderr,
				"%s: unrecognized option `%c%s'\n",
				argv[0], argv[d->custom_optind][0],
				d->nextchar);
		}
	}
	d->nextchar = (char *) "";
	d->custom_optind++;
	d->custom_optopt = 0;
	return '?';
}

static int check_short_opt(int argc, char *const *argv, const char *optstring,
		int print_errors, struct custom_getopt_data *d)
{
	char c = *d->nextchar++;
	const char *temp = strchr(optstring, c);

	/* Increment `custom_optind' when we start to process its last character.  */
	if (*d->nextchar == '\0')
		++d->custom_optind;
	if (!temp || c == ':') {
		if (print_errors)
			fprintf(stderr, "%s: invalid option -- %c\n", argv[0], c);

		d->custom_optopt = c;
		return '?';
	}
	if (temp[1] == ':') {
		if (temp[2] == ':') {
			/* This is an option that accepts an argument optionally.  */
			if (*d->nextchar != '\0') {
				d->custom_optarg = d->nextchar;
				d->custom_optind++;
			} else
				d->custom_optarg = NULL;
			d->nextchar = NULL;
		} else {
			/* This is an option that requires an argument.  */
			if (*d->nextchar != '\0') {
				d->custom_optarg = d->nextchar;
				/*
				 * If we end this ARGV-element by taking the
				 * rest as an arg, we must advance to the next
				 * element now.
				 */
				d->custom_optind++;
			} else if (d->custom_optind == argc) {
				if (print_errors) {
					fprintf(stderr,
						"%s: option requires an argument -- %c\n",
						argv[0], c);
				}
				d->custom_optopt = c;
				if (optstring[0] == ':')
					c = ':';
				else
					c = '?';
			} else
				/*
				 * We already incremented `custom_optind' once;
				 * increment it again when taking next ARGV-elt
				 * as argument.
				 */
				d->custom_optarg = argv[d->custom_optind++];
			d->nextchar = NULL;
		}
	}
	return c;
}

/*
 * Scan elements of ARGV for option characters given in OPTSTRING.
 *
 * If an element of ARGV starts with '-', and is not exactly "-" or "--",
 * then it is an option element.  The characters of this element
 * (aside from the initial '-') are option characters.  If `getopt'
 * is called repeatedly, it returns successively each of the option characters
 * from each of the option elements.
 *
 * If `getopt' finds another option character, it returns that character,
 * updating `custom_optind' and `nextchar' so that the next call to `getopt' can
 * resume the scan with the following option character or ARGV-element.
 *
 * If there are no more option characters, `getopt' returns -1.
 * Then `custom_optind' is the index in ARGV of the first ARGV-element
 * that is not an option.  (The ARGV-elements have been permuted
 * so that those that are not options now come last.)
 *
 * OPTSTRING is a string containing the legitimate option characters.
 * If an option character is seen that is not listed in OPTSTRING,
 * return '?' after printing an error message.  If you set `custom_opterr' to
 * zero, the error message is suppressed but we still return '?'.
 *
 * If a char in OPTSTRING is followed by a colon, that means it wants an arg,
 * so the following text in the same ARGV-element, or the text of the following
 * ARGV-element, is returned in `custom_optarg'.  Two colons mean an option that
 * wants an optional arg; if there is text in the current ARGV-element,
 * it is returned in `custom_optarg', otherwise `custom_optarg' is set to zero.
 *
 * If OPTSTRING starts with `-' or `+', it requests different methods of
 * handling the non-option ARGV-elements.
 * See the comments about RETURN_IN_ORDER and REQUIRE_ORDER, above.
 *
 * Long-named options begin with `--' instead of `-'.
 * Their names may be abbreviated as long as the abbreviation is unique
 * or is an exact match for some defined option.  If they have an
 * argument, it follows the option name in the same ARGV-element, separated
 * from the option name by a `=', or else the in next ARGV-element.
 * When `getopt' finds a long-named option, it returns 0 if that option's
 * `flag' field is nonzero, the value of the option's `val' field
 * if the `flag' field is zero.
 *
 * The elements of ARGV aren't really const, because we permute them.
 * But we pretend they're const in the prototype to be compatible
 * with other systems.
 *
 * LONGOPTS is a vector of `struct option' terminated by an
 * element containing a name which is zero.
 *
 * LONGIND returns the index in LONGOPT of the long-named option found.
 * It is only valid when a long-named option has been found by the most
 * recent call.
 *
 * Return the option character from OPTS just read.  Return -1 when there are
 * no more options.  For unrecognized options, or options missing arguments,
 * `custom_optopt' is set to the option letter, and '?' is returned.
 *
 * The OPTS string is a list of characters which are recognized option letters,
 * optionally followed by colons, specifying that that letter takes an
 * argument, to be placed in `custom_optarg'.
 *
 * If a letter in OPTS is followed by two colons, its argument is optional.
 * This behavior is specific to the GNU `getopt'.
 *
 * The argument `--' causes premature termination of argument scanning,
 * explicitly telling `getopt' that there are no more options.  If OPTS begins
 * with `--', then non-option arguments are treated as arguments to the option
 * '\0'.  This behavior is specific to the GNU `getopt'.
 */

static int getopt_internal_r(int argc, char *const *argv, const char *optstring,
		const struct option *longopts, int *longind,
		struct custom_getopt_data *d)
{
	int ret, print_errors = d->custom_opterr;

	if (optstring[0] == ':')
		print_errors = 0;
	if (argc < 1)
		return -1;
	d->custom_optarg = NULL;

	/* 
	 * This is a big difference with GNU getopt, since optind == 0
	 * means initialization while here 1 means first call.
	 */
	if (d->custom_optind == 0 || !d->initialized) {
		if (d->custom_optind == 0)
			d->custom_optind = 1;	/* Don't scan ARGV[0], the program name.  */
		custom_getopt_initialize(d);
	}
	if (d->nextchar == NULL || *d->nextchar == '\0') {
		ret = shuffle_argv(argc, argv, longopts, d);
		if (ret)
			return ret;
	}
	if (longopts && (argv[d->custom_optind][1] == '-' ))
		return check_long_opt(argc, argv, optstring, longopts,
			longind, print_errors, d);
	return check_short_opt(argc, argv, optstring, print_errors, d);
}

static int custom_getopt_internal(int argc, char *const *argv, const char *optstring,
	const struct option *longopts, int *longind)
{
	int result;
	/* Keep a global copy of all internal members of d */
	static struct custom_getopt_data d;

	d.custom_optind = custom_optind;
	d.custom_opterr = custom_opterr;
	result = getopt_internal_r(argc, argv, optstring, longopts,
		longind, &d);
	custom_optind = d.custom_optind;
	custom_optarg = d.custom_optarg;
	custom_optopt = d.custom_optopt;
	return result;
}

static int custom_getopt_long (int argc, char *const *argv, const char *options,
	const struct option *long_options, int *opt_index)
{
	return custom_getopt_internal(argc, argv, options, long_options,
		opt_index);
}


static char *package_name = 0;

/**
 * @brief updates an option
 * @param field the generic pointer to the field to update
 * @param orig_field the pointer to the orig field
 * @param field_given the pointer to the number of occurrence of this option
 * @param prev_given the pointer to the number of occurrence already seen
 * @param value the argument for this option (if null no arg was specified)
 * @param possible_values the possible values for this option (if specified)
 * @param default_value the default value (in case the option only accepts fixed values)
 * @param arg_type the type of this option
 * @param check_ambiguity @see RNAplex_cmdline_parser_params.check_ambiguity
 * @param override @see RNAplex_cmdline_parser_params.override
 * @param no_free whether to free a possible previous value
 * @param multiple_option whether this is a multiple option
 * @param long_opt the corresponding long option
 * @param short_opt the corresponding short option (or '-' if none)
 * @param additional_error possible further error specification
 */
static
int update_arg(void *field, char **orig_field,
               unsigned int *field_given, unsigned int *prev_given, 
               char *value, const char *possible_values[],
               const char *default_value,
               RNAplex_cmdline_parser_arg_type arg_type,
               int check_ambiguity, int override,
               int no_free, int multiple_option,
               const char *long_opt, char short_opt,
               const char *additional_error)
{
  char *stop_char = 0;
  const char *val = value;
  int found;
  char **string_field;
  FIX_UNUSED (field);

  stop_char = 0;
  found = 0;

  if (!multiple_option && prev_given && (*prev_given || (check_ambiguity && *field_given)))
    {
      if (short_opt != '-')
        fprintf (stderr, "%s: `--%s' (`-%c') option given more than once%s\n", 
               package_name, long_opt, short_opt,
               (additional_error ? additional_error : ""));
      else
        fprintf (stderr, "%s: `--%s' option given more than once%s\n", 
               package_name, long_opt,
               (additional_error ? additional_error : ""));
      return 1; /* failure */
    }

  FIX_UNUSED (default_value);
    
  if (field_given && *field_given && ! override)
    return 0;
  if (prev_given)
    (*prev_given)++;
  if (field_given)
    (*field_given)++;
  if (possible_values)
    val = possible_values[found];

  switch(arg_type) {
  case ARG_FLAG:
    *((int *)field) = !*((int *)field);
    break;
  case ARG_INT:
    if (val) *((int *)field) = strtol (val, &stop_char, 0);
    break;
  case ARG_DOUBLE:
    if (val) *((double *)field) = strtod (val, &stop_char);
    break;
  case ARG_STRING:
    if (val) {
      string_field = (char **)field;
      if (!no_free && *string_field)
        free (*string_field); /* free previous string */
      *string_field = gengetopt_strdup (val);
    }
    break;
  default:
    break;
  };

  /* check numeric conversion */
  switch(arg_type) {
  case ARG_INT:
  case ARG_DOUBLE:
    if (val && !(stop_char && *stop_char == '\0')) {
      fprintf(stderr, "%s: invalid numeric value: %s\n", package_name, val);
      return 1; /* failure */
    }
    break;
  default:
    ;
  };

  /* store the original value */
  switch(arg_type) {
  case ARG_NO:
  case ARG_FLAG:
    break;
  default:
    if (value && orig_field) {
      if (no_free) {
        *orig_field = value;
      } else {
        if (*orig_field)
          free (*orig_field); /* free previous string */
        *orig_field = gengetopt_strdup (value);
      }
    }
  };

  return 0; /* OK */
}


int
RNAplex_cmdline_parser_internal (
  int argc, char **argv, struct RNAplex_args_info *args_info,
                        struct RNAplex_cmdline_parser_params *params, const char *additional_error)
{
  int c;	/* Character of the parsed option.  */

  int error_occurred = 0;
  struct RNAplex_args_info local_args_info;
  
  int override;
  int initialize;
  int check_required;
  int check_ambiguity;

  char *optarg;
  int optind;
  int opterr;
  int optopt;
  
  package_name = argv[0];
  
  /* TODO: Why is this here? It is not used anywhere. */
  override = params->override;
  FIX_UNUSED(override);

  initialize = params->initialize;
  check_required = params->check_required;

  /* TODO: Why is this here? It is not used anywhere. */
  check_ambiguity = params->check_ambiguity;
  FIX_UNUSED(check_ambiguity);

  if (initialize)
    RNAplex_cmdline_parser_init (args_info);

  RNAplex_cmdline_parser_init (&local_args_info);

  optarg = 0;
  optind = 0;
  opterr = params->print_errors;
  optopt = '?';

  while (1)
    {
      int option_index = 0;

      static struct option long_options[] = {
        { "help",	0, NULL, 'h' },
        { "detailed-help",	0, NULL, 0 },
        { "version",	0, NULL, 0 },
        { "query",	1, NULL, 'q' },
        { "target",	1, NULL, 't' },
        { "accessibility-dir",	1, NULL, 'a' },
        { "binary",	0, NULL, 'b' },
        { "paramFile",	1, NULL, 'P' },
        { "temp",	1, NULL, 'T' },
        { "interaction-length",	1, NULL, 'l' },
        { "extension-cost",	1, NULL, 'c' },
        { "probe-mode",	0, NULL, 'p' },
        { "probe-concentration",	1, NULL, 'Q' },
        { "na-concentration",	1, NULL, 'N' },
        { "mg-concentration",	1, NULL, 'M' },
        { "k-concentration",	1, NULL, 'K' },
        { "tris-concentration",	1, NULL, 'U' },
        { "fast-folding",	1, NULL, 'f' },
        { "scale-accessibility",	1, NULL, 'V' },
        { "constraint",	0, NULL, 'C' },
        { "alignment-mode",	0, NULL, 'A' },
        { "convert-to-bin",	0, NULL, 'k' },
        { "duplex-distance",	1, NULL, 'z' },
        { "energy-threshold",	1, NULL, 'e' },
        { "produce-ps",	1, NULL, 'I' },
        { "WindowLength",	1, NULL, 'L' },
        { 0,  0, 0, 0 }
      };

      custom_optarg = optarg;
      custom_optind = optind;
      custom_opterr = opterr;
      custom_optopt = optopt;

      c = custom_getopt_long (argc, argv, "hq:t:a:bP:T:l:c:pQ:N:M:K:U:f:V:CAkz:e:I:L:", long_options, &option_index);

      optarg = custom_optarg;
      optind = custom_optind;
      opterr = custom_opterr;
      optopt = custom_optopt;

      if (c == -1) break;	/* Exit from `while (1)' loop.  */

      switch (c)
        {
        case 'h':	/* Print help and exit.  */
          RNAplex_cmdline_parser_print_help ();
          RNAplex_cmdline_parser_free (&local_args_info);
          exit (EXIT_SUCCESS);

        case 'q':	/* File containing the query sequence.
.  */
        
        
          if (update_arg( (void *)&(args_info->query_arg), 
               &(args_info->query_orig), &(args_info->query_given),
              &(local_args_info.query_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "query", 'q',
              additional_error))
            goto failure;
        
          break;
        case 't':	/* File containing the target sequence.
.  */
        
        
          if (update_arg( (void *)&(args_info->target_arg), 
               &(args_info->target_orig), &(args_info->target_given),
              &(local_args_info.target_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "target", 't',
              additional_error))
            goto failure;
        
          break;
        case 'a':	/* Location of the accessibility profiles.
.  */
        
        
          if (update_arg( (void *)&(args_info->accessibility_dir_arg), 
               &(args_info->accessibility_dir_orig), &(args_info->accessibility_dir_given),
              &(local_args_info.accessibility_dir_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "accessibility-dir", 'a',
              additional_error))
            goto failure;
        
          break;
        case 'b':	/* Allow the reading and parsing of memory dumped opening energy file
.  */
        
        
          if (update_arg((void *)&(args_info->binary_flag), 0, &(args_info->binary_given),
              &(local_args_info.binary_given), optarg, 0, 0, ARG_FLAG,
              check_ambiguity, override, 1, 0, "binary", 'b',
              additional_error))
            goto failure;
        
          break;
        case 'P':	/* Read energy parameters from paramfile, instead of using the default parameter set.
.  */
        
        
          if (update_arg( (void *)&(args_info->paramFile_arg), 
               &(args_info->paramFile_orig), &(args_info->paramFile_given),
              &(local_args_info.paramFile_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "paramFile", 'P',
              additional_error))
            goto failure;
        
          break;
        case 'T':	/* Rescale energy parameters to a temperature T. Default is 37C.
        
.  */
        
        
          if (update_arg( (void *)&(args_info->temp_arg), 
               &(args_info->temp_orig), &(args_info->temp_given),
              &(local_args_info.temp_given), optarg, 0, 0, ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "temp", 'T',
              additional_error))
            goto failure;
        
          break;
        case 'l':	/* Maximal length of an interaction
.  */
        
        
          if (update_arg( (void *)&(args_info->interaction_length_arg), 
               &(args_info->interaction_length_orig), &(args_info->interaction_length_given),
              &(local_args_info.interaction_length_given), optarg, 0, "40", ARG_INT,
              check_ambiguity, override, 0, 0,
              "interaction-length", 'l',
              additional_error))
            goto failure;
        
          break;
        case 'c':	/* Cost to add to each nucleotide in a duplex
.  */
        
        
          if (update_arg( (void *)&(args_info->extension_cost_arg), 
               &(args_info->extension_cost_orig), &(args_info->extension_cost_given),
              &(local_args_info.extension_cost_given), optarg, 0, "0", ARG_INT,
              check_ambiguity, override, 0, 0,
              "extension-cost", 'c',
              additional_error))
            goto failure;
        
          break;
        case 'p':	/* Compute Tm for probes.  */
        
        
          if (update_arg((void *)&(args_info->probe_mode_flag), 0, &(args_info->probe_mode_given),
              &(local_args_info.probe_mode_given), optarg, 0, 0, ARG_FLAG,
              check_ambiguity, override, 1, 0, "probe-mode", 'p',
              additional_error))
            goto failure;
        
          break;
        case 'Q':	/* Set the probe concentration for the Tm computation
        
.  */
        
        
          if (update_arg( (void *)&(args_info->probe_concentration_arg), 
               &(args_info->probe_concentration_orig), &(args_info->probe_concentration_given),
              &(local_args_info.probe_concentration_given), optarg, 0, "0.1", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "probe-concentration", 'Q',
              additional_error))
            goto failure;
        
          break;
        case 'N':	/* Set the Na+ concentration for the Tm computation
        
.  */
        
        
          if (update_arg( (void *)&(args_info->na_concentration_arg), 
               &(args_info->na_concentration_orig), &(args_info->na_concentration_given),
              &(local_args_info.na_concentration_given), optarg, 0, "1.0", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "na-concentration", 'N',
              additional_error))
            goto failure;
        
          break;
        case 'M':	/* Set the Mg2+ concentration for the Tm computation
        
.  */
        
        
          if (update_arg( (void *)&(args_info->mg_concentration_arg), 
               &(args_info->mg_concentration_orig), &(args_info->mg_concentration_given),
              &(local_args_info.mg_concentration_given), optarg, 0, "1.0", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "mg-concentration", 'M',
              additional_error))
            goto failure;
        
          break;
        case 'K':	/* Set the K+ concentration for the Tm computation
        
.  */
        
        
          if (update_arg( (void *)&(args_info->k_concentration_arg), 
               &(args_info->k_concentration_orig), &(args_info->k_concentration_given),
              &(local_args_info.k_concentration_given), optarg, 0, "1.0", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "k-concentration", 'K',
              additional_error))
            goto failure;
        
          break;
        case 'U':	/* Set the tris+ concentration for the Tm computation
        
.  */
        
        
          if (update_arg( (void *)&(args_info->tris_concentration_arg), 
               &(args_info->tris_concentration_orig), &(args_info->tris_concentration_given),
              &(local_args_info.tris_concentration_given), optarg, 0, "1.0", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "tris-concentration", 'U',
              additional_error))
            goto failure;
        
          break;
        case 'f':	/* Speedup of the target search
.  */
        
        
          if (update_arg( (void *)&(args_info->fast_folding_arg), 
               &(args_info->fast_folding_orig), &(args_info->fast_folding_given),
              &(local_args_info.fast_folding_given), optarg, 0, "0", ARG_INT,
              check_ambiguity, override, 0, 0,
              "fast-folding", 'f',
              additional_error))
            goto failure;
        
          break;
        case 'V':	/* Rescale all opening energy by a factor V
.  */
        
        
          if (update_arg( (void *)&(args_info->scale_accessibility_arg), 
               &(args_info->scale_accessibility_orig), &(args_info->scale_accessibility_given),
              &(local_args_info.scale_accessibility_given), optarg, 0, "1.0", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "scale-accessibility", 'V',
              additional_error))
            goto failure;
        
          break;
        case 'C':	/* Calculate structures subject to constraints.
.  */
        
        
          if (update_arg((void *)&(args_info->constraint_flag), 0, &(args_info->constraint_given),
              &(local_args_info.constraint_given), optarg, 0, 0, ARG_FLAG,
              check_ambiguity, override, 1, 0, "constraint", 'C',
              additional_error))
            goto failure;
        
          break;
        case 'A':	/* Tells RNAplex to compute interactions based on alignments
.  */
        
        
          if (update_arg((void *)&(args_info->alignment_mode_flag), 0, &(args_info->alignment_mode_given),
              &(local_args_info.alignment_mode_given), optarg, 0, 0, ARG_FLAG,
              check_ambiguity, override, 1, 0, "alignment-mode", 'A',
              additional_error))
            goto failure;
        
          break;
        case 'k':	/* If set, RNAplex will convert all opening energy file in a directory set by the -a option into binary opening energy files
.  */
        
        
          if (update_arg((void *)&(args_info->convert_to_bin_flag), 0, &(args_info->convert_to_bin_given),
              &(local_args_info.convert_to_bin_given), optarg, 0, 0, ARG_FLAG,
              check_ambiguity, override, 1, 0, "convert-to-bin", 'k',
              additional_error))
            goto failure;
        
          break;
        case 'z':	/* Distance between target 3' ends of two consecutive duplexes
.  */
        
        
          if (update_arg( (void *)&(args_info->duplex_distance_arg), 
               &(args_info->duplex_distance_orig), &(args_info->duplex_distance_given),
              &(local_args_info.duplex_distance_given), optarg, 0, "0", ARG_INT,
              check_ambiguity, override, 0, 0,
              "duplex-distance", 'z',
              additional_error))
            goto failure;
        
          break;
        case 'e':	/* Minimal energy for a duplex to be returned
.  */
        
        
          if (update_arg( (void *)&(args_info->energy_threshold_arg), 
               &(args_info->energy_threshold_orig), &(args_info->energy_threshold_given),
              &(local_args_info.energy_threshold_given), optarg, 0, "-100000", ARG_DOUBLE,
              check_ambiguity, override, 0, 0,
              "energy-threshold", 'e',
              additional_error))
            goto failure;
        
          break;
        case 'I':	/* Draw an alignment annotated interaction from RNAplex
.  */
        
        
          if (update_arg( (void *)&(args_info->produce_ps_arg), 
               &(args_info->produce_ps_orig), &(args_info->produce_ps_given),
              &(local_args_info.produce_ps_given), optarg, 0, 0, ARG_STRING,
              check_ambiguity, override, 0, 0,
              "produce-ps", 'I',
              additional_error))
            goto failure;
        
          break;
        case 'L':	/* Tells how large the region around the target site should be for redrawing the alignment interaction
.  */
        
        
          if (update_arg( (void *)&(args_info->WindowLength_arg), 
               &(args_info->WindowLength_orig), &(args_info->WindowLength_given),
              &(local_args_info.WindowLength_given), optarg, 0, "1", ARG_INT,
              check_ambiguity, override, 0, 0,
              "WindowLength", 'L',
              additional_error))
            goto failure;
        
          break;

        case 0:	/* Long option with no short option */
          if (strcmp (long_options[option_index].name, "detailed-help") == 0) {
            RNAplex_cmdline_parser_print_detailed_help ();
            RNAplex_cmdline_parser_free (&local_args_info);
            exit (EXIT_SUCCESS);
          }

          if (strcmp (long_options[option_index].name, "version") == 0) {
            RNAplex_cmdline_parser_print_version ();
            RNAplex_cmdline_parser_free (&local_args_info);
            exit (EXIT_SUCCESS);
          }

        case '?':	/* Invalid option.  */
          /* `getopt_long' already printed an error message.  */
          goto failure;

        default:	/* bug: option not considered.  */
          fprintf (stderr, "%s: option unknown: %c%s\n", RNAPLEX_CMDLINE_PARSER_PACKAGE, c, (additional_error ? additional_error : ""));
          abort ();
        } /* switch */
    } /* while */



	FIX_UNUSED(check_required);

  RNAplex_cmdline_parser_release (&local_args_info);

  if ( error_occurred )
    return (EXIT_FAILURE);

  return 0;

failure:
  
  RNAplex_cmdline_parser_release (&local_args_info);
  return (EXIT_FAILURE);
}
/* vim: set ft=c noet ts=8 sts=8 sw=8 tw=80 nojs spell : */
